// @stardust-ui/react/dist/commonjs/components/Status/Status.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _accessibility = require("@stardust-ui/accessibility");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _lib = require("../../lib");

var _types = require("../../types");

var Status =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Status, _UIComponent);

  function Status() {
    (0, _classCallCheck2.default)(this, Status);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Status).apply(this, arguments));
  }

  (0, _createClass2.default)(Status, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          variables = _ref.variables,
          styles = _ref.styles;
      var _ref2 = this.props,
          icon = _ref2.icon;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps), _Icon.default.create(icon, {
        defaultProps: function defaultProps() {
          return {
            size: 'smallest',
            styles: styles.icon,
            variables: variables.icon,
            xSpacing: 'none'
          };
        }
      }));
    }
  }]);
  return Status;
}(_lib.UIComponent);

(0, _defineProperty2.default)(Status, "create", void 0);
(0, _defineProperty2.default)(Status, "className", 'ui-status');
(0, _defineProperty2.default)(Status, "displayName", 'Status');
(0, _defineProperty2.default)(Status, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  children: false,
  content: false
}), {
  color: PropTypes.string,
  icon: customPropTypes.itemShorthandWithoutJSX,
  size: customPropTypes.size,
  state: PropTypes.oneOf(['success', 'info', 'warning', 'error', 'unknown'])
}));
(0, _defineProperty2.default)(Status, "defaultProps", {
  accessibility: _accessibility.statusBehavior,
  as: 'span',
  size: 'medium',
  state: 'unknown'
});
Status.create = (0, _lib.createShorthandFactory)({
  Component: Status,
  mappedProp: 'state'
});
/**
 * A Status represents someone's or something's state.
 *
 * @accessibility
 * Implements [ARIA img](https://www.w3.org/TR/wai-aria-1.1/#img) role.
 */

var _default = (0, _types.withSafeTypeForAs)(Status);

exports.default = _default;
