// @stardust-ui/accessibility/dist/commonjs/behaviors/Tooltip/tooltipBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

/**
 * @description
 * Implements ARIA Tooltip design pattern.
 *
 * @specification
 * Adds attribute 'role=tooltip' to 'tooltip' slot.
 * Adds attribute 'aria-hidden=false' to 'tooltip' slot if 'open' property is true. Sets the attribute to 'true' otherwise.
 * Adds attribute 'aria-describedby' based on the property 'aria-describedby' to 'trigger' slot.
 * Triggers 'close' action with 'Escape' on 'trigger'.
 */
var tooltipBehavior = function tooltipBehavior(props) {
  var defaultAriaDescribedBy = getDefaultAriaDescribedBy(props);
  return {
    attributes: {
      trigger: {
        'aria-describedby': defaultAriaDescribedBy || props['aria-describedby']
      },
      tooltip: {
        role: 'tooltip',
        id: defaultAriaDescribedBy,
        'aria-hidden': !props.open
      }
    },
    keyActions: {
      trigger: {
        close: {
          keyCombinations: [{
            keyCode: keyboardKey.Escape
          }]
        }
      }
    }
  };
};

var _default = tooltipBehavior;
/**
 * Returns the element id of the tooltip, it is used when user does not provide aria-describedby as props.
 */

exports.default = _default;

var getDefaultAriaDescribedBy = function getDefaultAriaDescribedBy(props) {
  if (props['aria-describedby']) {
    return undefined;
  }

  return props.contentId;
};
