// @stardust-ui/react/dist/commonjs/components/Menu/Menu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _filter2 = _interopRequireDefault(require("lodash/filter"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _mergeThemes = require("../../lib/mergeThemes");

var _MenuItem = _interopRequireDefault(require("./MenuItem"));

var _types = require("../../types");

var _MenuDivider = _interopRequireDefault(require("./MenuDivider"));

var Menu =
/*#__PURE__*/
function (_AutoControlledCompon) {
  (0, _inherits2.default)(Menu, _AutoControlledCompon);

  function Menu() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Menu);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Menu)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleItemOverrides", function (variables) {
      return function (predefinedProps) {
        return {
          onClick: function onClick(e, itemProps) {
            var index = itemProps.index;

            _this.setState({
              activeIndex: index
            });

            (0, _invoke2.default)(_this.props, 'onItemClick', e, itemProps);
            (0, _invoke2.default)(predefinedProps, 'onClick', e, itemProps);
          },
          onActiveChanged: function onActiveChanged(e, props) {
            var index = props.index,
                active = props.active;

            if (active) {
              _this.setState({
                activeIndex: index
              });
            } else if (_this.state.activeIndex === index) {
              _this.setState({
                activeIndex: null
              });
            }

            (0, _invoke2.default)(predefinedProps, 'onActiveChanged', e, props);
          },
          variables: (0, _mergeThemes.mergeComponentVariables)(variables, predefinedProps.variables)
        };
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDividerOverrides", function (variables) {
      return function (predefinedProps) {
        return {
          variables: (0, _mergeThemes.mergeComponentVariables)(variables, predefinedProps.variables)
        };
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderItems", function (styles, variables, accessibility) {
      var _this$props = _this.props,
          iconOnly = _this$props.iconOnly,
          items = _this$props.items,
          pills = _this$props.pills,
          pointing = _this$props.pointing,
          primary = _this$props.primary,
          secondary = _this$props.secondary,
          underlined = _this$props.underlined,
          vertical = _this$props.vertical,
          submenu = _this$props.submenu,
          indicator = _this$props.indicator;
      var activeIndex = _this.state.activeIndex;
      var itemsCount = (0, _filter2.default)(items, function (item) {
        return (0, _lib.getKindProp)(item, 'item') !== 'divider';
      }).length;
      var itemPosition = 0;

      var overrideItemProps = _this.handleItemOverrides(variables);

      var overrideDividerProps = _this.handleDividerOverrides(variables);

      return (0, _map2.default)(items, function (item, index) {
        var active = (typeof activeIndex === 'string' ? parseInt(activeIndex, 10) : activeIndex) === index;
        var kind = (0, _lib.getKindProp)(item, 'item');

        if (kind === 'divider') {
          return _MenuDivider.default.create(item, {
            defaultProps: function defaultProps() {
              return {
                className: Menu.slotClassNames.divider,
                primary: primary,
                secondary: secondary,
                vertical: vertical,
                styles: styles.divider,
                inSubmenu: submenu,
                accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.divider : undefined
              };
            },
            overrideProps: overrideDividerProps
          });
        }

        itemPosition++;
        return _MenuItem.default.create(item, {
          defaultProps: function defaultProps() {
            return {
              className: Menu.slotClassNames.item,
              iconOnly: iconOnly,
              pills: pills,
              pointing: pointing,
              primary: primary,
              secondary: secondary,
              underlined: underlined,
              vertical: vertical,
              index: index,
              itemPosition: itemPosition,
              itemsCount: itemsCount,
              active: active,
              inSubmenu: submenu,
              indicator: indicator,
              accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.item : undefined
            };
          },
          overrideProps: overrideItemProps
        });
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Menu, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          styles = _ref.styles,
          variables = _ref.variables,
          unhandledProps = _ref.unhandledProps;
      var children = this.props.children;
      return React.createElement(ElementType, (0, _extends2.default)({}, accessibility.attributes.root, _lib.rtlTextContainer.getAttributes({
        forElements: [children]
      }), unhandledProps, {
        className: classes.root
      }), (0, _lib.childrenExist)(children) ? children : this.renderItems(styles, variables, accessibility));
    }
  }]);
  return Menu;
}(_lib.AutoControlledComponent);

(0, _defineProperty2.default)(Menu, "displayName", 'Menu');
(0, _defineProperty2.default)(Menu, "className", 'ui-menu');
(0, _defineProperty2.default)(Menu, "slotClassNames", {
  divider: "".concat(Menu.className, "__divider"),
  item: "".concat(Menu.className, "__item")
});
(0, _defineProperty2.default)(Menu, "create", void 0);
(0, _defineProperty2.default)(Menu, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  activeIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  defaultActiveIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  fluid: PropTypes.bool,
  iconOnly: PropTypes.bool,
  items: customPropTypes.collectionShorthandWithKindProp(['divider', 'item']),
  onItemClick: PropTypes.func,
  pills: PropTypes.bool,
  pointing: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['start', 'end'])]),
  primary: customPropTypes.every([customPropTypes.disallow(['secondary']), PropTypes.bool]),
  secondary: customPropTypes.every([customPropTypes.disallow(['primary']), PropTypes.bool]),
  underlined: PropTypes.bool,
  vertical: PropTypes.bool,
  submenu: PropTypes.bool,
  indicator: customPropTypes.itemShorthandWithoutJSX
}));
(0, _defineProperty2.default)(Menu, "defaultProps", {
  as: 'ul',
  accessibility: _accessibility.menuBehavior
});
(0, _defineProperty2.default)(Menu, "autoControlledProps", ['activeIndex']);
(0, _defineProperty2.default)(Menu, "Item", _MenuItem.default);
(0, _defineProperty2.default)(Menu, "Divider", _MenuDivider.default);
Menu.create = (0, _lib.createShorthandFactory)({
  Component: Menu,
  mappedArrayProp: 'items'
});
/**
 * A Menu is a component that offers a grouped list of choices to the user.
 *
 * @accessibility
 * Implements ARIA [Menu](https://www.w3.org/TR/wai-aria-practices-1.1/#menu), [Toolbar](https://www.w3.org/TR/wai-aria-practices-1.1/#toolbar) or [Tabs](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel) design pattern, depending on the behavior used.
 * @accessibilityIssues
 * [JAWS - navigation instruction for menubar](https://github.com/FreedomScientific/VFO-standards-support/issues/203)
 * [JAWS - navigation instruction for menu with aria-orientation="horizontal"](https://github.com/FreedomScientific/VFO-standards-support/issues/204)
 * [JAWS [VC] doesn't narrate menu item, when it is open from menu button](https://github.com/FreedomScientific/VFO-standards-support/issues/324)
 * [JAWS [app mode] focus moves to second menu item, when it is open from menu button](https://github.com/FreedomScientific/VFO-standards-support/issues/325)
 * [Enter into a tablist JAWS narrates: To switch pages, press Control+PageDown](https://github.com/FreedomScientific/VFO-standards-support/issues/337)
 * 51114083 VoiceOver+Web narrate wrong position in menu / total count of menu items, when pseudo element ::after or ::before is used
 */

var _default = (0, _types.withSafeTypeForAs)(Menu);

exports.default = _default;
