// @stardust-ui/react/dist/es/themes/teams/components/Text/textStyles.js
import { getColorSchemeKey } from '../../colors';
import translateAlignProp from '../../../../styles/translateAlignProp';
export default {
  root: function root(_ref) {
    var _ref$props = _ref.props,
        as = _ref$props.as,
        atMention = _ref$props.atMention,
        color = _ref$props.color,
        important = _ref$props.important,
        timestamp = _ref$props.timestamp,
        animation = _ref$props.animation,
        truncated = _ref$props.truncated,
        disabled = _ref$props.disabled,
        error = _ref$props.error,
        success = _ref$props.success,
        temporary = _ref$props.temporary,
        align = _ref$props.align,
        weight = _ref$props.weight,
        size = _ref$props.size,
        v = _ref.variables;
    var colors = v.colorScheme[getColorSchemeKey(color)];
    return Object.assign({}, color && {
      color: colors.foreground
    }, animation && as === 'span' && {
      display: 'inline-block'
    }, atMention === true && {
      color: v.atMentionOtherColor
    }, truncated && {
      overflow: 'hidden',
      textOverflow: 'ellipsis',
      whiteSpace: 'nowrap'
    }, disabled && {
      color: v.disabledColor
    }, error && {
      color: v.errorColor
    }, success && {
      color: v.successColor
    }, temporary && {
      fontStyle: 'italic'
    }, align && {
      display: 'block',
      textAlign: translateAlignProp(align)
    }, weight === 'light' && {
      fontWeight: v.fontWeightLight
    }, weight === 'semilight' && {
      fontWeight: v.fontWeightSemilight
    }, weight === 'regular' && {
      fontWeight: v.fontWeightRegular
    }, weight === 'semibold' && {
      fontWeight: v.fontWeightSemibold
    }, weight === 'bold' && {
      fontWeight: v.fontWeightBold
    }, size === 'smallest' && {
      fontSize: v.fontSizeSmallest,
      lineHeight: v.fontLineHeightSmallest
    }, size === 'smaller' && {
      fontSize: v.fontSizeSmaller,
      lineHeight: v.fontLineHeightSmaller
    }, size === 'small' && {
      fontSize: v.fontSizeSmall,
      lineHeight: v.fontLineHeightSmall
    }, size === 'medium' && {
      fontSize: v.fontSizeMedium,
      lineHeight: v.fontLineHeightMedium
    }, size === 'large' && {
      fontSize: v.fontSizeLarge,
      lineHeight: v.fontLineHeightLarge
    }, size === 'larger' && {
      fontSize: v.fontSizeLarger,
      lineHeight: v.fontLineHeightLarger
    }, size === 'largest' && {
      fontSize: v.fontSizeLargest,
      lineHeight: v.fontLineHeightLargest
    }, atMention === 'me' && {
      color: v.atMentionMeColor,
      fontWeight: v.atMentionMeFontWeight
    }, timestamp && {
      color: v.timestampColor,
      ':hover': {
        color: v.timestampHoverColor
      }
    }, important && {
      color: v.importantColor,
      fontWeight: v.importantWeight
    });
  }
};
