// @stardust-ui/react/dist/es/themes/teams/components/Reaction/reactionStyles.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { default as Reaction } from '../../../../components/Reaction/Reaction';
import { pxToRem } from '../../../../lib';
var contentClassNameSelector = "& .".concat(Reaction.slotClassNames.content);
var reactionStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return {
      cursor: 'pointer',
      background: 'transparent',
      border: pxToRem(0),
      padding: pxToRem(0),
      display: 'inline-flex',
      justifyContent: 'center',
      alignItems: 'center',
      color: v.meReacting ? v.meReactingColor : v.otherReactingColor,
      ':hover': _defineProperty({
        color: v.meReacting ? v.meReactingColorHover : v.otherReactingColorHover
      }, contentClassNameSelector, {
        fontWeight: v.fontWeightHover
      }),
      position: 'relative',
      ':focus': {
        outline: 'none'
      },
      ':focus-visible': {
        ':after': {
          content: '""',
          position: 'absolute',
          top: "-".concat(pxToRem(2)),
          right: "-".concat(pxToRem(2)),
          bottom: "-".concat(pxToRem(2)),
          left: "-".concat(pxToRem(2)),
          borderWidth: '1px',
          borderStyle: 'solid',
          borderColor: v.borderColorFocus,
          borderRadius: pxToRem(2),
          boxShadow: "0px 0px 0px 1px ".concat(v.boxShadowColor, " inset")
        }
      }
    };
  },
  icon: function icon(_ref2) {
    var p = _ref2.props;
    return {
      marginRight: p.content ? pxToRem(4) : pxToRem(0)
    };
  },
  content: function content(_ref3) {
    var v = _ref3.variables;
    return {
      fontSize: v.contentFontSize
    };
  }
};
export default reactionStyles;
