// @stardust-ui/react/dist/es/components/Segment/Segment.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';
import * as PropTypes from 'prop-types';
import { UIComponent, childrenExist, commonPropTypes, rtlTextContainer } from '../../lib';
import { withSafeTypeForAs } from '../../types';
import Box from '../Box/Box';

var Segment =
/*#__PURE__*/
function (_UIComponent) {
  _inherits(Segment, _UIComponent);

  function Segment() {
    _classCallCheck(this, Segment);

    return _possibleConstructorReturn(this, _getPrototypeOf(Segment).apply(this, arguments));
  }

  _createClass(Segment, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content;
      return React.createElement(ElementType, _extends({}, rtlTextContainer.getAttributes({
        forElements: [children]
      }), accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), childrenExist(children) ? children : Box.create(content));
    }
  }]);

  return Segment;
}(UIComponent);
/**
 * A Segment visually groups related content.
 */


_defineProperty(Segment, "className", 'ui-segment');

_defineProperty(Segment, "displayName", 'Segment');

_defineProperty(Segment, "propTypes", Object.assign({}, commonPropTypes.createCommon({
  content: 'shorthand',
  color: true
}), {
  disabled: PropTypes.bool,
  inverted: PropTypes.bool,
  rtlAttributes: PropTypes.func
}));

_defineProperty(Segment, "defaultProps", {
  as: 'div'
});

export default withSafeTypeForAs(Segment);
