// @stardust-ui/react/dist/commonjs/themes/teams/components/Loader/loaderVariables.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lib = require("../../../../lib");

var _loaderSvgDataUrl = _interopRequireDefault(require("./loaderSvgDataUrl"));

var _default = function _default() {
  return {
    containerHeights: {
      smallest: (0, _lib.pxToRem)(24),
      smaller: (0, _lib.pxToRem)(24),
      small: (0, _lib.pxToRem)(24),
      medium: (0, _lib.pxToRem)(36),
      large: (0, _lib.pxToRem)(72),
      larger: (0, _lib.pxToRem)(72),
      largest: (0, _lib.pxToRem)(72)
    },
    containerWidths: {
      smallest: (0, _lib.pxToRem)(24),
      smaller: (0, _lib.pxToRem)(24),
      small: (0, _lib.pxToRem)(24),
      medium: (0, _lib.pxToRem)(36),
      large: (0, _lib.pxToRem)(72),
      larger: (0, _lib.pxToRem)(72),
      largest: (0, _lib.pxToRem)(72)
    },
    svgContent: _loaderSvgDataUrl.default,
    svgHeights: {
      smaller: (0, _lib.pxToRem)(1464),
      smallest: (0, _lib.pxToRem)(1464),
      small: (0, _lib.pxToRem)(1464),
      medium: (0, _lib.pxToRem)(2196),
      large: (0, _lib.pxToRem)(4392),
      larger: (0, _lib.pxToRem)(4392),
      largest: (0, _lib.pxToRem)(4392)
    },
    svgTranslatePosition: {
      smallest: (0, _lib.pxToRem)(-1440),
      smaller: (0, _lib.pxToRem)(-1440),
      small: (0, _lib.pxToRem)(-1440),
      medium: (0, _lib.pxToRem)(-2160),
      large: (0, _lib.pxToRem)(-4320),
      larger: (0, _lib.pxToRem)(-4320),
      largest: (0, _lib.pxToRem)(-4320)
    },
    svgWidths: {
      smallest: (0, _lib.pxToRem)(24),
      smaller: (0, _lib.pxToRem)(24),
      small: (0, _lib.pxToRem)(24),
      medium: (0, _lib.pxToRem)(36),
      large: (0, _lib.pxToRem)(72),
      larger: (0, _lib.pxToRem)(72),
      largest: (0, _lib.pxToRem)(72)
    }
  };
};

exports.default = _default;
