// @stardust-ui/react/dist/commonjs/components/Tree/TreeItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _reactComponentRef = require("@stardust-ui/react-component-ref");

var _TreeTitle = _interopRequireDefault(require("./TreeTitle"));

var _lib = require("../../lib");

var _types = require("../../types");

var _lib2 = require("./lib");

var TreeItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(TreeItem, _UIComponent);

  function TreeItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, TreeItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(TreeItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "state", {
      hasSubtree: false,
      treeSize: 0
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      performClick: function performClick(e) {
        e.preventDefault();
        e.stopPropagation();

        _this.handleTitleClick(e);
      },
      focusParent: function focusParent(e) {
        e.preventDefault();
        e.stopPropagation();
        (0, _invoke2.default)(_this.props, 'onFocusParent', e, _this.props);
      },
      collapse: function collapse(e) {
        e.preventDefault();
        e.stopPropagation();

        _this.handleTitleClick(e);
      },
      expand: function expand(e) {
        e.preventDefault();
        e.stopPropagation();

        _this.handleTitleClick(e);
      },
      focusFirstChild: function focusFirstChild(e) {
        e.preventDefault();
        e.stopPropagation();
        (0, _invoke2.default)(_this.props, 'onFocusFirstChild', e, _this.props);
      },
      expandSiblings: function expandSiblings(e) {
        e.preventDefault();
        e.stopPropagation();
        (0, _invoke2.default)(_this.props, 'onSiblingsExpand', e, _this.props);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleTitleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onTitleClick', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleTitleOverrides", function (predefinedProps) {
      return {
        onClick: function onClick(e, titleProps) {
          _this.handleTitleClick(e);

          (0, _invoke2.default)(predefinedProps, 'onClick', e, titleProps);
        }
      };
    });
    return _this;
  }

  (0, _createClass2.default)(TreeItem, [{
    key: "renderContent",
    value: function renderContent(accessibility) {
      var _this$props = this.props,
          title = _this$props.title,
          renderItemTitle = _this$props.renderItemTitle,
          open = _this$props.open,
          level = _this$props.level,
          index = _this$props.index;
      var _this$state = this.state,
          hasSubtree = _this$state.hasSubtree,
          treeSize = _this$state.treeSize;
      return _TreeTitle.default.create(title, {
        defaultProps: function defaultProps() {
          return {
            className: TreeItem.slotClassNames.title,
            open: open,
            hasSubtree: hasSubtree,
            as: hasSubtree ? 'span' : 'a',
            level: level,
            treeSize: treeSize,
            index: index,
            accessibility: accessibility.childBehaviors ? accessibility.childBehaviors.title : undefined
          };
        },
        render: renderItemTitle,
        overrideProps: this.handleTitleOverrides
      });
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          accessibility = _ref.accessibility,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props2 = this.props,
          children = _this$props2.children,
          contentRef = _this$props2.contentRef;
      var element = React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, accessibility.attributes.root, _lib.rtlTextContainer.getAttributes({
        forElements: [children]
      }), unhandledProps, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), (0, _lib.childrenExist)(children) ? children : this.renderContent(accessibility));
      return contentRef ? React.createElement(_reactComponentRef.Ref, {
        innerRef: contentRef
      }, element) : element;
    }
  }], [{
    key: "getDerivedStateFromProps",
    value: function getDerivedStateFromProps(props) {
      return {
        hasSubtree: (0, _lib2.hasSubtree)(props),
        treeSize: props.siblings.length + 1
      };
    }
  }]);
  return TreeItem;
}(_lib.UIComponent);

(0, _defineProperty2.default)(TreeItem, "create", void 0);
(0, _defineProperty2.default)(TreeItem, "displayName", 'TreeItem');
(0, _defineProperty2.default)(TreeItem, "className", 'ui-tree__item');
(0, _defineProperty2.default)(TreeItem, "slotClassNames", {
  title: "".concat(TreeItem.className, "__title"),
  subtree: "".concat(TreeItem.className, "__subtree")
});
(0, _defineProperty2.default)(TreeItem, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  contentRef: customPropTypes.ref,
  id: PropTypes.string.isRequired,
  index: PropTypes.number,
  items: customPropTypes.collectionShorthand,
  level: PropTypes.number,
  onTitleClick: PropTypes.func,
  onFocusFirstChild: PropTypes.func,
  onFocusParent: PropTypes.func,
  onSiblingsExpand: PropTypes.func,
  open: PropTypes.bool,
  parent: customPropTypes.itemShorthand,
  renderItemTitle: PropTypes.func,
  siblings: customPropTypes.collectionShorthand,
  title: customPropTypes.itemShorthand
}));
(0, _defineProperty2.default)(TreeItem, "defaultProps", {
  as: 'div',
  accessibility: _accessibility.treeItemBehavior
});
TreeItem.create = (0, _lib.createShorthandFactory)({
  Component: TreeItem,
  mappedProp: 'title'
});
/**
 * A TreeItem renders an item of a Tree.
 *
 * @accessibility
 * Implements [ARIA TreeView](https://www.w3.org/TR/wai-aria-practices-1.1/#TreeView) design pattern.
 */

var _default = (0, _types.withSafeTypeForAs)(TreeItem);

exports.default = _default;
