// @stardust-ui/accessibility/dist/commonjs/behaviors/HierarchicalTree/hierarchicalTreeTitleBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

var _attributes = require("../../attributes");

/**
 * @description
 * Adds role 'treeitem' if the title is a leaf node inside the tree.
 * Adds 'tabIndex' as '-1' if the title is a leaf node inside the tree.
 *
 * @specification
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var hierarchicalTreeTitleBehavior = function hierarchicalTreeTitleBehavior(props) {
  var _ref;

  return {
    attributes: {
      root: Object.assign({}, !props.hasSubtree && (_ref = {
        tabIndex: -1
      }, (0, _defineProperty2.default)(_ref, _attributes.IS_FOCUSABLE_ATTRIBUTE, true), (0, _defineProperty2.default)(_ref, "role", 'treeitem'), _ref))
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Enter
          }, {
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

var _default = hierarchicalTreeTitleBehavior;
exports.default = _default;
