// @stardust-ui/accessibility/dist/commonjs/behaviors/HierarchicalTree/hierarchicalTreeBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _types = require("../../focusZone/types");

var _hierarchicalSubtreeBehavior = _interopRequireDefault(require("./hierarchicalSubtreeBehavior"));

/**
 * @specification
 * Adds role 'tree' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in vertical direction.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 */
var hierarchicalTreeBehavior = function hierarchicalTreeBehavior(props) {
  var subtreeBehaviorData = (0, _hierarchicalSubtreeBehavior.default)({});
  return {
    attributes: {
      root: Object.assign({}, subtreeBehaviorData.attributes.root, {
        role: 'tree',
        'aria-labelledby': props['aria-labelledby']
      })
    },
    keyActions: {
      root: Object.assign({}, subtreeBehaviorData.keyActions.root)
    },
    focusZone: {
      mode: _types.FocusZoneMode.Embed,
      props: {
        direction: _types.FocusZoneDirection.vertical
      }
    }
  };
};

var _default = hierarchicalTreeBehavior;
exports.default = _default;
