// @stardust-ui/react/dist/commonjs/themes/teams-high-contrast/componentStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Attachment", {
  enumerable: true,
  get: function get() {
    return _attachmentStyles.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _buttonStyles.default;
  }
});
Object.defineProperty(exports, "MenuItem", {
  enumerable: true,
  get: function get() {
    return _menuItemStyles.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function get() {
    return _alertStyles.default;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function get() {
    return _dialogStyles.default;
  }
});
Object.defineProperty(exports, "Dropdown", {
  enumerable: true,
  get: function get() {
    return _dropdownStyles.default;
  }
});

var _attachmentStyles = _interopRequireDefault(require("./components/Attachment/attachmentStyles"));

var _buttonStyles = _interopRequireDefault(require("./components/Button/buttonStyles"));

var _menuItemStyles = _interopRequireDefault(require("./components/Menu/menuItemStyles"));

var _alertStyles = _interopRequireDefault(require("./components/Alert/alertStyles"));

var _dialogStyles = _interopRequireDefault(require("./components/Dialog/dialogStyles"));

var _dropdownStyles = _interopRequireDefault(require("./components/Dropdown/dropdownStyles"));
