// @stardust-ui/react/dist/commonjs/themes/teams/components/Popup/popupContentStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _getPointerStyles = _interopRequireDefault(require("../../getPointerStyles"));

var popupContentStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        rtl = _ref.rtl;
    return Object.assign({
      border: "".concat(v.borderSize, " solid ").concat(v.borderColor),
      borderRadius: v.borderRadius,
      boxShadow: v.boxShadow,
      display: 'block'
    }, p.pointing && (0, _getPointerStyles.default)(v.pointerOffset, v.pointerMargin, rtl, p.placement).root);
  },
  pointer: function pointer(_ref2) {
    var p = _ref2.props,
        v = _ref2.variables,
        rtl = _ref2.rtl;
    return Object.assign({
      display: 'block',
      position: 'absolute',
      backgroundColor: 'inherit',
      borderBottom: "".concat(v.borderSize, " solid ").concat(v.borderColor),
      borderRight: "".concat(v.borderSize, " solid ").concat(v.borderColor),
      height: v.pointerSize,
      width: v.pointerSize
    }, (0, _getPointerStyles.default)(v.pointerOffset, v.pointerMargin, rtl, p.placement).pointer);
  },
  content: function content(_ref3) {
    var p = _ref3.props,
        v = _ref3.variables;
    return {
      display: 'block',
      padding: v.padding
    };
  }
};
var _default = popupContentStyles;
exports.default = _default;
