// @stardust-ui/react/dist/commonjs/themes/teams/components/Reaction/reactionStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Reaction = _interopRequireDefault(require("../../../../components/Reaction/Reaction"));

var _lib = require("../../../../lib");

var contentClassNameSelector = "& .".concat(_Reaction.default.slotClassNames.content);
var reactionStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return {
      cursor: 'pointer',
      background: 'transparent',
      border: (0, _lib.pxToRem)(0),
      padding: (0, _lib.pxToRem)(0),
      display: 'inline-flex',
      justifyContent: 'center',
      alignItems: 'center',
      color: v.meReacting ? v.meReactingColor : v.otherReactingColor,
      ':hover': (0, _defineProperty2.default)({
        color: v.meReacting ? v.meReactingColorHover : v.otherReactingColorHover
      }, contentClassNameSelector, {
        fontWeight: v.fontWeightHover
      }),
      position: 'relative',
      ':focus': {
        outline: 'none'
      },
      ':focus-visible': {
        ':after': {
          content: '""',
          position: 'absolute',
          top: "-".concat((0, _lib.pxToRem)(2)),
          right: "-".concat((0, _lib.pxToRem)(2)),
          bottom: "-".concat((0, _lib.pxToRem)(2)),
          left: "-".concat((0, _lib.pxToRem)(2)),
          borderWidth: '1px',
          borderStyle: 'solid',
          borderColor: v.borderColorFocus,
          borderRadius: (0, _lib.pxToRem)(2),
          boxShadow: "0px 0px 0px 1px ".concat(v.boxShadowColor, " inset")
        }
      }
    };
  },
  icon: function icon(_ref2) {
    var p = _ref2.props;
    return {
      marginRight: p.content ? (0, _lib.pxToRem)(4) : (0, _lib.pxToRem)(0)
    };
  },
  content: function content(_ref3) {
    var v = _ref3.variables;
    return {
      fontSize: v.contentFontSize
    };
  }
};
var _default = reactionStyles;
exports.default = _default;
