// @stardust-ui/react/dist/commonjs/components/Video/Video.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _reactComponentRef = require("@stardust-ui/react-component-ref");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var Video =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Video, _UIComponent);

  function Video() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Video);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Video)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "videoRef", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setVideoAttributes", function () {
      // React doesn't guaranty that props will be set:
      // https://github.com/facebook/react/issues/10389
      if (_this.videoRef.current) {
        _this.videoRef.current.muted = !!_this.props.muted;
      }
    });
    return _this;
  }

  (0, _createClass2.default)(Video, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.setVideoAttributes();
    }
  }, {
    key: "componentDidUpdate",
    value: function componentDidUpdate() {
      this.setVideoAttributes();
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          controls = _this$props.controls,
          autoPlay = _this$props.autoPlay,
          loop = _this$props.loop,
          poster = _this$props.poster,
          src = _this$props.src;
      return React.createElement(_reactComponentRef.Ref, {
        innerRef: this.videoRef
      }, React.createElement(ElementType, (0, _extends2.default)({
        autoPlay: autoPlay,
        className: classes.root,
        controls: controls,
        loop: loop,
        poster: poster,
        src: src
      }, accessibility.attributes.root, unhandledProps)));
    }
  }]);
  return Video;
}(_lib.UIComponent);

(0, _defineProperty2.default)(Video, "create", void 0);
(0, _defineProperty2.default)(Video, "className", 'ui-video');
(0, _defineProperty2.default)(Video, "displayName", 'Video');
(0, _defineProperty2.default)(Video, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  children: false,
  content: false
}), {
  autoPlay: PropTypes.bool,
  controls: PropTypes.bool,
  loop: PropTypes.bool,
  muted: PropTypes.bool,
  poster: PropTypes.string,
  src: PropTypes.string
}));
(0, _defineProperty2.default)(Video, "defaultProps", {
  as: 'video',
  controls: true,
  autoPlay: false,
  muted: false
});
Video.create = (0, _lib.createShorthandFactory)({
  Component: Video,
  mappedProp: 'src'
});
/**
 * A Video provides ability to embed video content.
 */

var _default = (0, _types.withSafeTypeForAs)(Video);

exports.default = _default;
