// @stardust-ui/react/dist/commonjs/components/Input/Input.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var _reactComponentRef = require("@stardust-ui/react-component-ref");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _lib = require("../../lib");

var _types = require("../../types");

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Box = _interopRequireDefault(require("../Box/Box"));

var Input =
/*#__PURE__*/
function (_AutoControlledCompon) {
  (0, _inherits2.default)(Input, _AutoControlledCompon);

  function Input() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Input);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Input)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "inputRef", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      clear: function clear(e) {
        if (_this.props.clearable && _this.state.value !== '') {
          e.stopPropagation();
          e.nativeEvent && e.nativeEvent.stopPropagation();

          _this.handleOnClear(e);
        }
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleIconOverrides", function (predefinedProps) {
      return {
        onClick: function onClick(e) {
          _this.handleOnClear(e);

          _this.inputRef.current.focus();

          (0, _invoke2.default)(predefinedProps, 'onClick', e, _this.props);
        }
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleChange", function (e) {
      var value = (0, _get2.default)(e, 'target.value');
      (0, _invoke2.default)(_this.props, 'onChange', e, Object.assign({}, _this.props, {
        value: value
      }));

      _this.setState({
        value: value
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleOnClear", function (e) {
      if (_this.props.clearable) {
        (0, _invoke2.default)(_this.props, 'onChange', e, Object.assign({}, _this.props, {
          value: ''
        }));

        _this.setState({
          value: ''
        });
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "computeIcon", function () {
      var _this$props = _this.props,
          clearable = _this$props.clearable,
          icon = _this$props.icon;
      var value = _this.state.value;

      if (clearable && value.length !== 0) {
        return 'stardust-close';
      }

      return icon || null;
    });
    return _this;
  }

  (0, _createClass2.default)(Input, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var _this2 = this;

      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles,
          variables = _ref.variables;
      var _this$props2 = this.props,
          className = _this$props2.className,
          input = _this$props2.input,
          inputRef = _this$props2.inputRef,
          type = _this$props2.type,
          wrapper = _this$props2.wrapper;
      var _this$state$value = this.state.value,
          value = _this$state$value === void 0 ? '' : _this$state$value;

      var _partitionHTMLProps = (0, _lib.partitionHTMLProps)(unhandledProps),
          _partitionHTMLProps2 = (0, _slicedToArray2.default)(_partitionHTMLProps, 2),
          htmlInputProps = _partitionHTMLProps2[0],
          restProps = _partitionHTMLProps2[1];

      return _Box.default.create(wrapper, {
        defaultProps: function defaultProps() {
          return Object.assign({}, accessibility.attributes.root, {
            className: (0, _classnames.default)(Input.className, className),
            children: React.createElement(React.Fragment, null, React.createElement(_reactComponentRef.Ref, {
              innerRef: function innerRef(inputElement) {
                (0, _reactComponentRef.handleRef)(_this2.inputRef, inputElement);
                (0, _reactComponentRef.handleRef)(inputRef, inputElement);
              }
            }, _Box.default.create(input || type, {
              defaultProps: function defaultProps() {
                return Object.assign({}, htmlInputProps, {
                  as: 'input',
                  type: type,
                  value: value,
                  className: Input.slotClassNames.input,
                  styles: styles.input,
                  onChange: _this2.handleChange
                }, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.input, htmlInputProps));
              }
            })), _Icon.default.create(_this2.computeIcon(), {
              defaultProps: function defaultProps() {
                return {
                  styles: styles.icon,
                  variables: variables.icon
                };
              },
              overrideProps: _this2.handleIconOverrides
            })),
            styles: styles.root
          }, restProps);
        },
        overrideProps: {
          as: wrapper && wrapper.as || ElementType
        }
      });
    }
  }]);
  return Input;
}(_lib.AutoControlledComponent);

(0, _defineProperty2.default)(Input, "className", 'ui-input');
(0, _defineProperty2.default)(Input, "displayName", 'Input');
(0, _defineProperty2.default)(Input, "slotClassNames", void 0);
(0, _defineProperty2.default)(Input, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  clearable: PropTypes.bool,
  defaultValue: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  fluid: PropTypes.bool,
  icon: customPropTypes.itemShorthandWithoutJSX,
  iconPosition: PropTypes.oneOf(['start', 'end']),
  input: customPropTypes.itemShorthand,
  inputRef: customPropTypes.ref,
  inline: PropTypes.bool,
  inverted: PropTypes.bool,
  onChange: PropTypes.func,
  type: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  wrapper: customPropTypes.wrapperShorthand
}));
(0, _defineProperty2.default)(Input, "defaultProps", {
  accessibility: _accessibility.inputBehavior,
  type: 'text',
  wrapper: {},
  iconPosition: 'end'
});
(0, _defineProperty2.default)(Input, "autoControlledProps", ['value']);
Input.slotClassNames = {
  input: "".concat(Input.className, "__input")
  /**
   * An Input is a field used to elicit an input from a user.
   *
   * @accessibility
   * For good screen reader experience set `aria-label` or `aria-labelledby` attribute for input.
   */

};

var _default = (0, _types.withSafeTypeForAs)(Input);

exports.default = _default;
