// @stardust-ui/accessibility/dist/es/behaviors/HierarchicalTree/hierarchicalTreeBehavior.js
import { FocusZoneMode, FocusZoneDirection } from '../../focusZone/types';
import hierarchicalSubtreeBehavior from './hierarchicalSubtreeBehavior';
/**
 * @specification
 * Adds role 'tree' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in vertical direction.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 */

var hierarchicalTreeBehavior = function hierarchicalTreeBehavior(props) {
  var subtreeBehaviorData = hierarchicalSubtreeBehavior({});
  return {
    attributes: {
      root: Object.assign({}, subtreeBehaviorData.attributes.root, {
        role: 'tree',
        'aria-labelledby': props['aria-labelledby']
      })
    },
    keyActions: {
      root: Object.assign({}, subtreeBehaviorData.keyActions.root)
    },
    focusZone: {
      mode: FocusZoneMode.Embed,
      props: {
        direction: FocusZoneDirection.vertical
      }
    }
  };
};

export default hierarchicalTreeBehavior;
