// @stardust-ui/react/dist/commonjs/themes/teams/components/SplitButton/splitButtonStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _getBorderFocusStyles = _interopRequireDefault(require("../../getBorderFocusStyles"));

var _SplitButton = _interopRequireDefault(require("../../../../components/SplitButton/SplitButton"));

var splitButtonStyles = {
  button: function button(_ref) {
    var v = _ref.variables;
    return {
      border: 0,
      padding: v.padding,
      ':focus-visible': {
        '::before': {
          border: 0
        },
        '::after': {
          border: 0
        }
      },
      ':active': {
        backgroundColor: v.backgroundColorActive
      }
    };
  },
  root: function root(_ref2) {
    var _Object$assign, _ref3;

    var p = _ref2.props,
        v = _ref2.variables,
        siteVariables = _ref2.theme.siteVariables;
    var borderWidth = siteVariables.borderWidth;
    var borderFocusStyles = (0, _getBorderFocusStyles.default)({
      siteVariables: siteVariables
    });
    return _ref3 = {
      border: "".concat(borderWidth, " solid ").concat(v.borderColor),
      borderRadius: v.borderRadius,
      position: 'relative',
      display: 'inline-block'
    }, (0, _defineProperty2.default)(_ref3, "& .".concat(_SplitButton.default.slotClassNames.toggleButton), {
      borderRight: 0,
      borderTop: 0,
      borderBottom: 0
    }), (0, _defineProperty2.default)(_ref3, ':focus-within', Object.assign({
      boxShadow: 'none'
    }, p.isFromKeyboard ? Object.assign((_Object$assign = {}, (0, _defineProperty2.default)(_Object$assign, "& .".concat(_SplitButton.default.slotClassNames.toggleButton), {
      color: p.primary ? v.primaryColorFocus : v.colorFocus,
      backgroundColor: p.primary ? v.primaryBackgroundColorFocus : v.backgroundColorFocus,
      borderColor: siteVariables.colors.white
    }), (0, _defineProperty2.default)(_Object$assign, "color", v.colorFocus), (0, _defineProperty2.default)(_Object$assign, "backgroundColor", v.backgroundColorFocus), _Object$assign), borderFocusStyles[':focus-visible']) : {
      ':active': {
        backgroundColor: v.backgroundColorActive
      }
    })), _ref3;
  }
};
var _default = splitButtonStyles;
exports.default = _default;
