// @stardust-ui/react/dist/commonjs/themes/teams/componentVariables.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Attachment", {
  enumerable: true,
  get: function get() {
    return _attachmentVariables.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function get() {
    return _alertVariables.default;
  }
});
Object.defineProperty(exports, "Avatar", {
  enumerable: true,
  get: function get() {
    return _avatarVariables.default;
  }
});
Object.defineProperty(exports, "Button", {
  enumerable: true,
  get: function get() {
    return _buttonVariables.default;
  }
});
Object.defineProperty(exports, "ButtonGroup", {
  enumerable: true,
  get: function get() {
    return _buttonVariables.default;
  }
});
Object.defineProperty(exports, "Chat", {
  enumerable: true,
  get: function get() {
    return _chatVariables.default;
  }
});
Object.defineProperty(exports, "ChatItem", {
  enumerable: true,
  get: function get() {
    return _chatItemVariables.default;
  }
});
Object.defineProperty(exports, "ChatMessage", {
  enumerable: true,
  get: function get() {
    return _chatMessageVariables.default;
  }
});
Object.defineProperty(exports, "Checkbox", {
  enumerable: true,
  get: function get() {
    return _checkboxVariables.default;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function get() {
    return _dialogVariables.default;
  }
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function get() {
    return _dividerVariables.default;
  }
});
Object.defineProperty(exports, "Dropdown", {
  enumerable: true,
  get: function get() {
    return _dropdownVariables.default;
  }
});
Object.defineProperty(exports, "DropdownItem", {
  enumerable: true,
  get: function get() {
    return _dropdownItemVariables.default;
  }
});
Object.defineProperty(exports, "DropdownSearchInput", {
  enumerable: true,
  get: function get() {
    return _dropdownSearchInputVariables.default;
  }
});
Object.defineProperty(exports, "DropdownSelectedItem", {
  enumerable: true,
  get: function get() {
    return _dropdownSelectedItemVariables.default;
  }
});
Object.defineProperty(exports, "Embed", {
  enumerable: true,
  get: function get() {
    return _embedVariables.default;
  }
});
Object.defineProperty(exports, "Flex", {
  enumerable: true,
  get: function get() {
    return _flexVariables.default;
  }
});
Object.defineProperty(exports, "FlexItem", {
  enumerable: true,
  get: function get() {
    return _flexItemVariables.default;
  }
});
Object.defineProperty(exports, "Grid", {
  enumerable: true,
  get: function get() {
    return _gridVariables.default;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function get() {
    return _headerVariables.default;
  }
});
Object.defineProperty(exports, "HeaderDescription", {
  enumerable: true,
  get: function get() {
    return _headerDescriptionVariables.default;
  }
});
Object.defineProperty(exports, "HierarchicalTreeTitle", {
  enumerable: true,
  get: function get() {
    return _hierarchicalTreeTitleVariables.default;
  }
});
Object.defineProperty(exports, "Icon", {
  enumerable: true,
  get: function get() {
    return _iconVariables.default;
  }
});
Object.defineProperty(exports, "Image", {
  enumerable: true,
  get: function get() {
    return _imageVariables.default;
  }
});
Object.defineProperty(exports, "Input", {
  enumerable: true,
  get: function get() {
    return _inputVariables.default;
  }
});
Object.defineProperty(exports, "Label", {
  enumerable: true,
  get: function get() {
    return _labelVariables.default;
  }
});
Object.defineProperty(exports, "Layout", {
  enumerable: true,
  get: function get() {
    return _layoutVariables.default;
  }
});
Object.defineProperty(exports, "Loader", {
  enumerable: true,
  get: function get() {
    return _loaderVariables.default;
  }
});
Object.defineProperty(exports, "ItemLayout", {
  enumerable: true,
  get: function get() {
    return _itemLayoutVariables.default;
  }
});
Object.defineProperty(exports, "ListItem", {
  enumerable: true,
  get: function get() {
    return _listItemVariables.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function get() {
    return _menuVariables.default;
  }
});
Object.defineProperty(exports, "MenuItem", {
  enumerable: true,
  get: function get() {
    return _menuItemVariables.default;
  }
});
Object.defineProperty(exports, "MenuDivider", {
  enumerable: true,
  get: function get() {
    return _menuDividerVariables.default;
  }
});
Object.defineProperty(exports, "Popup", {
  enumerable: true,
  get: function get() {
    return _popupVariables.default;
  }
});
Object.defineProperty(exports, "PopupContent", {
  enumerable: true,
  get: function get() {
    return _popupContentVariables.default;
  }
});
Object.defineProperty(exports, "ProviderBox", {
  enumerable: true,
  get: function get() {
    return _providerBoxVariables.default;
  }
});
Object.defineProperty(exports, "RadioGroupItem", {
  enumerable: true,
  get: function get() {
    return _radioGroupItemVariables.default;
  }
});
Object.defineProperty(exports, "Reaction", {
  enumerable: true,
  get: function get() {
    return _reactionVariables.default;
  }
});
Object.defineProperty(exports, "ReactionGroup", {
  enumerable: true,
  get: function get() {
    return _reactionGroupVariables.default;
  }
});
Object.defineProperty(exports, "Segment", {
  enumerable: true,
  get: function get() {
    return _segmentVariables.default;
  }
});
Object.defineProperty(exports, "Slider", {
  enumerable: true,
  get: function get() {
    return _sliderVariables.default;
  }
});
Object.defineProperty(exports, "Status", {
  enumerable: true,
  get: function get() {
    return _statusVariables.default;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function get() {
    return _textVariables.default;
  }
});
Object.defineProperty(exports, "TextArea", {
  enumerable: true,
  get: function get() {
    return _textAreaVariables.default;
  }
});
Object.defineProperty(exports, "Toolbar", {
  enumerable: true,
  get: function get() {
    return _toolbarVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarCustomItem", {
  enumerable: true,
  get: function get() {
    return _toolbarCustomItemVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarItem", {
  enumerable: true,
  get: function get() {
    return _toolbarItemVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarDivider", {
  enumerable: true,
  get: function get() {
    return _toolbarDividerVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarMenu", {
  enumerable: true,
  get: function get() {
    return _toolbarMenuVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarMenuDivider", {
  enumerable: true,
  get: function get() {
    return _toolbarMenuDividerVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarMenuItem", {
  enumerable: true,
  get: function get() {
    return _toolbarMenuItemVariables.default;
  }
});
Object.defineProperty(exports, "ToolbarMenuRadioGroup", {
  enumerable: true,
  get: function get() {
    return _toolbarMenuRadioGroupVariables.default;
  }
});
Object.defineProperty(exports, "TreeTitle", {
  enumerable: true,
  get: function get() {
    return _treeTitleVariables.default;
  }
});
Object.defineProperty(exports, "Animation", {
  enumerable: true,
  get: function get() {
    return _animationVariables.default;
  }
});
Object.defineProperty(exports, "Video", {
  enumerable: true,
  get: function get() {
    return _videoVariables.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _tooltipVariables.default;
  }
});
Object.defineProperty(exports, "TooltipContent", {
  enumerable: true,
  get: function get() {
    return _tooltipContentVariables.default;
  }
});
Object.defineProperty(exports, "SplitButton", {
  enumerable: true,
  get: function get() {
    return _splitButtonVariables.default;
  }
});

var _attachmentVariables = _interopRequireDefault(require("./components/Attachment/attachmentVariables"));

var _alertVariables = _interopRequireDefault(require("./components/Alert/alertVariables"));

var _avatarVariables = _interopRequireDefault(require("./components/Avatar/avatarVariables"));

var _buttonVariables = _interopRequireDefault(require("./components/Button/buttonVariables"));

var _chatVariables = _interopRequireDefault(require("./components/Chat/chatVariables"));

var _chatItemVariables = _interopRequireDefault(require("./components/Chat/chatItemVariables"));

var _chatMessageVariables = _interopRequireDefault(require("./components/Chat/chatMessageVariables"));

var _checkboxVariables = _interopRequireDefault(require("./components/Checkbox/checkboxVariables"));

var _dialogVariables = _interopRequireDefault(require("./components/Dialog/dialogVariables"));

var _dividerVariables = _interopRequireDefault(require("./components/Divider/dividerVariables"));

var _dropdownVariables = _interopRequireDefault(require("./components/Dropdown/dropdownVariables"));

var _dropdownItemVariables = _interopRequireDefault(require("./components/Dropdown/dropdownItemVariables"));

var _dropdownSearchInputVariables = _interopRequireDefault(require("./components/Dropdown/dropdownSearchInputVariables"));

var _dropdownSelectedItemVariables = _interopRequireDefault(require("./components/Dropdown/dropdownSelectedItemVariables"));

var _embedVariables = _interopRequireDefault(require("./components/Embed/embedVariables"));

var _flexVariables = _interopRequireDefault(require("./components/Flex/flexVariables"));

var _flexItemVariables = _interopRequireDefault(require("./components/Flex/flexItemVariables"));

var _gridVariables = _interopRequireDefault(require("./components/Grid/gridVariables"));

var _headerVariables = _interopRequireDefault(require("./components/Header/headerVariables"));

var _headerDescriptionVariables = _interopRequireDefault(require("./components/Header/headerDescriptionVariables"));

var _hierarchicalTreeTitleVariables = _interopRequireDefault(require("./components/HierarchicalTree/hierarchicalTreeTitleVariables"));

var _iconVariables = _interopRequireDefault(require("./components/Icon/iconVariables"));

var _imageVariables = _interopRequireDefault(require("./components/Image/imageVariables"));

var _inputVariables = _interopRequireDefault(require("./components/Input/inputVariables"));

var _labelVariables = _interopRequireDefault(require("./components/Label/labelVariables"));

var _layoutVariables = _interopRequireDefault(require("./components/Layout/layoutVariables"));

var _loaderVariables = _interopRequireDefault(require("./components/Loader/loaderVariables"));

var _itemLayoutVariables = _interopRequireDefault(require("./components/ItemLayout/itemLayoutVariables"));

var _listItemVariables = _interopRequireDefault(require("./components/List/listItemVariables"));

var _menuVariables = _interopRequireDefault(require("./components/Menu/menuVariables"));

var _menuItemVariables = _interopRequireDefault(require("./components/Menu/menuItemVariables"));

var _menuDividerVariables = _interopRequireDefault(require("./components/Menu/menuDividerVariables"));

var _popupVariables = _interopRequireDefault(require("./components/Popup/popupVariables"));

var _popupContentVariables = _interopRequireDefault(require("./components/Popup/popupContentVariables"));

var _providerBoxVariables = _interopRequireDefault(require("./components/Provider/providerBoxVariables"));

var _radioGroupItemVariables = _interopRequireDefault(require("./components/RadioGroup/radioGroupItemVariables"));

var _reactionVariables = _interopRequireDefault(require("./components/Reaction/reactionVariables"));

var _reactionGroupVariables = _interopRequireDefault(require("./components/Reaction/reactionGroupVariables"));

var _segmentVariables = _interopRequireDefault(require("./components/Segment/segmentVariables"));

var _sliderVariables = _interopRequireDefault(require("./components/Slider/sliderVariables"));

var _statusVariables = _interopRequireDefault(require("./components/Status/statusVariables"));

var _textVariables = _interopRequireDefault(require("./components/Text/textVariables"));

var _textAreaVariables = _interopRequireDefault(require("./components/TextArea/textAreaVariables"));

var _toolbarVariables = _interopRequireDefault(require("./components/Toolbar/toolbarVariables"));

var _toolbarCustomItemVariables = _interopRequireDefault(require("./components/Toolbar/toolbarCustomItemVariables"));

var _toolbarItemVariables = _interopRequireDefault(require("./components/Toolbar/toolbarItemVariables"));

var _toolbarDividerVariables = _interopRequireDefault(require("./components/Toolbar/toolbarDividerVariables"));

var _toolbarMenuVariables = _interopRequireDefault(require("./components/Toolbar/toolbarMenuVariables"));

var _toolbarMenuDividerVariables = _interopRequireDefault(require("./components/Toolbar/toolbarMenuDividerVariables"));

var _toolbarMenuItemVariables = _interopRequireDefault(require("./components/Toolbar/toolbarMenuItemVariables"));

var _toolbarMenuRadioGroupVariables = _interopRequireDefault(require("./components/Toolbar/toolbarMenuRadioGroupVariables"));

var _treeTitleVariables = _interopRequireDefault(require("./components/Tree/treeTitleVariables"));

var _animationVariables = _interopRequireDefault(require("./components/Animation/animationVariables"));

var _videoVariables = _interopRequireDefault(require("./components/Video/videoVariables"));

var _tooltipVariables = _interopRequireDefault(require("./components/Tooltip/tooltipVariables"));

var _tooltipContentVariables = _interopRequireDefault(require("./components/Tooltip/tooltipContentVariables"));

var _splitButtonVariables = _interopRequireDefault(require("./components/SplitButton/splitButtonVariables"));
