// @stardust-ui/react/dist/commonjs/themes/teams/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getIcon = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _animations = _interopRequireDefault(require("./animations"));

var siteVariables = _interopRequireWildcard(require("./siteVariables"));

var componentVariables = _interopRequireWildcard(require("./componentVariables"));

var componentStyles = _interopRequireWildcard(require("./componentStyles"));

var _fontFaces = _interopRequireDefault(require("./fontFaces"));

var _staticStyles = _interopRequireDefault(require("./staticStyles"));

var _svg = _interopRequireDefault(require("./components/Icon/svg"));

var _createTheme = require("../createTheme");

var declareSvg = function declareSvg(svgIcon) {
  return {
    isSvg: true,
    icon: svgIcon
  };
};

var getIcon = function getIcon(iconAndMaybeStyles) {
  return iconAndMaybeStyles.styles ? iconAndMaybeStyles.icon : iconAndMaybeStyles;
};

exports.getIcon = getIcon;
var themeIcons = Object.keys(_svg.default).reduce(function (accIcons, iconName) {
  var iconAndMaybeStyles = _svg.default[iconName];
  var icon = getIcon(iconAndMaybeStyles);
  return Object.assign({}, accIcons, (0, _defineProperty2.default)({}, iconName, declareSvg(icon)));
}, {});
var icons = Object.assign({}, themeIcons, {
  'stardust-checkmark': themeIcons['accept'],
  'stardust-circle': themeIcons['stardust-circle'],
  'stardust-close': themeIcons['close'],
  'stardust-arrow-up': themeIcons['triangle-up'],
  'stardust-arrow-down': themeIcons['triangle-down'],
  'stardust-arrow-end': themeIcons['triangle-right'],
  'stardust-menu-arrow-down': themeIcons['chevron-down-medium'],
  'stardust-menu-arrow-end': themeIcons['chevron-right-medium'],
  'stardust-pause': themeIcons['pause'],
  'stardust-play': themeIcons['play']
});
var teamsTheme = (0, _createTheme.createTheme)({
  siteVariables: siteVariables,
  componentVariables: componentVariables,
  componentStyles: componentStyles,
  fontFaces: _fontFaces.default,
  staticStyles: _staticStyles.default,
  icons: icons,
  animations: _animations.default
}, 'teams');
var _default = teamsTheme;
exports.default = _default;
