// @stardust-ui/react/dist/commonjs/themes/teams/components/Dialog/dialogVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lib = require("../../../../lib");

var _default = function _default(siteVariables) {
  return {
    rootBackground: siteVariables.colors.white,
    rootBorderRadius: (0, _lib.pxToRem)(3),
    rootWidth: '50vw',
    rootPadding: "".concat((0, _lib.pxToRem)(27), " ").concat((0, _lib.pxToRem)(32), " ").concat((0, _lib.pxToRem)(20), " ").concat((0, _lib.pxToRem)(32)),
    contentMargin: "0 0 ".concat((0, _lib.pxToRem)(20), " 0"),
    boxShadow: siteVariables.shadowLevel4,
    foregroundColor: siteVariables.colors.grey[900],
    headerFontSize: siteVariables.fontSizes.large,
    headerFontWeight: siteVariables.fontWeightBold,
    headerMargin: "0 0 ".concat((0, _lib.pxToRem)(8), " 0"),
    overlayBackground: 'rgba(37, 36, 36, .75)',
    // todo: update to a palette value when daisy has mapped one
    overlayZIndex: 1000,
    headerActionMargin: "".concat((0, _lib.pxToRem)(-3), " ").concat((0, _lib.pxToRem)(-8), " 0 0")
  };
};

exports.default = _default;
