// @stardust-ui/docs-components/dist/commonjs/CodeSnippet/formatCode.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formatCode = exports.prettifyCode = void 0;

var _isObject2 = _interopRequireDefault(require("lodash/isObject"));

var _prettier = _interopRequireWildcard(require("prettier/standalone"));

require("prettier/parser-babylon");

require("prettier/parser-html");

require("prettier/parser-typescript");

// `prettier` is a CJS library, there are known issues with them:
// https://github.com/rollup/rollup/issues/1267#issuecomment-446681320
var prettier = _prettier.default || _prettier;
var prettierConfig = {
  htmlWhitespaceSensitivity: 'ignore',
  printWidth: 100,
  tabWidth: 2,
  semi: false,
  singleQuote: true,
  trailingComma: 'all',
  plugins: typeof window !== 'undefined' ? window.prettierPlugins : []
};

var normalizeToString = function normalizeToString(value) {
  if (Array.isArray(value)) return value.join('\n');
  return (0, _isObject2.default)(value) ? JSON.stringify(value, null, 2) : value;
};

var prettifyCode = function prettifyCode(code, parser) {
  var formatted = prettier.format(code, Object.assign({}, prettierConfig, {
    // a narrower print width is more friendly to doc examples
    parser: parser
  }));
  return formatted.replace(/^;</m, '<'); // remove beginning semi in JSX/HTML
};

exports.prettifyCode = prettifyCode;
var formatters = {
  bash: function bash() {
    var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    return val.replace(/^/g, '$  ');
  },
  json: function json(val) {
    return prettifyCode(val, 'json');
  },
  js: function js() {
    var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    return prettifyCode(val, 'babel');
  },
  jsx: function jsx() {
    var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    return prettifyCode(val, 'babel');
  },
  html: function html() {
    var val = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
    return prettifyCode(val, 'html');
  }
};

var formatCode = function formatCode(code, mode) {
  if (!code) return '';
  var formatter = formatters[mode];
  return formatter(normalizeToString(code)) // remove eof line break, they are not helpful for snippets
  .replace(/\n$/, '');
};

exports.formatCode = formatCode;
