// @stardust-ui/react/dist/commonjs/themes/teams/components/Label/labelVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.labelColorAreas = void 0;

var _lib = require("../../../../lib");

var _colorUtils = require("../../../colorUtils");

var labelColorAreas = (0, _lib.stringLiteralsArray)('foreground', 'background');
exports.labelColorAreas = labelColorAreas;

var _default = function _default(siteVars) {
  var colorScheme = (0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
    default: {
      background: 'rgba(0, 0, 0, 0.6)',
      foreground: 'rgb(232, 232, 232)'
    },
    brand: {
      background: siteVars.colorScheme.brand.foreground4
    },
    red: {
      background: siteVars.colorScheme.red.foreground1
    }
  });
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)(colorScheme, labelColorAreas),
    circularRadius: (0, _lib.pxToRem)(9999),
    padding: "0 ".concat((0, _lib.pxToRem)(4), " 0 ").concat((0, _lib.pxToRem)(4)),
    startPaddingLeft: '0px',
    endPaddingRight: '0px',
    height: (0, _lib.pxToRem)(20)
  };
};

exports.default = _default;
