// @stardust-ui/react/dist/commonjs/themes/teams/components/Label/labelStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _typeof2 = _interopRequireDefault(require("@babel/runtime/helpers/typeof"));

var _lib = require("../../../../lib");

var _colors = require("../../colors");

var labelStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, p.color);
    return Object.assign({
      display: 'inline-flex',
      alignItems: 'center',
      overflow: 'hidden',
      height: v.height,
      lineHeight: v.height,
      color: colors.background,
      backgroundColor: colors.foreground,
      fontSize: (0, _lib.pxToRem)(14),
      borderRadius: (0, _lib.pxToRem)(3),
      padding: v.padding
    }, p.image && (p.imagePosition === 'start' ? {
      paddingLeft: v.startPaddingLeft
    } : {
      paddingRight: v.endPaddingRight
    }), p.circular && {
      borderRadius: v.circularRadius
    });
  },
  image: function image(_ref2) {
    var v = _ref2.variables;
    return {
      height: v.height,
      width: v.height
    };
  },
  icon: function icon(_ref3) {
    var p = _ref3.props;
    return p.icon && (0, _typeof2.default)(p.icon) === 'object' && p.icon.onClick && {
      cursor: 'pointer'
    };
  }
};
var _default = labelStyles;
exports.default = _default;
