// @stardust-ui/react/dist/es/themes/teams/animations/timingFunctions.js
export var easeIn = 'cubic-bezier(1.00,0.00,0.78,1.00)';
export var easeOut = 'cubic-bezier(0.33,0.00,0.00,1.00)';
export var easeEasy = 'cubic-bezier(0.33,0.00,0.67,1.00)';
export var shotgun = 'cubic-bezier(0.00,0.00,0.00,1.00)';
export var easeIn90 = 'cubic-bezier(0.90,0.00,0.78,1.00)';
export var easeOut90 = 'cubic-bezier(0.33,0.00,0.10,1.00)';
export var easeIn80 = 'cubic-bezier(0.80,0.00,0.78,1.0)';
export var easeOut80 = 'cubic-bezier(0.33,0.00,0.20,1.00)';
export var easeIn70 = 'cubic-bezier(0.70,0.00,0.78,1.00)';
export var easeOut70 = 'cubic-bezier(0.33,0.00,0.30,1.00)';
export var easeIn60 = 'cubic-bezier(0.60,0.00,0.78,1.00)';
export var easeOut60 = 'cubic-bezier(0.33,0.00,0.40,1.00)';
export var easeInAlt = 'cubic-bezier(1.00,0.00,0.67,1.00)';
export var easeInSmooth = 'cubic-bezier(0.60,0.00,0.67,1.00)';
export var easeOutSmooth = 'cubic-bezier(0.60,0.00,0.00,1.00)';
export var expoIn5 = 'cubic-bezier(0.67,0.01,0.83,0.21)';
export var expoOut5 = 'cubic-bezier(0.17,0.8,0.33,0.99)';
export var expoInOut5 = 'cubic-bezier(0.78,0,0.22,1)';
export var bounce1 = 'cubic-bezier(0.65, 0.00, 0.26, 1.31)';
export var bounce2 = 'cubic-bezier(1.00, 0.00, 0.35, 1.36)';
