// @stardust-ui/react/dist/commonjs/themes/teams/components/Menu/menuStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lib = require("../../../../lib");

var _menuItemStyles = require("./menuItemStyles");

var _colors = require("../../colors");

var _default = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    var iconOnly = p.iconOnly,
        fluid = p.fluid,
        pointing = p.pointing,
        pills = p.pills,
        primary = p.primary,
        underlined = p.underlined,
        vertical = p.vertical,
        submenu = p.submenu;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, null, primary);
    return Object.assign({
      display: 'flex',
      minHeight: (0, _lib.pxToRem)(24),
      margin: 0,
      padding: 0,
      color: v.color,
      backgroundColor: v.backgroundColor || 'inherit',
      listStyleType: 'none'
    }, iconOnly && {
      alignItems: 'center'
    }, vertical && Object.assign({
      flexDirection: 'column',
      backgroundColor: v.verticalBackgroundColor,
      padding: "".concat((0, _lib.pxToRem)(8), " 0")
    }, submenu && {
      boxShadow: v.verticalBoxShadow
    }, !fluid && !submenu && {
      width: 'fit-content'
    }, iconOnly && {
      display: 'inline-block',
      width: 'auto'
    }), !pills && !iconOnly && !(pointing && vertical) && !underlined && {
      // primary has hardcoded grey border color
      border: "".concat(v.borderWidth, " solid ").concat(primary ? v.primaryBorderColor : v.borderColor || colors.border),
      borderRadius: (0, _lib.pxToRem)(4)
    }, underlined && {
      borderBottom: "".concat(v.underlinedBottomBorderWidth, " solid ").concat(v.underlinedBorderColor)
    });
  },
  divider: function divider(_ref2) {
    var _ref2$props = _ref2.props,
        pointing = _ref2$props.pointing,
        vertical = _ref2$props.vertical,
        pills = _ref2$props.pills;
    return Object.assign({}, pointing && vertical && {
      marginBottom: _menuItemStyles.verticalPointingBottomMargin
    }, pills && Object.assign({}, vertical ? {
      margin: "0 0 ".concat(_menuItemStyles.verticalPillsBottomMargin, " 0")
    } : {
      margin: "0 ".concat(_menuItemStyles.horizontalPillsRightMargin, " 0 0")
    }));
  }
};
exports.default = _default;
