// @stardust-ui/react/dist/commonjs/themes/teams/components/Menu/menuVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.menuColorAreas = void 0;

var _lib = require("../../../../lib");

var _colorUtils = require("../../../colorUtils");

var menuColorAreas = (0, _lib.stringLiteralsArray)('border', 'borderActive', 'foregroundActive', 'foregroundFocus', 'foregroundHover', 'backgroundHover', 'backgroundActive', 'backgroundFocus', 'foregroundDisabled');
exports.menuColorAreas = menuColorAreas;

var _default = function _default(siteVars) {
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)((0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
      default: {
        borderActive: siteVars.colorScheme.default.border2,
        backgroundActive: siteVars.colorScheme.default.backgroundActive1,
        backgroundFocus: siteVars.colorScheme.default.backgroundFocus1,
        foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1
      },
      brand: {
        foregroundHover: siteVars.colors.white,
        backgroundHover: siteVars.colors.brand[300],
        foregroundActive: siteVars.colors.white,
        borderActive: siteVars.colors.brand[600],
        backgroundActive: siteVars.colors.brand[500],
        // it's 600 in the color scheme
        foregroundFocus: siteVars.colors.white,
        backgroundFocus: siteVars.colors.brand[300],
        foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1
      }
    }), menuColorAreas),
    color: siteVars.colors.grey[500],
    colorActive: siteVars.colors.black,
    colorFocus: siteVars.colors.white,
    colorDisabled: undefined,
    borderColor: undefined,
    borderColorHover: undefined,
    borderColorActive: undefined,
    borderColorFocus: siteVars.colors.white,
    outlineColorFocus: siteVars.colors.black,
    backgroundColor: undefined,
    backgroundColorFocus: undefined,
    backgroundColorHover: undefined,
    backgroundColorActive: undefined,
    iconOnlyColorActive: siteVars.colors.brand[600],
    underlinedBorderColor: siteVars.colors.grey[200],
    lineHeightBase: siteVars.lineHeightMedium,
    horizontalPadding: "".concat((0, _lib.pxToRem)(14), " ").concat((0, _lib.pxToRem)(18)),
    verticalBackgroundColor: siteVars.colors.white,
    verticalItemPadding: "".concat((0, _lib.pxToRem)(7), " ").concat((0, _lib.pxToRem)(14)),
    verticalBoxShadow: siteVars.shadowLevel3,
    verticalDividerMargin: "".concat((0, _lib.pxToRem)(8), " 0"),
    verticalItemBorderWidth: (0, _lib.pxToRem)(2),
    verticalItemBorderColor: 'transparent',
    verticalPointingBorderColor: siteVars.colorScheme.brand.borderActive,
    verticalBackgroundColorFocus: siteVars.colors.grey[150],
    pointingIndicatorBackgroundColor: siteVars.colors.brand[600],
    underlinedBottomBorderWidth: (0, _lib.pxToRem)(2),
    primaryBorderColor: siteVars.colorScheme.default.border2,
    dividerHeight: (0, _lib.pxToRem)(1),
    borderWidth: (0, _lib.pxToRem)(1)
  };
};

exports.default = _default;
