// @stardust-ui/react/dist/commonjs/themes/teams/components/Menu/menuDividerStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _colors = require("../../colors");

var menuDividerStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    var colors = (0, _colors.getColorScheme)(v.colorScheme, null, p.primary);
    var borderColor = p.primary ? v.primaryBorderColor : v.borderColor || colors.border;
    var borderType = p.vertical ? 'borderTop' : 'borderLeft';
    return p.content ? {
      display: 'flex',
      justifyContent: 'center',
      flexDirection: 'column',
      textAlign: 'center'
    } : Object.assign((0, _defineProperty2.default)({}, borderType, "1px solid ".concat(borderColor)), !p.vertical && {
      alignSelf: 'stretch'
    }, p.vertical && p.inSubmenu && {
      margin: '8px 0'
    });
  }
};
var _default = menuDividerStyles;
exports.default = _default;
