// @stardust-ui/react/dist/commonjs/components/Button/ButtonGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _isNil2 = _interopRequireDefault(require("lodash/isNil"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _types = require("../../types");

var _lib = require("../../lib");

var _Button = _interopRequireDefault(require("./Button"));

var ButtonGroup =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(ButtonGroup, _UIComponent);

  function ButtonGroup() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ButtonGroup);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ButtonGroup)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "getStyleForButtonIndex", function (styles, isFirst, isLast) {
      var resultStyles = {};

      if (isFirst) {
        resultStyles = styles.firstButton;
      }

      if (isLast) {
        resultStyles = Object.assign({}, resultStyles, styles.lastButton);
      }

      if (!isFirst && !isLast) {
        resultStyles = styles.middleButton;
      }

      return resultStyles;
    });
    return _this;
  }

  (0, _createClass2.default)(ButtonGroup, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var _this2 = this;

      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          styles = _ref.styles,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          buttons = _this$props.buttons,
          circular = _this$props.circular,
          content = _this$props.content;

      if ((0, _isNil2.default)(buttons)) {
        return React.createElement(ElementType, (0, _extends2.default)({}, accessibility.attributes.root, _lib.rtlTextContainer.getAttributes({
          forElements: [children, content]
        }), unhandledProps, {
          className: classes.root
        }), (0, _lib.childrenExist)(children) ? children : content);
      }

      return React.createElement(ElementType, (0, _extends2.default)({}, unhandledProps, {
        className: classes.root
      }), (0, _map2.default)(buttons, function (button, idx) {
        return _Button.default.create(button, {
          defaultProps: function defaultProps() {
            return {
              circular: circular,
              styles: _this2.getStyleForButtonIndex(styles, idx === 0, idx === buttons.length - 1)
            };
          }
        });
      }));
    }
  }]);
  return ButtonGroup;
}(_lib.UIComponent);

(0, _defineProperty2.default)(ButtonGroup, "create", void 0);
(0, _defineProperty2.default)(ButtonGroup, "displayName", 'ButtonGroup');
(0, _defineProperty2.default)(ButtonGroup, "className", 'ui-buttons');
(0, _defineProperty2.default)(ButtonGroup, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon(), {
  buttons: customPropTypes.collectionShorthand,
  circular: PropTypes.bool
}));
(0, _defineProperty2.default)(ButtonGroup, "defaultProps", {
  as: 'div'
});
ButtonGroup.create = (0, _lib.createShorthandFactory)({
  Component: ButtonGroup,
  mappedProp: 'content',
  mappedArrayProp: 'buttons'
});
/**
 * A ButtonGroup represents multiple related actions as a group.
 */

var _default = (0, _types.withSafeTypeForAs)(ButtonGroup);

exports.default = _default;
