// @stardust-ui/react/dist/commonjs/components/Design/Design.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var _reactFela = require("react-fela");

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

// @ts-ignore

/**
 * The Design component provides a theme safe subset of CSS for designing layouts.
 */
function Design(_ref) {
  var config = _ref.config,
      children = _ref.children;
  var theme = React.useContext(_reactFela.ThemeContext);
  var getConfig = React.useCallback(function () {
    return config;
  }, [config]); // Heads Up! Keep in sync with renderComponent.tsx

  var styleParam = {
    theme: {
      direction: theme.rtl ? 'rtl' : 'ltr'
    }
  };
  var className = theme.renderer.renderRule(getConfig, styleParam);
  return children({
    className: className
  });
}

Design.displayName = 'Design';
Design.propTypes = {
  children: PropTypes.func.isRequired,
  config: customPropTypes.design.isRequired
};
var _default = Design;
exports.default = _default;
