// @stardust-ui/react/dist/es/themes/teams/components/Menu/menuVariables.js
import { pxToRem, stringLiteralsArray } from '../../../../lib';
import { extendColorScheme, pickValuesFromColorScheme } from '../../../colorUtils';
export var menuColorAreas = stringLiteralsArray('border', 'borderActive', 'foregroundActive', 'foregroundFocus', 'foregroundHover', 'backgroundHover', 'backgroundActive', 'backgroundFocus', 'foregroundDisabled');
export default (function (siteVars) {
  return {
    colorScheme: pickValuesFromColorScheme(extendColorScheme(siteVars.colorScheme, {
      default: {
        borderActive: siteVars.colorScheme.default.border2,
        backgroundActive: siteVars.colorScheme.default.backgroundActive1,
        backgroundFocus: siteVars.colorScheme.default.backgroundFocus1,
        foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1
      },
      brand: {
        foregroundHover: siteVars.colors.white,
        backgroundHover: siteVars.colors.brand[300],
        foregroundActive: siteVars.colors.white,
        borderActive: siteVars.colors.brand[600],
        backgroundActive: siteVars.colors.brand[500],
        // it's 600 in the color scheme
        foregroundFocus: siteVars.colors.white,
        backgroundFocus: siteVars.colors.brand[300],
        foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1
      }
    }), menuColorAreas),
    color: siteVars.colors.grey[500],
    colorActive: siteVars.colors.black,
    colorFocus: siteVars.colors.white,
    colorDisabled: undefined,
    borderColor: undefined,
    borderColorHover: undefined,
    borderColorActive: undefined,
    borderColorFocus: siteVars.colors.white,
    outlineColorFocus: siteVars.colors.black,
    backgroundColor: undefined,
    backgroundColorFocus: undefined,
    backgroundColorHover: undefined,
    backgroundColorActive: undefined,
    iconOnlyColorActive: siteVars.colors.brand[600],
    underlinedBorderColor: siteVars.colors.grey[200],
    lineHeightBase: siteVars.lineHeightMedium,
    horizontalPadding: "".concat(pxToRem(14), " ").concat(pxToRem(18)),
    verticalBackgroundColor: siteVars.colors.white,
    verticalItemPadding: "".concat(pxToRem(7), " ").concat(pxToRem(14)),
    verticalBoxShadow: siteVars.shadowLevel3,
    verticalDividerMargin: "".concat(pxToRem(8), " 0"),
    verticalItemBorderWidth: pxToRem(2),
    verticalItemBorderColor: 'transparent',
    verticalPointingBorderColor: siteVars.colorScheme.brand.borderActive,
    verticalBackgroundColorFocus: siteVars.colors.grey[150],
    pointingIndicatorBackgroundColor: siteVars.colors.brand[600],
    underlinedBottomBorderWidth: pxToRem(2),
    primaryBorderColor: siteVars.colorScheme.default.border2,
    dividerHeight: pxToRem(1),
    borderWidth: pxToRem(1)
  };
});
