// @stardust-ui/react/dist/es/themes/teams/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import animations from './animations';
import * as siteVariables from './siteVariables';
import * as componentVariables from './componentVariables';
import * as componentStyles from './componentStyles';
import fontFaces from './fontFaces';
import staticStyles from './staticStyles';
import { default as svgIconsAndStyles } from './components/Icon/svg';
import { createTheme } from '../createTheme';

var declareSvg = function declareSvg(svgIcon) {
  return {
    isSvg: true,
    icon: svgIcon
  };
};

export var getIcon = function getIcon(iconAndMaybeStyles) {
  return iconAndMaybeStyles.styles ? iconAndMaybeStyles.icon : iconAndMaybeStyles;
};
var themeIcons = Object.keys(svgIconsAndStyles).reduce(function (accIcons, iconName) {
  var iconAndMaybeStyles = svgIconsAndStyles[iconName];
  var icon = getIcon(iconAndMaybeStyles);
  return Object.assign({}, accIcons, _defineProperty({}, iconName, declareSvg(icon)));
}, {});
var icons = Object.assign({}, themeIcons, {
  'stardust-checkmark': themeIcons['accept'],
  'stardust-circle': themeIcons['stardust-circle'],
  'stardust-close': themeIcons['close'],
  'stardust-arrow-up': themeIcons['triangle-up'],
  'stardust-arrow-down': themeIcons['triangle-down'],
  'stardust-arrow-end': themeIcons['triangle-right'],
  'stardust-menu-arrow-down': themeIcons['chevron-down-medium'],
  'stardust-menu-arrow-end': themeIcons['chevron-right-medium'],
  'stardust-pause': themeIcons['pause'],
  'stardust-play': themeIcons['play']
});
var teamsTheme = createTheme({
  siteVariables: siteVariables,
  componentVariables: componentVariables,
  componentStyles: componentStyles,
  fontFaces: fontFaces,
  staticStyles: staticStyles,
  icons: icons,
  animations: animations
}, 'teams');
export default teamsTheme;
