// @stardust-ui/react/dist/es/themes/teams/siteVariables.js
import { pxToRem } from '../../lib';
import { colors } from './colors'; //
// COLORS
//

export { categoryColors, categoryColorScheme } from './categoryColors';
export { colors, contextualColors, naturalColors, primitiveColors, colorScheme, transparentColors } from './colors'; //
// BORDER STYLES
//

export var borderWidth = '1px';
export var borderRadius = '2px';
export var focusInnerBorderColor = colors.white;
export var focusOuterBorderColor = colors.black; //
// SHADOW LEVELS
//

export var shadowLevel1 = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .1)';
export var shadowLevel2 = '0 .4rem .7rem -.1rem rgba(0, 0, 0, .1)';
export var shadowLevel3 = '0 .8rem 1rem -.2rem rgba(0, 0, 0, .1)';
export var shadowLevel4 = '0 1.6rem 1.8rem -.4rem rgba(0, 0, 0, .1)';
export var shadowLevel1Dark = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .25)';
export var shadowLevel1Darker = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .5)'; //
// FONT SIZES
//

export var fontSizes = {
  smaller: pxToRem(10),
  small: pxToRem(12),
  medium: pxToRem(14),
  large: pxToRem(18),
  larger: pxToRem(24) //
  // FONT WEIGHTS
  //

};
export var fontWeightLight = 200;
export var fontWeightSemilight = 300;
export var fontWeightRegular = 400;
export var fontWeightSemibold = 600;
export var fontWeightBold = 700; //
// LINE HEIGHTS
//

export var lineHeightSmaller = 1.2;
export var lineHeightSmall = 1.3333;
export var lineHeightMedium = 1.4286;
export var lineHeightLarge = 1.3333;
export var lineHeightLarger = 1.3333; //
// SEMANTIC ASSIGNMENTS
//

export var bodyPadding = 0;
export var bodyMargin = 0;
export var bodyFontFamily = '"Segoe UI", "Helvetica Neue", "Apple Color Emoji", "Segoe UI Emoji", Helvetica, Arial, sans-serif';
export var bodyFontSize = fontSizes.medium;
export var bodyBackground = colors.white;
export var bodyColor = colors.grey[750];
export var bodyLineHeight = lineHeightMedium;
