// @stardust-ui/state/dist/commonjs/managers/dropdownManager.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDropdownManager = void 0;

var _createManager = _interopRequireDefault(require("../createManager"));

var createDropdownManager = function createDropdownManager() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _createManager.default)(Object.assign({}, config, {
    state: Object.assign({
      multiple: false,
      open: false,
      searchQuery: '',
      items: [],
      values: []
    }, config.state),
    actions: Object.assign({
      clear: function clear() {
        return function () {
          return {
            values: []
          };
        };
      },
      close: function close() {
        return function () {
          return {
            open: false
          };
        };
      },
      open: function open() {
        return function () {
          return {
            open: true
          };
        };
      },
      select: function select(item) {
        return function (state) {
          return {
            values: [state.items.indexOf(item).toString()]
          };
        };
      },
      unselect: function unselect(item) {
        return function (state) {
          return {
            values: [state.items.indexOf(item).toString()]
          };
        };
      }
    }, config.actions)
  }));
};

exports.createDropdownManager = createDropdownManager;
