// @stardust-ui/react/dist/commonjs/themes/teams/components/Flex/flexStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _utils = require("./utils");

var flexStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({
      display: 'flex'
    }, p.debug && {
      border: '1px dotted grey',
      background: 'lightgrey'
    }, p.inline && {
      display: 'inline-flex'
    }, p.column && {
      flexDirection: 'column'
    }, p.hAlign && (p.column ? {
      alignItems: (0, _utils.toFlexAlignment)(p.hAlign)
    } : {
      justifyContent: (0, _utils.toFlexAlignment)(p.hAlign)
    }), p.vAlign && (p.column ? {
      justifyContent: (0, _utils.toFlexAlignment)(p.vAlign)
    } : {
      alignItems: (0, _utils.toFlexAlignment)(p.vAlign)
    }), p.space && {
      justifyContent: "space-".concat(p.space)
    }, p.wrap && {
      flexWrap: 'wrap'
    }, p.fill && {
      width: '100%',
      height: '100%'
    }, p.padding && {
      padding: v[p.padding]
    }, p.gap && {
      '> *:not(:last-child)': (0, _defineProperty2.default)({}, p.column ? 'marginBottom' : 'marginRight', v[p.gap])
    });
  }
};
var _default = flexStyles;
exports.default = _default;
