// @stardust-ui/accessibility/dist/commonjs/focusZone/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FocusZoneDirection = exports.FocusZoneTabbableElements = exports.FocusZoneMode = void 0;
var FocusZoneMode;
exports.FocusZoneMode = FocusZoneMode;

(function (FocusZoneMode) {
  FocusZoneMode[FocusZoneMode["Custom"] = 0] = "Custom";
  FocusZoneMode[FocusZoneMode["Wrap"] = 1] = "Wrap";
  FocusZoneMode[FocusZoneMode["Embed"] = 2] = "Embed";
})(FocusZoneMode || (exports.FocusZoneMode = FocusZoneMode = {}));

var FocusZoneTabbableElements;
exports.FocusZoneTabbableElements = FocusZoneTabbableElements;

(function (FocusZoneTabbableElements) {
  FocusZoneTabbableElements[FocusZoneTabbableElements["none"] = 0] = "none";
  FocusZoneTabbableElements[FocusZoneTabbableElements["all"] = 1] = "all";
  FocusZoneTabbableElements[FocusZoneTabbableElements["inputOnly"] = 2] = "inputOnly";
})(FocusZoneTabbableElements || (exports.FocusZoneTabbableElements = FocusZoneTabbableElements = {}));

var FocusZoneDirection;
exports.FocusZoneDirection = FocusZoneDirection;

(function (FocusZoneDirection) {
  FocusZoneDirection[FocusZoneDirection["vertical"] = 0] = "vertical";
  FocusZoneDirection[FocusZoneDirection["horizontal"] = 1] = "horizontal";
  FocusZoneDirection[FocusZoneDirection["bidirectional"] = 2] = "bidirectional";
  FocusZoneDirection[FocusZoneDirection["bidirectionalDomOrder"] = 3] = "bidirectionalDomOrder";
})(FocusZoneDirection || (exports.FocusZoneDirection = FocusZoneDirection = {}));
