// @stardust-ui/react/dist/es/components/Debug/DebugLine.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import * as React from 'react';

var DebugLine = function DebugLine(_ref) {
  var active = _ref.active,
      _ref$indent = _ref.indent,
      indent = _ref$indent === void 0 ? 0 : _ref$indent,
      actionable = _ref.actionable,
      children = _ref.children,
      style = _ref.style,
      badge = _ref.badge,
      rest = _objectWithoutProperties(_ref, ["active", "indent", "actionable", "children", "style", "badge"]);

  return React.createElement("a", _extends({
    style: Object.assign({
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      paddingLeft: "".concat(indent * 2, "ch"),
      outline: 0
    }, actionable && {
      color: 'cornflowerblue',
      cursor: 'pointer'
    }, active && {
      background: 'rgba(255, 255, 255, 0.1)'
    }, style)
  }, rest), children, badge && React.createElement("span", {
    style: {
      padding: '0 4px',
      fontSize: 10,
      opacity: 0.75
    }
  }, badge));
};

export default DebugLine;
