// @stardust-ui/react/dist/es/components/Debug/ScrollToBottom.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _assertThisInitialized from "@babel/runtime/helpers/esm/assertThisInitialized";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';

var ScrollToBottom =
/*#__PURE__*/
function (_React$Component) {
  _inherits(ScrollToBottom, _React$Component);

  function ScrollToBottom() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, ScrollToBottom);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(ScrollToBottom)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "ref", React.createRef());

    return _this;
  }

  _createClass(ScrollToBottom, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      this.scrollToBottom();
    }
  }, {
    key: "scrollToBottom",
    value: function scrollToBottom() {
      this.ref.current.scrollTo({
        behavior: 'smooth',
        top: 999999
      });
    }
  }, {
    key: "render",
    value: function render() {
      return React.createElement("div", _extends({
        ref: this.ref
      }, this.props));
    }
  }]);

  return ScrollToBottom;
}(React.Component);

export default ScrollToBottom;
