// @stardust-ui/react/dist/commonjs/themes/teams/components/Avatar/avatarStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lib = require("../../../../lib");

var sizeToPxValue = {
  smallest: 24,
  smaller: 24,
  small: 24,
  medium: 32,
  large: 36,
  larger: 42,
  largest: 48
};
var avatarStyles = {
  root: function root(_ref) {
    var size = _ref.props.size;
    var sizeInRem = (0, _lib.pxToRem)(sizeToPxValue[size]);
    return {
      position: 'relative',
      backgroundColor: 'inherit',
      display: 'inline-block',
      verticalAlign: 'middle',
      height: sizeInRem,
      width: sizeInRem
    };
  },
  image: function image(_ref2) {
    var v = _ref2.variables;
    return {
      borderColor: v.avatarBorderColor,
      borderStyle: 'solid',
      borderWidth: "".concat(v.avatarBorderWidth, "px"),
      height: '100%',
      objectFit: 'cover',
      verticalAlign: 'top',
      width: '100%'
    };
  },
  label: function label(_ref3) {
    var size = _ref3.props.size;
    var sizeInRem = (0, _lib.pxToRem)(sizeToPxValue[size]);
    return {
      display: 'inline-block',
      width: sizeInRem,
      height: sizeInRem,
      lineHeight: sizeInRem,
      fontSize: (0, _lib.pxToRem)(sizeToPxValue[size] / 2.333),
      verticalAlign: 'top',
      textAlign: 'center',
      padding: '0px'
    };
  },
  status: function status(_ref4) {
    var v = _ref4.variables;
    return {
      position: 'absolute',
      bottom: "-".concat(v.statusBorderWidth, "px"),
      right: "-".concat(v.statusBorderWidth, "px")
    };
  }
};
var _default = avatarStyles;
exports.default = _default;
