// @stardust-ui/react/dist/commonjs/components/Form/Form.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var _FormField = _interopRequireDefault(require("./FormField"));

var Form =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Form, _UIComponent);

  function Form() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Form);

    for (var _len = arguments.length, _args = new Array(_len), _key = 0; _key < _len; _key++) {
      _args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Form)).call.apply(_getPrototypeOf2, [this].concat(_args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleSubmit", function (e) {
      var action = _this.props.action; // Heads up! Third party libs can pass own data as first argument, we need to check that it has preventDefault()
      // method.

      if (!action) (0, _invoke2.default)(e, 'preventDefault');

      for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
        args[_key2 - 1] = arguments[_key2];
      }

      _invoke2.default.apply(void 0, [_this.props, 'onSubmit', e, _this.props].concat(args));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderFields", function () {
      var fields = _this.props.fields;
      return (0, _map2.default)(fields, function (field) {
        return _FormField.default.create(field, {
          defaultProps: function defaultProps() {
            return {
              className: Form.slotClassNames.field
            };
          }
        });
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Form, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          action = _this$props.action,
          children = _this$props.children;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root,
        action: action,
        onSubmit: this.handleSubmit
      }, _lib.rtlTextContainer.getAttributes({
        forElements: [children]
      }), accessibility.attributes.root, unhandledProps), (0, _lib.childrenExist)(children) ? children : this.renderFields());
    }
  }]);
  return Form;
}(_lib.UIComponent);
/**
 * A Form is used to collect, oprionally validate, and submit the user input, in a structured way.
 */


(0, _defineProperty2.default)(Form, "create", void 0);
(0, _defineProperty2.default)(Form, "displayName", 'Form');
(0, _defineProperty2.default)(Form, "className", 'ui-form');
(0, _defineProperty2.default)(Form, "slotClassNames", {
  field: "".concat(Form.className, "__field")
});
(0, _defineProperty2.default)(Form, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  action: PropTypes.string,
  fields: customPropTypes.collectionShorthand,
  onSubmit: PropTypes.func
}));
(0, _defineProperty2.default)(Form, "defaultProps", {
  as: 'form'
});
(0, _defineProperty2.default)(Form, "Field", _FormField.default);

var _default = (0, _types.withSafeTypeForAs)(Form);

exports.default = _default;
