// @stardust-ui/react/dist/commonjs/themes/teams-dark/componentVariables.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Attachment", {
  enumerable: true,
  get: function get() {
    return _attachmentVariables.default;
  }
});
Object.defineProperty(exports, "Chat", {
  enumerable: true,
  get: function get() {
    return _chatVariables.default;
  }
});
Object.defineProperty(exports, "ChatMessage", {
  enumerable: true,
  get: function get() {
    return _chatMessageVariables.default;
  }
});
Object.defineProperty(exports, "Dialog", {
  enumerable: true,
  get: function get() {
    return _dialogVariables.default;
  }
});
Object.defineProperty(exports, "Divider", {
  enumerable: true,
  get: function get() {
    return _dividerVariables.default;
  }
});
Object.defineProperty(exports, "Header", {
  enumerable: true,
  get: function get() {
    return _headerVariables.default;
  }
});
Object.defineProperty(exports, "HeaderDescription", {
  enumerable: true,
  get: function get() {
    return _headerDescriptionVariables.default;
  }
});
Object.defineProperty(exports, "ListItem", {
  enumerable: true,
  get: function get() {
    return _listItemVariables.default;
  }
});
Object.defineProperty(exports, "RadioGroupItem", {
  enumerable: true,
  get: function get() {
    return _radioGroupItemVariables.default;
  }
});
Object.defineProperty(exports, "Segment", {
  enumerable: true,
  get: function get() {
    return _segmentVariables.default;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function get() {
    return _textVariables.default;
  }
});
Object.defineProperty(exports, "HierarchicalTreeTitle", {
  enumerable: true,
  get: function get() {
    return _hierarchicalTreeTitleVariables.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function get() {
    return _menuVariables.default;
  }
});
Object.defineProperty(exports, "Icon", {
  enumerable: true,
  get: function get() {
    return _iconVariables.default;
  }
});
Object.defineProperty(exports, "Reaction", {
  enumerable: true,
  get: function get() {
    return _reactionVariables.default;
  }
});
Object.defineProperty(exports, "Alert", {
  enumerable: true,
  get: function get() {
    return _alertVariables.default;
  }
});
Object.defineProperty(exports, "ProviderBox", {
  enumerable: true,
  get: function get() {
    return _providerBoxVariables.default;
  }
});
Object.defineProperty(exports, "Dropdown", {
  enumerable: true,
  get: function get() {
    return _dropdownVariables.default;
  }
});
Object.defineProperty(exports, "Label", {
  enumerable: true,
  get: function get() {
    return _labelVariables.default;
  }
});

var _attachmentVariables = _interopRequireDefault(require("./components/Attachment/attachmentVariables"));

var _chatVariables = _interopRequireDefault(require("./components/Chat/chatVariables"));

var _chatMessageVariables = _interopRequireDefault(require("./components/Chat/chatMessageVariables"));

var _dialogVariables = _interopRequireDefault(require("./components/Dialog/dialogVariables"));

var _dividerVariables = _interopRequireDefault(require("./components/Divider/dividerVariables"));

var _headerVariables = _interopRequireDefault(require("./components/Header/headerVariables"));

var _headerDescriptionVariables = _interopRequireDefault(require("./components/Header/headerDescriptionVariables"));

var _listItemVariables = _interopRequireDefault(require("./components/List/listItemVariables"));

var _radioGroupItemVariables = _interopRequireDefault(require("./components/RadioGroup/radioGroupItemVariables"));

var _segmentVariables = _interopRequireDefault(require("./components/Segment/segmentVariables"));

var _textVariables = _interopRequireDefault(require("./components/Text/textVariables"));

var _hierarchicalTreeTitleVariables = _interopRequireDefault(require("./components/HierarchicalTree/hierarchicalTreeTitleVariables"));

var _menuVariables = _interopRequireDefault(require("./components/Menu/menuVariables"));

var _iconVariables = _interopRequireDefault(require("./components/Icon/iconVariables"));

var _reactionVariables = _interopRequireDefault(require("./components/Reaction/reactionVariables"));

var _alertVariables = _interopRequireDefault(require("./components/Alert/alertVariables"));

var _providerBoxVariables = _interopRequireDefault(require("./components/Provider/providerBoxVariables"));

var _dropdownVariables = _interopRequireDefault(require("./components/Dropdown/dropdownVariables"));

var _labelVariables = _interopRequireDefault(require("./components/Label/labelVariables"));
