// @stardust-ui/react/dist/commonjs/themes/teams-dark/siteVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "colorScheme", {
  enumerable: true,
  get: function get() {
    return _colors.colorScheme;
  }
});
Object.defineProperty(exports, "categoryColorScheme", {
  enumerable: true,
  get: function get() {
    return _colors.categoryColorScheme;
  }
});
exports.bodyColor = exports.bodyBackground = exports.shadowLevel4 = exports.shadowLevel3 = exports.shadowLevel2 = exports.shadowLevel1 = exports.focusOuterBorderColor = exports.focusInnerBorderColor = void 0;

var _siteVariables = require("../teams/siteVariables");

var _colors = require("./colors");

//
// BORDER STYLES
//
var focusInnerBorderColor = _siteVariables.colors.black;
exports.focusInnerBorderColor = focusInnerBorderColor;
var focusOuterBorderColor = _siteVariables.colors.white; //
// SHADOW LEVELS
//

exports.focusOuterBorderColor = focusOuterBorderColor;
var shadowLevel1 = '0 .2rem .4rem -.075rem rgba(0, 0, 0, .25)';
exports.shadowLevel1 = shadowLevel1;
var shadowLevel2 = '0 .4rem 0.7rem -0.1rem rgba(0, 0, 0, .25)';
exports.shadowLevel2 = shadowLevel2;
var shadowLevel3 = '0 .8rem 1rem -0.2rem rgba(0, 0, 0, .25)';
exports.shadowLevel3 = shadowLevel3;
var shadowLevel4 = '0 .6rem 1.8rem -0.4rem rgba(0, 0, 0, .25)'; //
// SEMANTIC ASSIGNMENTS
//

exports.shadowLevel4 = shadowLevel4;
var bodyBackground = _siteVariables.colors.grey[650];
exports.bodyBackground = bodyBackground;
var bodyColor = _siteVariables.colors.white;
exports.bodyColor = bodyColor;
