// @stardust-ui/react/dist/commonjs/themes/teams/components/Status/statusStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.getSizeStyles = void 0;

var _lib = require("../../../../lib");

var getBackgroundColor = function getBackgroundColor(state, variables) {
  switch (state) {
    case 'success':
      return variables.successBackgroundColor;

    case 'info':
      return variables.infoBackgroundColor;

    case 'warning':
      return variables.warningBackgroundColor;

    case 'error':
      return variables.errorBackgroundColor;

    case 'unknown':
    default:
      return variables.defaultBackgroundColor;
  }
};

var getTextColor = function getTextColor(state, variables) {
  switch (state) {
    case 'success':
      return variables.successTextColor;

    case 'info':
      return variables.infoTextColor;

    case 'warning':
      return variables.warningTextColor;

    case 'error':
      return variables.errorTextColor;

    case 'unknown':
    default:
      return variables.defaultTextColor;
  }
};

var sizeToPxValue = {
  smallest: 8,
  smaller: 8,
  small: 8,
  medium: 10,
  large: 12,
  larger: 14,
  largest: 16
};

var getSizeStyles = function getSizeStyles(sizeInPx, variables) {
  var borderWidth = variables.borderColor && variables.borderWidth || 0;
  var sizeInRem = (0, _lib.pxToRem)(sizeInPx + 2 * borderWidth);
  return {
    height: sizeInRem,
    width: sizeInRem
  };
};

exports.getSizeStyles = getSizeStyles;
var statusStyles = {
  root: function root(_ref) {
    var _ref$props = _ref.props,
        color = _ref$props.color,
        size = _ref$props.size,
        state = _ref$props.state,
        variables = _ref.variables;
    return Object.assign({
      display: 'inline-flex',
      alignItems: 'center',
      justifyContent: 'center'
    }, getSizeStyles(sizeToPxValue[size], variables), {
      verticalAlign: 'middle',
      borderRadius: '9999px'
    }, variables.borderColor && {
      borderColor: variables.borderColor,
      borderWidth: (0, _lib.pxToRem)(variables.borderWidth),
      borderStyle: 'solid'
    }, {
      backgroundColor: color || getBackgroundColor(state, variables)
    });
  },
  icon: function icon(_ref2) {
    var state = _ref2.props.state,
        variables = _ref2.variables;
    return {
      color: getTextColor(state, variables)
    };
  }
};
var _default = statusStyles;
exports.default = _default;
