// @stardust-ui/react/dist/commonjs/themes/teams/components/Dropdown/dropdownSelectedItemStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _getIconFillOrOutlineStyles = _interopRequireDefault(require("../../getIconFillOrOutlineStyles"));

var dropdownSelectedItemStyles = {
  root: function root(_ref) {
    var v = _ref.variables;
    return Object.assign({
      margin: '.4rem 0 0 .4rem',
      color: v.selectedItemColor
    }, v.selectedItemBackgroundColor && {
      backgroundColor: v.selectedItemBackgroundColor
    }, {
      ':focus': {
        color: v.selectedItemColorFocus,
        backgroundColor: v.selectedItemBackgroundColorFocus,
        outline: '0'
      },
      ':hover': {
        color: v.selectedItemColorFocus,
        backgroundColor: v.selectedItemBackgroundColorFocus
      }
    });
  },
  icon: function icon(_ref2) {
    var v = _ref2.variables;
    return Object.assign({}, (0, _getIconFillOrOutlineStyles.default)({
      outline: true
    }), {
      ':hover': Object.assign({
        color: v.selectedItemColorFocus
      }, (0, _getIconFillOrOutlineStyles.default)({
        outline: false
      }))
    });
  }
};
var _default = dropdownSelectedItemStyles;
exports.default = _default;
