// @stardust-ui/react/dist/commonjs/themes/teams/components/Dropdown/dropdownVariables.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _lib = require("../../../../lib");

var _map = [3, 12].map(function (v) {
  return (0, _lib.pxToRem)(v);
}),
    _map2 = (0, _slicedToArray2.default)(_map, 2),
    cornerRadius = _map2[0],
    _12px_asRem = _map2[1];

var _default = function _default(siteVars) {
  return {
    backgroundColor: siteVars.colors.grey[100],
    backgroundColorHover: siteVars.colors.grey[150],
    borderColor: 'transparent',
    borderColorFocus: siteVars.colors.brand[600],
    borderWidth: '0px',
    containerBorderRadius: "".concat(cornerRadius),
    openAboveContainerBorderRadius: "0 0 ".concat(cornerRadius, " ").concat(cornerRadius),
    openBelowContainerBorderRadius: "".concat(cornerRadius, " ").concat(cornerRadius, " 0 0"),
    searchBorderBottomWidth: (0, _lib.pxToRem)(2),
    color: siteVars.bodyColor,
    selectedItemColor: siteVars.bodyColor,
    comboboxPaddingButton: "0 ".concat(_12px_asRem),
    comboboxFlexBasis: (0, _lib.pxToRem)(50),
    aboveListBorderRadius: "".concat(cornerRadius, " ").concat(cornerRadius, " 0 0"),
    belowListBorderRadius: "0 0 ".concat(cornerRadius, " ").concat(cornerRadius),
    listBackgroundColor: siteVars.colors.white,
    listBorderColor: 'transparent',
    listBorderWidth: '0px',
    listPadding: "".concat((0, _lib.pxToRem)(8), " 0 ").concat((0, _lib.pxToRem)(6)),
    listBoxShadow: siteVars.shadowLevel3,
    listMaxHeight: (0, _lib.pxToRem)(296),
    listItemFocusBorderWidth: (0, _lib.pxToRem)(1),
    listItemBackgroundColor: siteVars.colors.white,
    listItemColorHover: siteVars.colors.grey[750],
    listItemBackgroundColorHover: siteVars.colors.grey[100],
    listItemBackgroundColorActive: siteVars.colors.grey[100],
    listItemColorActive: siteVars.colors.grey[750],
    listItemSelectedFontWeight: siteVars.fontWeightSemibold,
    listItemSelectedColor: siteVars.colors.grey[750],
    selectedItemBackgroundColor: 'undefined',
    selectedItemColorFocus: siteVars.bodyColor,
    selectedItemBackgroundColorFocus: siteVars.colors.brand[200],
    selectedItemsMaxHeight: (0, _lib.pxToRem)(82),
    toggleIndicatorSize: (0, _lib.pxToRem)(32),
    triggerButtonColorHover: siteVars.bodyColor,
    width: (0, _lib.pxToRem)(356),
    // these should only apply when there is content in the image/media slot:
    listItemHeaderFontSize: siteVars.fontSizes.medium,
    listItemHeaderColor: siteVars.colors.grey[1000],
    listItemContentFontSize: siteVars.fontSizes.small,
    listItemContentColor: siteVars.colors.grey[450]
  };
};

exports.default = _default;
