// @stardust-ui/react/dist/commonjs/components/ItemLayout/ItemLayout.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _lib = require("../../lib");

var _Layout = _interopRequireDefault(require("../Layout/Layout"));

var _types = require("../../types");

var ItemLayout =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(ItemLayout, _UIComponent);

  function ItemLayout() {
    (0, _classCallCheck2.default)(this, ItemLayout);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ItemLayout).apply(this, arguments));
  }

  (0, _createClass2.default)(ItemLayout, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles;
      var _this$props = this.props,
          as = _this$props.as,
          debug = _this$props.debug,
          endMedia = _this$props.endMedia,
          media = _this$props.media,
          renderMainArea = _this$props.renderMainArea,
          rootCSS = _this$props.rootCSS,
          mediaCSS = _this$props.mediaCSS,
          endMediaCSS = _this$props.endMediaCSS;
      var startArea = media;
      var mainArea = renderMainArea(this.props, this.state, classes);
      var endArea = endMedia;
      var mergedMediaClasses = (0, _classnames.default)(ItemLayout.slotClassNames.media, classes.media);
      var mergedEndMediaClasses = (0, _classnames.default)(ItemLayout.slotClassNames.endMedia, classes.endMedia);
      return React.createElement(_Layout.default, (0, _extends2.default)({
        as: as,
        className: classes.root,
        styles: styles.root,
        rootCSS: rootCSS,
        alignItems: "center",
        gap: (0, _lib.pxToRem)(8),
        debug: debug,
        reducing: true,
        start: startArea && React.createElement("span", {
          style: mediaCSS,
          className: mergedMediaClasses
        }, _lib.rtlTextContainer.createFor({
          element: startArea
        })),
        main: mainArea,
        end: endArea && React.createElement("span", {
          style: endMediaCSS,
          className: mergedEndMediaClasses
        }, _lib.rtlTextContainer.createFor({
          element: endArea
        }))
      }, unhandledProps));
    }
  }]);
  return ItemLayout;
}(_lib.UIComponent);

(0, _defineProperty2.default)(ItemLayout, "create", void 0);
(0, _defineProperty2.default)(ItemLayout, "displayName", 'ItemLayout');
(0, _defineProperty2.default)(ItemLayout, "className", 'ui-itemlayout');
(0, _defineProperty2.default)(ItemLayout, "slotClassNames", void 0);
(0, _defineProperty2.default)(ItemLayout, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  accessibility: false,
  children: false,
  content: false
}), {
  contentMedia: PropTypes.any,
  content: PropTypes.any,
  debug: PropTypes.bool,
  header: PropTypes.any,
  endMedia: PropTypes.any,
  headerMedia: PropTypes.any,
  media: PropTypes.any,
  renderContentArea: PropTypes.func,
  renderHeaderArea: PropTypes.func,
  renderMainArea: PropTypes.func,
  rootCSS: PropTypes.object,
  mediaCSS: PropTypes.object,
  headerCSS: PropTypes.object,
  headerMediaCSS: PropTypes.object,
  contentCSS: PropTypes.object,
  contentMediaCSS: PropTypes.object,
  endMediaCSS: PropTypes.object
}));
(0, _defineProperty2.default)(ItemLayout, "defaultProps", {
  as: 'div',
  renderMainArea: function renderMainArea(props, state, classes) {
    var renderHeaderArea = props.renderHeaderArea,
        renderContentArea = props.renderContentArea;
    var headerArea = renderHeaderArea(props, state, classes);
    var contentArea = renderContentArea(props, state, classes);
    return React.createElement("div", {
      className: ItemLayout.slotClassNames.main,
      style: {
        gridTemplateRows: '1fr 1fr'
      }
    }, headerArea, contentArea);
  },
  renderHeaderArea: function renderHeaderArea(props, state, classes) {
    var debug = props.debug,
        header = props.header,
        headerMedia = props.headerMedia,
        headerCSS = props.headerCSS,
        headerMediaCSS = props.headerMediaCSS;
    var mergedClasses = (0, _classnames.default)(ItemLayout.slotClassNames.header, classes.header);
    var mediaClasses = (0, _classnames.default)(ItemLayout.slotClassNames.headerMedia, classes.headerMedia);
    return !header && !headerMedia ? null : React.createElement(_Layout.default, {
      className: mergedClasses,
      alignItems: "end",
      gap: (0, _lib.pxToRem)(8),
      debug: debug,
      main: _lib.rtlTextContainer.createFor({
        element: header
      }),
      rootCSS: headerCSS,
      end: headerMedia && React.createElement("span", {
        style: headerMediaCSS,
        className: mediaClasses
      }, _lib.rtlTextContainer.createFor({
        element: headerMedia
      }))
    });
  },
  renderContentArea: function renderContentArea(props, state, classes) {
    var debug = props.debug,
        content = props.content,
        contentMedia = props.contentMedia,
        contentCSS = props.contentCSS,
        contentMediaCSS = props.contentMediaCSS;
    var mergedClasses = (0, _classnames.default)(ItemLayout.slotClassNames.content, classes.content);
    var mediaClasses = (0, _classnames.default)(ItemLayout.slotClassNames.contentMedia, classes.contentMedia);
    return !content && !contentMedia ? null : React.createElement(_Layout.default, {
      className: mergedClasses,
      alignItems: "start",
      gap: (0, _lib.pxToRem)(8),
      debug: debug,
      rootCSS: contentCSS,
      main: _lib.rtlTextContainer.createFor({
        element: content
      }),
      end: contentMedia && React.createElement("span", {
        style: contentMediaCSS,
        className: mediaClasses
      }, _lib.rtlTextContainer.createFor({
        element: contentMedia
      }))
    });
  }
});
ItemLayout.create = (0, _lib.createShorthandFactory)({
  Component: ItemLayout,
  mappedProp: 'content'
});
ItemLayout.slotClassNames = {
  header: "".concat(ItemLayout.className, "__header"),
  headerMedia: "".concat(ItemLayout.className, "__headerMedia"),
  main: "".concat(ItemLayout.className, "__main"),
  content: "".concat(ItemLayout.className, "__content"),
  contentMedia: "".concat(ItemLayout.className, "__contentMedia"),
  media: "".concat(ItemLayout.className, "__media"),
  endMedia: "".concat(ItemLayout.className, "__endMedia")
  /**
   * (DEPRECATED) The Item Layout handles layout styles for menu items, list items and other similar item templates.
   */

};

var _default = (0, _types.withSafeTypeForAs)(ItemLayout);

exports.default = _default;
