// @stardust-ui/accessibility/dist/commonjs/behaviors/MenuButton/menuButtonBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _includes2 = _interopRequireDefault(require("lodash/includes"));

var _merge2 = _interopRequireDefault(require("lodash/merge"));

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

var _popupBehavior = _interopRequireDefault(require("../Popup/popupBehavior"));

/**
 * @description
 * Implements ARIA [MenuButton](https://www.w3.org/TR/wai-aria-practices/#menubutton) design pattern.
 *
 * @specification
 * Adds role='none'.
 * Adds attribute 'aria-haspopup=true' to 'trigger' slot if 'contextMenu' property is not set.
 * Adds attribute 'tabIndex=-1' based on the property 'open' to 'trigger' slot.
 * Adds attribute 'aria-controls=menu-id' based on the property 'menuId' to 'trigger' slot.
 * Adds attribute 'aria-expanded=true' based on the property 'open' to 'trigger' slot.
 * Adds attribute 'id=trigger-id' based on the property 'triggerId' to 'trigger' slot.
 * Adds attribute 'id=menu-id' based on the property 'menuId' to 'menu' slot.
 * Adds attribute 'aria-labelledby=trigger-id' based on the property 'triggerId' to 'menu' slot.
 */
var menuButtonBehavior = function menuButtonBehavior(props) {
  var behavior = (0, _popupBehavior.default)(props);
  return (0, _merge2.default)(behavior, {
    attributes: {
      root: {
        role: 'none'
      },
      trigger: Object.assign({
        'aria-controls': props.menuId,
        'aria-expanded': props.open || undefined,
        'aria-haspopup': props.contextMenu ? undefined : 'true',
        id: props.triggerId
      }, !props.contextMenu && props.open && {
        tabIndex: -1
      }),
      menu: {
        'aria-labelledby': props.triggerId,
        id: props.menuId
      }
    },
    keyActions: {
      root: Object.assign({}, props.open ? {
        closeMenu: {
          keyCombinations: [{
            keyCode: keyboardKey.Tab,
            shiftKey: false
          }, {
            keyCode: keyboardKey.Tab,
            shiftKey: true
          }]
        }
      } : (0, _includes2.default)(props.on, 'click') && {
        openAndFocusFirst: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowDown
          }]
        },
        openAndFocusLast: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowUp
          }]
        }
      })
    }
  });
};

var _default = menuButtonBehavior;
exports.default = _default;
