// @stardust-ui/react/dist/commonjs/components/Avatar/Avatar.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _Image = _interopRequireDefault(require("../Image/Image"));

var _Label = _interopRequireDefault(require("../Label/Label"));

var _Status = _interopRequireDefault(require("../Status/Status"));

var _types = require("../../types");

var _lib = require("../../lib");

var Avatar =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Avatar, _UIComponent);

  function Avatar() {
    (0, _classCallCheck2.default)(this, Avatar);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Avatar).apply(this, arguments));
  }

  (0, _createClass2.default)(Avatar, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles,
          variables = _ref.variables;
      var _ref2 = this.props,
          name = _ref2.name,
          status = _ref2.status,
          image = _ref2.image,
          label = _ref2.label,
          getInitials = _ref2.getInitials,
          size = _ref2.size;
      return React.createElement(ElementType, (0, _extends2.default)({}, accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), _Image.default.create(image, {
        defaultProps: function defaultProps() {
          return {
            fluid: true,
            avatar: true,
            title: name,
            styles: styles.image
          };
        }
      }), !image && _Label.default.create(label || {}, {
        defaultProps: function defaultProps() {
          return {
            content: getInitials(name),
            circular: true,
            title: name,
            styles: styles.label
          };
        }
      }), _Status.default.create(status, {
        defaultProps: function defaultProps() {
          return {
            size: size,
            styles: styles.status,
            variables: {
              borderColor: variables.statusBorderColor,
              borderWidth: variables.statusBorderWidth
            }
          };
        }
      }));
    }
  }]);
  return Avatar;
}(_lib.UIComponent);

(0, _defineProperty2.default)(Avatar, "create", void 0);
(0, _defineProperty2.default)(Avatar, "className", 'ui-avatar');
(0, _defineProperty2.default)(Avatar, "displayName", 'Avatar');
(0, _defineProperty2.default)(Avatar, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  children: false,
  content: false
}), {
  name: PropTypes.string,
  image: customPropTypes.itemShorthandWithoutJSX,
  label: customPropTypes.itemShorthand,
  size: customPropTypes.size,
  status: customPropTypes.itemShorthand,
  getInitials: PropTypes.func
}));
(0, _defineProperty2.default)(Avatar, "defaultProps", {
  size: 'medium',
  getInitials: function getInitials(name) {
    if (!name) {
      return '';
    }

    var reducedName = name.replace(/\s*\(.*?\)\s*/g, ' ').replace(/\s*{.*?}\s*/g, ' ').replace(/\s*\[.*?]\s*/g, ' ');
    var initials = reducedName.split(' ').filter(function (item) {
      return item !== '';
    }).map(function (item) {
      return item.charAt(0);
    }).reduce(function (accumulator, currentValue) {
      return accumulator + currentValue;
    });

    if (initials.length > 2) {
      return initials.charAt(0) + initials.charAt(initials.length - 1);
    }

    return initials;
  }
});
Avatar.create = (0, _lib.createShorthandFactory)({
  Component: Avatar,
  mappedProp: 'name'
});
/**
 * An Avatar is a graphical representation of a user.
 */

var _default = (0, _types.withSafeTypeForAs)(Avatar);

exports.default = _default;
