// @stardust-ui/react/dist/commonjs/components/Portal/PortalInner.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var ReactDOM = _interopRequireWildcard(require("react-dom"));

var _reactFela = require("react-fela");

var _lib = require("../../lib");

// @ts-ignore

/**
 * A PortalInner is a container for Portal's content.
 */
var PortalInner =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(PortalInner, _React$Component);

  function PortalInner() {
    (0, _classCallCheck2.default)(this, PortalInner);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(PortalInner).apply(this, arguments));
  }

  (0, _createClass2.default)(PortalInner, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      (0, _invoke2.default)(this.props, 'onMount', this.props);
    }
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      (0, _invoke2.default)(this.props, 'onUnmount', this.props);
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          children = _this$props.children,
          mountNode = _this$props.mountNode;
      var target = (0, _lib.isBrowser)() ? this.context.target.body : null;
      var container = mountNode || target;
      return container && ReactDOM.createPortal(children, container);
    }
  }]);
  return PortalInner;
}(React.Component);

(0, _defineProperty2.default)(PortalInner, "contextType", _reactFela.ThemeContext);
(0, _defineProperty2.default)(PortalInner, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  accessibility: false,
  animated: false,
  as: false,
  className: false,
  content: false,
  styled: false
}), {
  mountNode: PropTypes.object,
  onMount: PropTypes.func,
  onUnmount: PropTypes.func
}));
var _default = PortalInner;
exports.default = _default;
