// @stardust-ui/react/dist/commonjs/components/Label/Label.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Image = _interopRequireDefault(require("../Image/Image"));

var _Layout = _interopRequireDefault(require("../Layout/Layout"));

var _types = require("../../types");

var Label =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Label, _UIComponent);

  function Label() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Label);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Label)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleIconOverrides", function (iconProps) {
      return Object.assign({}, !iconProps.xSpacing && {
        xSpacing: 'none'
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Label, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          variables = _ref.variables,
          styles = _ref.styles;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content,
          icon = _this$props.icon,
          iconPosition = _this$props.iconPosition,
          image = _this$props.image,
          imagePosition = _this$props.imagePosition;

      if ((0, _lib.childrenExist)(children)) {
        return React.createElement(ElementType, (0, _extends2.default)({}, _lib.rtlTextContainer.getAttributes({
          forElements: [children]
        }), accessibility.attributes.root, unhandledProps, {
          className: classes.root
        }), children);
      }

      var imageElement = _Image.default.create(image, {
        defaultProps: function defaultProps() {
          return {
            styles: styles.image,
            variables: variables.image
          };
        }
      });

      var iconElement = _Icon.default.create(icon, {
        defaultProps: function defaultProps() {
          return {
            styles: styles.icon,
            variables: variables.icon
          };
        },
        overrideProps: this.handleIconOverrides
      });

      var startImage = imagePosition === 'start' && imageElement;
      var startIcon = iconPosition === 'start' && iconElement;
      var endIcon = iconPosition === 'end' && iconElement;
      var endImage = imagePosition === 'end' && imageElement;
      var hasStartElement = startImage || startIcon;
      var hasEndElement = endIcon || endImage;
      return React.createElement(ElementType, (0, _extends2.default)({}, accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), React.createElement(_Layout.default, {
        start: hasStartElement && React.createElement(React.Fragment, null, startImage, startIcon),
        main: content,
        end: hasEndElement && React.createElement(React.Fragment, null, endIcon, endImage),
        gap: (0, _lib.pxToRem)(3)
      }));
    }
  }]);
  return Label;
}(_lib.UIComponent);

(0, _defineProperty2.default)(Label, "displayName", 'Label');
(0, _defineProperty2.default)(Label, "create", void 0);
(0, _defineProperty2.default)(Label, "className", 'ui-label');
(0, _defineProperty2.default)(Label, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  color: true
}), {
  circular: PropTypes.bool,
  icon: customPropTypes.itemShorthandWithoutJSX,
  iconPosition: PropTypes.oneOf(['start', 'end']),
  image: customPropTypes.itemShorthandWithoutJSX,
  imagePosition: PropTypes.oneOf(['start', 'end']),
  fluid: PropTypes.bool
}));
(0, _defineProperty2.default)(Label, "defaultProps", {
  as: 'span',
  imagePosition: 'start',
  iconPosition: 'end'
});
Label.create = (0, _lib.createShorthandFactory)({
  Component: Label,
  mappedProp: 'content'
});
/**
 * A Label allows user to classify content.
 */

var _default = (0, _types.withSafeTypeForAs)(Label);

exports.default = _default;
