// @stardust-ui/react/dist/commonjs/components/Flex/Flex.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _get2 = _interopRequireDefault(require("lodash/get"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var _FlexItem = _interopRequireDefault(require("./FlexItem"));

var Flex =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Flex, _UIComponent);

  function Flex() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Flex);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Flex)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderChildren", function () {
      var _this$props = _this.props,
          column = _this$props.column,
          children = _this$props.children;
      return React.Children.map(children, function (child) {
        var isFlexItemElement = (0, _get2.default)(child, 'type.__isFlexItem');
        return isFlexItemElement ? React.cloneElement(child, {
          flexDirection: column ? 'column' : 'row'
        }) : child;
      });
    });
    return _this;
  }

  (0, _createClass2.default)(Flex, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, unhandledProps), this.renderChildren());
    }
  }]);
  return Flex;
}(_lib.UIComponent);
/**
 * A Flex is a layout component that arranges group of items aligned towards common direction (either row or column).
 */


(0, _defineProperty2.default)(Flex, "Item", _FlexItem.default);
(0, _defineProperty2.default)(Flex, "displayName", 'Flex');
(0, _defineProperty2.default)(Flex, "className", 'ui-flex');
(0, _defineProperty2.default)(Flex, "defaultProps", {
  as: 'div'
});
(0, _defineProperty2.default)(Flex, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  accessibility: false,
  content: false
}), {
  inline: PropTypes.bool,
  column: PropTypes.bool,
  wrap: PropTypes.bool,
  hAlign: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  vAlign: PropTypes.oneOf(['start', 'center', 'end', 'stretch']),
  space: PropTypes.oneOf(['around', 'between', 'evenly']),
  gap: PropTypes.oneOf(['gap.smaller', 'gap.small', 'gap.medium', 'gap.large']),
  padding: PropTypes.oneOf(['padding.medium']),
  fill: PropTypes.bool,
  debug: PropTypes.bool
}));

var _default = (0, _types.withSafeTypeForAs)(Flex);

exports.default = _default;
