// @stardust-ui/react/dist/commonjs/themes/teams/components/Toolbar/toolbarVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.toolbarColorAreas = void 0;

var _lib = require("../../../../lib");

var _colorUtils = require("../../../colorUtils");

var toolbarColorAreas = (0, _lib.stringLiteralsArray)('foreground1', 'background', 'border', 'foregroundActive', 'foregroundHover', 'backgroundHover', 'foregroundDisabled1', // custom
'menuItemForegroundHover', 'menuItemBackgroundHover');
exports.toolbarColorAreas = toolbarColorAreas;

var _default = function _default(siteVars) {
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)((0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
      default: {
        foregroundHover: siteVars.colorScheme.brand.foregroundHover,
        backgroundHover: 'transparent',
        foregroundActive: siteVars.colorScheme.brand.foregroundActive,
        menuItemForegroundHover: siteVars.colorScheme.default.foregroundHover,
        menuItemBackgroundHover: siteVars.colorScheme.default.backgroundHover
      }
    }), toolbarColorAreas),
    foreground: undefined,
    background: 'transparent',
    dividerBorder: undefined,
    foregroundHover: undefined,
    backgroundHover: undefined,
    foregroundActive: undefined,
    backgroundActive: 'transparent',
    foregroundDisabled: undefined,
    backgroundDisabled: 'transparent',
    itemHeight: (0, _lib.pxToRem)(32),
    itemPadding: '0',
    borderWidth: '2px',
    borderRadius: '50%',
    dividerMargin: "".concat((0, _lib.pxToRem)(10), " ").concat((0, _lib.pxToRem)(4)),
    menuWidth: (0, _lib.pxToRem)(200),
    menuPadding: "".concat((0, _lib.pxToRem)(8), " 0"),
    menuBackground: undefined,
    menuBorder: undefined,
    menuBorderWidth: '1px',
    menuBorderRadius: (0, _lib.pxToRem)(4),
    menuBoxShadow: siteVars.shadowLevel3,
    menuItemForeground: undefined,
    menuItemForegroundHover: undefined,
    menuItemBackgroundHover: undefined,
    menuItemForegroundDisabled: undefined,
    menuItemBackgroundDisabled: 'transparent',
    menuItemPadding: "".concat((0, _lib.pxToRem)(9), " ").concat((0, _lib.pxToRem)(16)),
    menuDividerBorder: undefined,
    menuDividerMargin: "".concat((0, _lib.pxToRem)(8), " 0"),
    customItemHorizontalPadding: (0, _lib.pxToRem)(16),
    customItemVerticalPadding: (0, _lib.pxToRem)(4)
  };
};

exports.default = _default;
