// @stardust-ui/react/dist/es/components/Status/Status.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { statusBehavior } from '@stardust-ui/accessibility';
import * as customPropTypes from '@stardust-ui/react-proptypes';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import Icon from '../Icon/Icon';
import { UIComponent, createShorthandFactory, commonPropTypes } from '../../lib';
import { withSafeTypeForAs } from '../../types';

var Status =
/*#__PURE__*/
function (_UIComponent) {
  _inherits(Status, _UIComponent);

  function Status() {
    _classCallCheck(this, Status);

    return _possibleConstructorReturn(this, _getPrototypeOf(Status).apply(this, arguments));
  }

  _createClass(Status, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          variables = _ref.variables,
          styles = _ref.styles;
      var _ref2 = this.props,
          icon = _ref2.icon;
      return React.createElement(ElementType, _extends({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps), Icon.create(icon, {
        defaultProps: function defaultProps() {
          return {
            size: 'smallest',
            styles: styles.icon,
            variables: variables.icon,
            xSpacing: 'none'
          };
        }
      }));
    }
  }]);

  return Status;
}(UIComponent);

_defineProperty(Status, "create", void 0);

_defineProperty(Status, "className", 'ui-status');

_defineProperty(Status, "displayName", 'Status');

_defineProperty(Status, "propTypes", Object.assign({}, commonPropTypes.createCommon({
  children: false,
  content: false
}), {
  color: PropTypes.string,
  icon: customPropTypes.itemShorthandWithoutJSX,
  size: customPropTypes.size,
  state: PropTypes.oneOf(['success', 'info', 'warning', 'error', 'unknown'])
}));

_defineProperty(Status, "defaultProps", {
  accessibility: statusBehavior,
  as: 'span',
  size: 'medium',
  state: 'unknown'
});

Status.create = createShorthandFactory({
  Component: Status,
  mappedProp: 'state'
});
/**
 * A Status represents someone's or something's state.
 *
 * @accessibility
 * Implements [ARIA img](https://www.w3.org/TR/wai-aria-1.1/#img) role.
 */

export default withSafeTypeForAs(Status);
