// @stardust-ui/docs-components/dist/es/knobs/KnobProvider.js
import _toConsumableArray from "@babel/runtime/helpers/esm/toConsumableArray";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import * as React from 'react';
import defaultComponents from './defaultComponents';
import { KnobContext, LogContext } from './KnobContexts';

var KnobProvider = function KnobProvider(props) {
  var children = props.children,
      components = props.components;

  var _React$useState = React.useState({}),
      _React$useState2 = _slicedToArray(_React$useState, 2),
      knobs = _React$useState2[0],
      setKnobs = _React$useState2[1];

  var _React$useState3 = React.useState([]),
      _React$useState4 = _slicedToArray(_React$useState3, 2),
      items = _React$useState4[0],
      setItems = _React$useState4[1];

  var registerKnob = function registerKnob(knob) {
    setKnobs(function (prevKnobs) {
      if (process.env.NODE_ENV !== 'production') {
        if (prevKnobs[knob.name]) {
          throw new Error("Knob with name \"".concat(knob.name, "\" has been already registered"));
        }
      }

      return Object.assign({}, prevKnobs, _defineProperty({}, knob.name, knob));
    });
  };

  var setKnobValue = function setKnobValue(knobName, knobValue) {
    setKnobs(function (prevKnob) {
      return Object.assign({}, prevKnob, _defineProperty({}, knobName, Object.assign({}, prevKnob[knobName], {
        value: knobValue
      })));
    });
  };

  var unregisterKnob = function unregisterKnob(knobName) {
    setKnobs(function (prevKnobs) {
      var newKnobs = Object.assign({}, prevKnobs);
      delete newKnobs[knobName];
      return newKnobs;
    });
  };

  var appendLog = React.useCallback(function (value) {
    return setItems(function (prevLog) {
      return [].concat(_toConsumableArray(prevLog), [value]);
    });
  }, []);
  var clearLog = React.useCallback(function () {
    return setItems([]);
  }, []);
  var knobValue = React.useMemo(function () {
    return {
      components: Object.assign({}, defaultComponents, components),
      knobs: knobs,
      registerKnob: registerKnob,
      setKnobValue: setKnobValue,
      unregisterKnob: unregisterKnob
    };
  }, [knobs, components]);
  var logValue = React.useMemo(function () {
    return {
      appendLog: appendLog,
      clearLog: clearLog,
      items: items
    };
  }, [items]);
  return React.createElement(KnobContext.Provider, {
    value: knobValue
  }, React.createElement(LogContext.Provider, {
    value: logValue
  }, children));
};

KnobProvider.defaultProps = {
  components: {}
};
export default KnobProvider;
