// @stardust-ui/react/dist/es/lib/commonPropTypes.js
import * as customPropTypes from '@stardust-ui/react-proptypes';
import * as PropTypes from 'prop-types';
export var createCommon = function createCommon() {
  var config = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var _config$accessibility = config.accessibility,
      accessibility = _config$accessibility === void 0 ? true : _config$accessibility,
      _config$animated = config.animated,
      animated = _config$animated === void 0 ? true : _config$animated,
      _config$as = config.as,
      as = _config$as === void 0 ? true : _config$as,
      _config$children = config.children,
      children = _config$children === void 0 ? 'node' : _config$children,
      _config$className = config.className,
      className = _config$className === void 0 ? true : _config$className,
      _config$color = config.color,
      color = _config$color === void 0 ? false : _config$color,
      _config$content = config.content,
      content = _config$content === void 0 ? 'node' : _config$content,
      _config$styled = config.styled,
      styled = _config$styled === void 0 ? true : _config$styled;
  return Object.assign({}, accessibility && {
    accessibility: customPropTypes.accessibility
  }, animated && {
    animation: customPropTypes.animation
  }, as && {
    as: PropTypes.elementType
  }, children && {
    children: children === 'element' ? PropTypes.element : PropTypes.node
  }, className && {
    className: PropTypes.string
  }, color && {
    color: PropTypes.string
  }, content && {
    content: content === 'shorthand' ? customPropTypes.itemShorthand : customPropTypes.nodeContent
  }, styled && {
    styles: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    variables: PropTypes.oneOfType([PropTypes.object, PropTypes.func]),
    design: customPropTypes.design
  });
};
