// @stardust-ui/react/dist/es/themes/teams/components/Popup/popupContentStyles.js
import getPointerStyles from '../../getPointerStyles';
var popupContentStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        rtl = _ref.rtl;
    return Object.assign({
      border: "".concat(v.borderSize, " solid ").concat(v.borderColor),
      borderRadius: v.borderRadius,
      boxShadow: v.boxShadow,
      display: 'block'
    }, p.pointing && getPointerStyles(v.pointerOffset, v.pointerMargin, rtl, p.placement).root);
  },
  pointer: function pointer(_ref2) {
    var p = _ref2.props,
        v = _ref2.variables,
        rtl = _ref2.rtl;
    return Object.assign({
      display: 'block',
      position: 'absolute',
      backgroundColor: 'inherit',
      borderBottom: "".concat(v.borderSize, " solid ").concat(v.borderColor),
      borderRight: "".concat(v.borderSize, " solid ").concat(v.borderColor),
      height: v.pointerSize,
      width: v.pointerSize
    }, getPointerStyles(v.pointerOffset, v.pointerMargin, rtl, p.placement).pointer);
  },
  content: function content(_ref3) {
    var p = _ref3.props,
        v = _ref3.variables;
    return {
      display: 'block',
      padding: v.padding
    };
  }
};
export default popupContentStyles;
