// @stardust-ui/accessibility/dist/commonjs/behaviors/Tab/tabListBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _types = require("../../focusZone/types");

var _tabBehavior = _interopRequireDefault(require("./tabBehavior"));

/**
 * @description
 * Implements ARIA Tabs design pattern.
 * Child item components need to have tabBehavior assigned.
 * @specification
 * Adds role 'tablist' to 'root' slot.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in bidirectionalDomOrder direction.
 * When component's container element receives focus, focus will be set to the default focusable child element of the component.
 */
var tabListBehavior = function tabListBehavior() {
  return {
    attributes: {
      root: {
        role: 'tablist'
      }
    },
    focusZone: {
      mode: _types.FocusZoneMode.Embed,
      props: {
        shouldFocusInnerElementWhenReceivedFocus: true,
        direction: _types.FocusZoneDirection.bidirectionalDomOrder
      }
    },
    childBehaviors: {
      item: _tabBehavior.default
    }
  };
};

var _default = tabListBehavior;
exports.default = _default;
