// @stardust-ui/react/dist/commonjs/components/Box/Box.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _createComponent = _interopRequireDefault(require("../../lib/createComponent"));

var _types = require("../../types");

var Box = (0, _createComponent.default)({
  displayName: 'Box',
  className: 'ui-box',
  propTypes: Object.assign({}, _lib.commonPropTypes.createCommon()),
  render: function render(config, props) {
    var ElementType = config.ElementType,
        classes = config.classes,
        unhandledProps = config.unhandledProps;
    var children = props.children,
        content = props.content;
    return React.createElement(ElementType, (0, _extends2.default)({}, _lib.rtlTextContainer.getAttributes({
      forElements: [children, content]
    }), unhandledProps, {
      className: classes.root
    }), (0, _lib.childrenExist)(children) ? children : content);
  }
});
Box.create = (0, _lib.createShorthandFactory)({
  Component: Box
});
/**
 * A Box is a basic component, commonly used for slots in other Stardust components.
 * By default it just renders a `div`.
 */

var _default = (0, _types.withSafeTypeForAs)(Box);

exports.default = _default;
