// @stardust-ui/accessibility/dist/commonjs/behaviors/Tree/treeBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

var _types = require("../../focusZone/types");

var _treeItemBehavior = _interopRequireDefault(require("./treeItemBehavior"));

/**
 * @specification
 * Adds role 'tree' to 'root' slot.
 * Adds attribute 'tabIndex=-1' to 'root' slot.
 * Adds attribute 'aria-labelledby' based on the property 'aria-labelledby' to 'root' slot.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in vertical direction.
 * Triggers 'expandSiblings' action with '*' on 'root'.
 */
var treeBehavior = function treeBehavior(props) {
  return {
    attributes: {
      root: {
        role: 'tree',
        'aria-labelledby': props['aria-labelledby'],
        tabIndex: -1
      }
    },
    keyActions: {
      root: {
        expandSiblings: {
          keyCombinations: [{
            keyCode: keyboardKey['*']
          }]
        }
      }
    },
    focusZone: {
      mode: _types.FocusZoneMode.Embed,
      props: {
        direction: _types.FocusZoneDirection.vertical
      }
    },
    childBehaviors: {
      item: _treeItemBehavior.default
    }
  };
};

var _default = treeBehavior;
exports.default = _default;
