// @stardust-ui/react/dist/commonjs/components/List/ListItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var _classnames = _interopRequireDefault(require("classnames"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _lib = require("../../lib");

var _types = require("../../types");

var _Box = _interopRequireDefault(require("../Box/Box"));

var ListItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(ListItem, _UIComponent);

  function ListItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ListItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ListItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      performClick: function performClick(event) {
        _this.handleClick(event);

        event.preventDefault();
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    return _this;
  }

  (0, _createClass2.default)(ListItem, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          accessibility = _ref.accessibility,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles;
      var _this$props = this.props,
          endMedia = _this$props.endMedia,
          media = _this$props.media,
          content = _this$props.content,
          contentMedia = _this$props.contentMedia,
          header = _this$props.header,
          headerMedia = _this$props.headerMedia;

      var contentElement = _Box.default.create(content, {
        defaultProps: function defaultProps() {
          return {
            className: ListItem.slotClassNames.content,
            styles: styles.content
          };
        }
      });

      var contentMediaElement = _Box.default.create(contentMedia, {
        defaultProps: function defaultProps() {
          return {
            className: ListItem.slotClassNames.contentMedia,
            styles: styles.contentMedia
          };
        }
      });

      var headerElement = _Box.default.create(header, {
        defaultProps: function defaultProps() {
          return {
            className: ListItem.slotClassNames.header,
            styles: styles.header
          };
        }
      });

      var headerMediaElement = _Box.default.create(headerMedia, {
        defaultProps: function defaultProps() {
          return {
            className: ListItem.slotClassNames.headerMedia,
            styles: styles.headerMedia
          };
        }
      });

      var endMediaElement = _Box.default.create(endMedia, {
        defaultProps: function defaultProps() {
          return {
            className: ListItem.slotClassNames.endMedia,
            styles: styles.endMedia
          };
        }
      });

      var mediaElement = _Box.default.create(media, {
        defaultProps: function defaultProps() {
          return {
            className: ListItem.slotClassNames.media,
            styles: styles.media
          };
        }
      });

      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root,
        onClick: this.handleClick
      }, accessibility.attributes.root, unhandledProps, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), mediaElement, React.createElement("div", {
        className: (0, _classnames.default)(ListItem.slotClassNames.main, classes.main)
      }, (headerElement || headerMediaElement) && React.createElement("div", {
        className: (0, _classnames.default)(ListItem.slotClassNames.headerWrapper, classes.headerWrapper)
      }, headerElement, headerMediaElement), (contentElement || contentMediaElement) && React.createElement("div", {
        className: (0, _classnames.default)(ListItem.slotClassNames.contentWrapper, classes.contentWrapper)
      }, contentElement, contentMediaElement)), endMediaElement);
    }
  }]);
  return ListItem;
}(_lib.UIComponent);

(0, _defineProperty2.default)(ListItem, "create", void 0);
(0, _defineProperty2.default)(ListItem, "displayName", 'ListItem');
(0, _defineProperty2.default)(ListItem, "className", 'ui-list__item');
(0, _defineProperty2.default)(ListItem, "slotClassNames", void 0);
(0, _defineProperty2.default)(ListItem, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  contentMedia: PropTypes.any,
  content: PropTypes.any,
  debug: PropTypes.bool,
  header: PropTypes.any,
  endMedia: PropTypes.any,
  headerMedia: PropTypes.any,
  important: PropTypes.bool,
  media: PropTypes.any,
  selectable: PropTypes.bool,
  navigable: PropTypes.bool,
  index: PropTypes.number,
  selected: PropTypes.bool,
  truncateContent: PropTypes.bool,
  truncateHeader: PropTypes.bool,
  onClick: PropTypes.func
}));
(0, _defineProperty2.default)(ListItem, "defaultProps", {
  as: 'li',
  accessibility: _accessibility.listItemBehavior
});
ListItem.create = (0, _lib.createShorthandFactory)({
  Component: ListItem,
  mappedProp: 'content'
});
ListItem.slotClassNames = {
  header: "".concat(ListItem.className, "__header"),
  headerMedia: "".concat(ListItem.className, "__headerMedia"),
  headerWrapper: "".concat(ListItem.className, "__headerWrapper"),
  main: "".concat(ListItem.className, "__main"),
  content: "".concat(ListItem.className, "__content"),
  contentMedia: "".concat(ListItem.className, "__contentMedia"),
  contentWrapper: "".concat(ListItem.className, "__contentWrapper"),
  media: "".concat(ListItem.className, "__media"),
  endMedia: "".concat(ListItem.className, "__endMedia")
  /**
   * A ListItem contains a single piece of content within a List.
   */

};

var _default = (0, _types.withSafeTypeForAs)(ListItem);

exports.default = _default;
