// @stardust-ui/react/dist/commonjs/themes/teams-high-contrast/components/Menu/menuVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _menuVariables = require("../../../teams/components/Menu/menuVariables");

var _colorUtils = require("../../../colorUtils");

var _default = function _default(siteVars) {
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)((0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
      default: {
        backgroundActive: siteVars.colorScheme.default.backgroundActive1,
        backgroundFocus: siteVars.colorScheme.default.backgroundFocus1,
        foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1
      },
      brand: {
        foregroundActive: siteVars.colors.black,
        foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1
      }
    }), _menuVariables.menuColorAreas),
    color: siteVars.colors.white,
    colorActive: siteVars.colors.black,
    backgroundColorFocus: siteVars.accessibleYellow,
    backgroundColorActive: siteVars.accessibleCyan,
    primaryBorderColor: siteVars.colors.white,
    verticalBackgroundColor: siteVars.colors.black,
    iconOnlyColorActive: siteVars.colors.black
  };
};

exports.default = _default;
