// @stardust-ui/react/dist/es/themes/teams/components/Flex/flexStyles.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import { toFlexAlignment } from './utils';
var flexStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({
      display: 'flex'
    }, p.debug && {
      border: '1px dotted grey',
      background: 'lightgrey'
    }, p.inline && {
      display: 'inline-flex'
    }, p.column && {
      flexDirection: 'column'
    }, p.hAlign && (p.column ? {
      alignItems: toFlexAlignment(p.hAlign)
    } : {
      justifyContent: toFlexAlignment(p.hAlign)
    }), p.vAlign && (p.column ? {
      justifyContent: toFlexAlignment(p.vAlign)
    } : {
      alignItems: toFlexAlignment(p.vAlign)
    }), p.space && {
      justifyContent: "space-".concat(p.space)
    }, p.wrap && {
      flexWrap: 'wrap'
    }, p.fill && {
      width: '100%',
      height: '100%'
    }, p.padding && {
      padding: v[p.padding]
    }, p.gap && {
      '> *:not(:last-child)': _defineProperty({}, p.column ? 'marginBottom' : 'marginRight', v[p.gap])
    });
  }
};
export default flexStyles;
