// @stardust-ui/react/dist/es/themes/teams/components/List/listItemVariables.js
import { pxToRem } from '../../../../lib';
export default (function (siteVariables) {
  return {
    minHeight: pxToRem(48),
    rootPadding: "0 ".concat(pxToRem(18), " 0 ").concat(pxToRem(20)),
    // Header
    // TODO: prod app uses 17.5px here, it should be 16px per the design guide!
    headerLineHeight: siteVariables.lineHeightSmall,
    headerFontSize: siteVariables.fontSizes.medium,
    // Header Media
    headerMediaFontSize: siteVariables.fontSizes.small,
    // TODO: prod app uses 17.5px here, it should be 16px per the design guide!
    headerMediaLineHeight: siteVariables.lineHeightSmall,
    // Content
    contentFontSize: siteVariables.fontSizes.small,
    contentLineHeight: siteVariables.lineHeightSmall,
    // Content Media
    contentMediaFontSize: siteVariables.fontSizes.small,
    contentMediaLineHeight: siteVariables.lineHeightSmall,
    // Selectable
    selectableFocusHoverColor: siteVariables.colors.white,
    selectableFocusHoverBackgroundColor: siteVariables.colors.brand[500],
    selectedColor: siteVariables.bodyColor,
    selectedBackgroundColor: siteVariables.colors.grey[100]
  };
});
