// @stardust-ui/react/dist/es/themes/teams/components/Button/buttonVariables.js
import { pxToRem } from '../../../../lib';
export default (function (siteVars) {
  return {
    padding: "0 ".concat(pxToRem(20)),
    height: pxToRem(32),
    minWidth: pxToRem(96),
    loadingMinWidth: pxToRem(118),
    maxWidth: pxToRem(280),
    borderRadius: siteVars.borderRadius,
    contentFontSize: siteVars.fontSizes.medium,
    contentFontWeight: siteVars.fontWeightSemibold,
    contentLineHeight: siteVars.lineHeightMedium,
    color: siteVars.colorScheme.default.foreground,
    colorHover: siteVars.colorScheme.default.foregroundHover,
    colorActive: siteVars.colorScheme.default.foregroundPressed,
    colorDisabled: siteVars.colorScheme.brand.foregroundDisabled,
    backgroundColor: siteVars.colorScheme.default.background,
    backgroundColorActive: siteVars.colorScheme.default.backgroundPressed,
    backgroundColorHover: siteVars.colorScheme.default.backgroundHover1,
    backgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled1,
    borderColor: siteVars.colorScheme.default.border,
    borderColorHover: siteVars.colorScheme.default.borderHover,
    borderColorActive: siteVars.colorScheme.default.borderPressed,
    borderColorDisabled: 'transparent',
    primaryColor: siteVars.colorScheme.brand.foreground4,
    primaryColorHover: siteVars.colorScheme.brand.foreground4,
    primaryBackgroundColor: siteVars.colorScheme.brand.background,
    primaryBackgroundColorActive: siteVars.colorScheme.brand.backgroundPressed,
    primaryBackgroundColorHover: siteVars.colorScheme.brand.backgroundHover,
    primaryBackgroundColorDisabled: siteVars.colorScheme.default.backgroundDisabled,
    primaryBorderColor: 'transparent',
    circularBorderRadius: pxToRem(999),
    textColor: siteVars.colorScheme.default.foreground1,
    textColorHover: siteVars.colorScheme.brand.foreground1,
    textPrimaryColor: siteVars.colorScheme.brand.foreground,
    textPrimaryColorHover: siteVars.colorScheme.brand.foreground1,
    textColorDisabled: siteVars.colorScheme.brand.foregroundDisabled1,
    primaryBoxShadow: siteVars.shadowLevel1Darker,
    boxShadow: siteVars.shadowLevel1,
    loaderBorderSize: pxToRem(2),
    loaderSize: pxToRem(20),
    loaderSvgHeight: pxToRem(1220),
    loaderSvgAnimationHeight: pxToRem(-1200),
    sizeSmallContentFontSize: siteVars.fontSizes.small,
    sizeSmallContentLineHeight: siteVars.lineHeightSmall,
    sizeSmallHeight: pxToRem(24),
    sizeSmallMinWidth: pxToRem(72),
    sizeSmallPadding: "0 ".concat(pxToRem(8)),
    sizeSmallLoaderBorderSize: pxToRem(2),
    sizeSmallLoaderSize: pxToRem(15),
    sizeSmallLoaderSvgHeight: pxToRem(895),
    sizeSmallLoaderSvgAnimationHeight: pxToRem(-880)
  };
});
