// @stardust-ui/react/dist/commonjs/components/Alert/Alert.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _uniqueId2 = _interopRequireDefault(require("lodash/uniqueId"));

var _accessibility = require("@stardust-ui/accessibility");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var _Box = _interopRequireDefault(require("../Box/Box"));

var _Button = _interopRequireDefault(require("../Button/Button"));

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Text = _interopRequireDefault(require("../Text/Text"));

var _ButtonGroup = _interopRequireDefault(require("../Button/ButtonGroup"));

var Alert =
/*#__PURE__*/
function (_AutoControlledCompon) {
  (0, _inherits2.default)(Alert, _AutoControlledCompon);

  function Alert() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, Alert);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(Alert)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleDismissOverrides", function (predefinedProps) {
      return {
        onClick: function onClick(e, buttonProps) {
          (0, _invoke2.default)(predefinedProps, 'onClick', e, buttonProps);
          (0, _invoke2.default)(_this.props, 'onDismiss', e, Object.assign({}, _this.props, {
            visible: false
          }));

          _this.setState({
            visible: false
          });
        }
      };
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleFocus", function (e) {
      (0, _invoke2.default)(_this.props, 'onFocus', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "renderContent", function (_ref) {
      var styles = _ref.styles,
          accessibility = _ref.accessibility;
      var _this$props = _this.props,
          actions = _this$props.actions,
          dismissible = _this$props.dismissible,
          dismissAction = _this$props.dismissAction,
          content = _this$props.content,
          icon = _this$props.icon,
          header = _this$props.header,
          body = _this$props.body;
      var bodyContent = React.createElement(React.Fragment, null, _Text.default.create(header, {
        defaultProps: function defaultProps() {
          return Object.assign({
            className: Alert.slotClassNames.header,
            styles: styles.header
          }, accessibility.attributes.header);
        }
      }), _Box.default.create(content, {
        defaultProps: function defaultProps() {
          return Object.assign({
            className: Alert.slotClassNames.content,
            styles: styles.content
          }, accessibility.attributes.content);
        }
      }));
      return React.createElement(React.Fragment, null, _Icon.default.create(icon, {
        defaultProps: function defaultProps() {
          return {
            className: Alert.slotClassNames.icon,
            styles: styles.icon
          };
        }
      }), _Box.default.create(body, {
        defaultProps: function defaultProps() {
          return Object.assign({
            id: _this.state.bodyId,
            className: Alert.slotClassNames.body
          }, accessibility.attributes.body, {
            styles: styles.body
          });
        },
        overrideProps: {
          children: bodyContent
        }
      }), _ButtonGroup.default.create(actions, {
        defaultProps: function defaultProps() {
          return {
            className: Alert.slotClassNames.actions,
            styles: styles.actions
          };
        }
      }), dismissible && _Button.default.create(dismissAction, {
        defaultProps: function defaultProps() {
          return Object.assign({
            iconOnly: true,
            text: true,
            className: Alert.slotClassNames.dismissAction,
            styles: styles.dismissAction
          }, accessibility.attributes.dismissAction);
        },
        overrideProps: _this.handleDismissOverrides
      }));
    });
    return _this;
  }

  (0, _createClass2.default)(Alert, [{
    key: "getInitialAutoControlledState",
    value: function getInitialAutoControlledState() {
      return {
        visible: true,
        bodyId: (0, _uniqueId2.default)('alert-body-')
      };
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(config) {
      var accessibility = config.accessibility,
          classes = config.classes,
          ElementType = config.ElementType,
          unhandledProps = config.unhandledProps;
      var children = this.props.children;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root,
        onFocus: this.handleFocus
      }, accessibility.attributes.root, _lib.rtlTextContainer.getAttributes({
        forElements: [children]
      }), unhandledProps), (0, _lib.childrenExist)(children) ? children : this.renderContent(config));
    }
  }]);
  return Alert;
}(_lib.AutoControlledComponent);
/**
 * An Alert displays a brief, important message to attract a user's attention without interrupting their current task.
 *
 * @accessibility
 * Implements [ARIA Alert](https://www.w3.org/TR/wai-aria-practices-1.1/#alert) design pattern.
 */


(0, _defineProperty2.default)(Alert, "displayName", 'Alert');
(0, _defineProperty2.default)(Alert, "className", 'ui-alert');
(0, _defineProperty2.default)(Alert, "slotClassNames", {
  content: "".concat(Alert.className, "__content"),
  actions: "".concat(Alert.className, "__actions"),
  dismissAction: "".concat(Alert.className, "__dismissAction"),
  icon: "".concat(Alert.className, "__icon"),
  header: "".concat(Alert.className, "__header"),
  body: "".concat(Alert.className, "__body")
});
(0, _defineProperty2.default)(Alert, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: 'shorthand'
}), {
  actions: PropTypes.oneOfType([customPropTypes.itemShorthand, customPropTypes.collectionShorthand]),
  icon: customPropTypes.itemShorthandWithoutJSX,
  header: customPropTypes.itemShorthand,
  attached: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['top', 'bottom'])]),
  fitted: PropTypes.bool,
  danger: PropTypes.bool,
  defaultVisible: PropTypes.bool,
  dismissible: PropTypes.bool,
  dismissAction: customPropTypes.itemShorthand,
  info: PropTypes.bool,
  onDismiss: PropTypes.func,
  onFocus: PropTypes.func,
  success: PropTypes.bool,
  visible: PropTypes.bool,
  warning: PropTypes.bool,
  body: customPropTypes.itemShorthand
}));
(0, _defineProperty2.default)(Alert, "defaultProps", {
  accessibility: _accessibility.alertBehavior,
  dismissAction: {
    icon: 'close'
  },
  body: {}
});
(0, _defineProperty2.default)(Alert, "autoControlledProps", ['visible']);

var _default = (0, _types.withSafeTypeForAs)(Alert);

exports.default = _default;
