// @stardust-ui/accessibility/dist/es/behaviors/List/selectableListItemBehavior.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as keyboardKey from 'keyboard-key';
import { IS_FOCUSABLE_ATTRIBUTE } from '../../attributes';

/**
 * @specification
 * Adds role='option'. This role is used for a selectable item in a list.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Adds attribute 'aria-selected=true' based on the property 'selected'. Based on this screen readers will recognize the selected state of the item.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var selectableListItemBehavior = function selectableListItemBehavior(props) {
  return {
    attributes: {
      root: _defineProperty({
        role: 'option',
        'aria-selected': !!props.selected
      }, IS_FOCUSABLE_ATTRIBUTE, true)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Enter
          }, {
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

export default selectableListItemBehavior;
