// @stardust-ui/react/dist/commonjs/components/Header/HeaderDescription.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var HeaderDescription =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(HeaderDescription, _UIComponent);

  function HeaderDescription() {
    (0, _classCallCheck2.default)(this, HeaderDescription);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(HeaderDescription).apply(this, arguments));
  }

  (0, _createClass2.default)(HeaderDescription, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content;
      return React.createElement(ElementType, (0, _extends2.default)({}, _lib.rtlTextContainer.getAttributes({
        forElements: [children, content]
      }), accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), (0, _lib.childrenExist)(children) ? children : content);
    }
  }]);
  return HeaderDescription;
}(_lib.UIComponent);

(0, _defineProperty2.default)(HeaderDescription, "create", void 0);
(0, _defineProperty2.default)(HeaderDescription, "className", 'ui-header__description');
(0, _defineProperty2.default)(HeaderDescription, "displayName", 'HeaderDescription');
(0, _defineProperty2.default)(HeaderDescription, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  color: true
})));
(0, _defineProperty2.default)(HeaderDescription, "defaultProps", {
  as: 'p'
});
HeaderDescription.create = (0, _lib.createShorthandFactory)({
  Component: HeaderDescription,
  mappedProp: 'content'
});
/**
 * A HeaderDescription provides more detailed information about the Header.
 */

var _default = (0, _types.withSafeTypeForAs)(HeaderDescription);

exports.default = _default;
