// @stardust-ui/react/dist/commonjs/themes/teams/components/Divider/dividerStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _get2 = _interopRequireDefault(require("lodash/get"));

var _lib = require("../../../../lib");

var beforeAndAfter = function beforeAndAfter(size, variables, colors) {
  return {
    content: '""',
    flex: 1,
    height: "".concat(size + 1, "px"),
    background: (0, _get2.default)(colors, 'foreground', variables.dividerColor)
  };
};

var dividerStyles = {
  root: function root(_ref) {
    var props = _ref.props,
        variables = _ref.variables;
    var children = props.children,
        color = props.color,
        fitted = props.fitted,
        size = props.size,
        important = props.important,
        content = props.content;
    var colors = variables.colorScheme[color];
    return Object.assign({
      color: (0, _get2.default)(colors, 'foreground', variables.textColor),
      display: 'flex',
      alignItems: 'center'
    }, !fitted && {
      paddingTop: variables.dividerPadding,
      paddingBottom: variables.dividerPadding
    }, important && {
      fontWeight: variables.importantFontWeight
    }, (0, _lib.childrenExist)(children) || content ? {
      textAlign: 'center',
      fontSize: (0, _lib.pxToRem)(12 + size),
      lineHeight: variables.textLineHeight,
      '::before': Object.assign({}, beforeAndAfter(size, variables, colors), {
        marginRight: (0, _lib.pxToRem)(20)
      }),
      '::after': Object.assign({}, beforeAndAfter(size, variables, colors), {
        marginLeft: (0, _lib.pxToRem)(20)
      })
    } : {
      '::before': Object.assign({}, beforeAndAfter(size, variables, colors))
    });
  }
};
var _default = dividerStyles;
exports.default = _default;
