// @stardust-ui/react/dist/commonjs/themes/teams-high-contrast/components/Dropdown/dropdownStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _Dropdown = _interopRequireDefault(require("../../../../components/Dropdown/Dropdown"));

var transparentColorStyle = {
  backgroundColor: 'transparent',
  borderColor: 'transparent',
  borderBottomColor: 'transparent'
};
var dropdownStyles = {
  container: function container(_ref) {
    var p = _ref.props,
        v = _ref.variables;
    return Object.assign({}, !p.open && {
      ':hover': (0, _defineProperty2.default)({
        backgroundColor: v.backgroundColorHover,
        borderColor: v.borderColorHover
      }, "& .".concat(_Dropdown.default.slotClassNames.triggerButton), {
        color: v.triggerButtonColorHover
      })
    });
  },
  triggerButton: function triggerButton(_ref2) {
    var p = _ref2.props,
        v = _ref2.variables;
    return {
      ':hover': Object.assign({}, transparentColorStyle, {
        color: v.triggerButtonColorHover
      }),
      ':focus': {
        color: v.color,
        ':active': {
          color: v.color
        }
      }
    };
  }
};
var _default = dropdownStyles;
exports.default = _default;
