// @stardust-ui/react/dist/commonjs/components/Debug/DebugLine.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var React = _interopRequireWildcard(require("react"));

var DebugLine = function DebugLine(_ref) {
  var active = _ref.active,
      _ref$indent = _ref.indent,
      indent = _ref$indent === void 0 ? 0 : _ref$indent,
      actionable = _ref.actionable,
      children = _ref.children,
      style = _ref.style,
      badge = _ref.badge,
      rest = (0, _objectWithoutProperties2.default)(_ref, ["active", "indent", "actionable", "children", "style", "badge"]);
  return React.createElement("a", (0, _extends2.default)({
    style: Object.assign({
      display: 'flex',
      justifyContent: 'space-between',
      alignItems: 'center',
      paddingLeft: "".concat(indent * 2, "ch"),
      outline: 0
    }, actionable && {
      color: 'cornflowerblue',
      cursor: 'pointer'
    }, active && {
      background: 'rgba(255, 255, 255, 0.1)'
    }, style)
  }, rest), children, badge && React.createElement("span", {
    style: {
      padding: '0 4px',
      fontSize: 10,
      opacity: 0.75
    }
  }, badge));
};

var _default = DebugLine;
exports.default = _default;
