// @stardust-ui/react/dist/commonjs/components/Debug/DebugComponentViewer.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var React = _interopRequireWildcard(require("react"));

var _DebugLine = _interopRequireDefault(require("./DebugLine"));

var _ScrollToBottom = _interopRequireDefault(require("./ScrollToBottom"));

var style = {
  padding: '8px',
  whiteSpace: 'pre',
  lineHeight: 1.4,
  background: '#222',
  overflowY: 'auto',
  color: '#CCC',
  fontFamily: 'monospace',
  fontWeight: 'bold'
};

var DebugComponentViewer = function DebugComponentViewer(props) {
  var fiberNav = props.fiberNav,
      onFiberChanged = props.onFiberChanged,
      onFiberSelected = props.onFiberSelected;
  var ownerNav = fiberNav.owner || {
    jsxString: 'unknown'
  };
  var parentNavs = [];

  if (fiberNav.owner) {
    var parentNav = fiberNav.parent;

    while (parentNav && !parentNav.isEqual(ownerNav)) {
      if (parentNav.stardustDebug) parentNavs.unshift(parentNav);
      parentNav = parentNav.parent;
    }
  }

  var component = fiberNav.name && React.createElement(_DebugLine.default, null, fiberNav.jsxString);
  return React.createElement(_ScrollToBottom.default, {
    style: style
  }, React.createElement(_DebugLine.default, (0, _extends2.default)({
    indent: 0
  }, ownerNav.stardustDebug && {
    actionable: true,
    tabIndex: 0,
    onClick: function onClick(e) {
      e.preventDefault();
      onFiberChanged(ownerNav);
    },
    onMouseEnter: function onMouseEnter(e) {
      return onFiberSelected(ownerNav);
    },
    onMouseLeave: function onMouseLeave(e) {
      return onFiberSelected(null);
    }
  }), ownerNav.jsxString), React.createElement(_DebugLine.default, {
    indent: 1,
    style: {
      color: '#ba645e'
    }
  }, "render()"), parentNavs.map(function (parent, i) {
    return React.createElement(_DebugLine.default, {
      key: i,
      indent: 2 + i,
      actionable: true,
      tabIndex: "0",
      onClick: function onClick(e) {
        e.preventDefault();
        onFiberChanged(parent);
      },
      onMouseEnter: function onMouseEnter(e) {
        return onFiberSelected(parent);
      },
      onMouseLeave: function onMouseLeave(e) {
        return onFiberSelected(null);
      }
    }, parent.jsxString);
  }), React.createElement(_DebugLine.default, {
    indent: 3 + (parentNavs.length - 1),
    active: true,
    badge: "selected",
    actionable: true,
    tabIndex: "0",
    onClick: function onClick(e) {
      e.preventDefault();
      onFiberChanged(fiberNav);
    },
    onMouseEnter: function onMouseEnter(e) {
      return onFiberSelected(fiberNav);
    },
    onMouseLeave: function onMouseLeave(e) {
      return onFiberSelected(null);
    }
  }, component));
};

var _default = DebugComponentViewer;
exports.default = _default;
