// @stardust-ui/react/dist/commonjs/components/Debug/FiberNavigator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

// ========================================================
// react/packages/shared/ReactTypes.js
// ========================================================
// ========================================================
// react/packages/react-reconciler/src/ReactFiberHooks.js
// ========================================================
var WorkTag;

(function (WorkTag) {
  WorkTag[WorkTag["FunctionComponent"] = 0] = "FunctionComponent";
  WorkTag[WorkTag["ClassComponent"] = 1] = "ClassComponent";
  WorkTag[WorkTag["IndeterminateComponent"] = 2] = "IndeterminateComponent";
  WorkTag[WorkTag["HostRoot"] = 3] = "HostRoot";
  WorkTag[WorkTag["HostPortal"] = 4] = "HostPortal";
  WorkTag[WorkTag["HostComponent"] = 5] = "HostComponent";
  WorkTag[WorkTag["HostText"] = 6] = "HostText";
  WorkTag[WorkTag["Fragment"] = 7] = "Fragment";
  WorkTag[WorkTag["Mode"] = 8] = "Mode";
  WorkTag[WorkTag["ContextConsumer"] = 9] = "ContextConsumer";
  WorkTag[WorkTag["ContextProvider"] = 10] = "ContextProvider";
  WorkTag[WorkTag["ForwardRef"] = 11] = "ForwardRef";
  WorkTag[WorkTag["Profiler"] = 12] = "Profiler";
  WorkTag[WorkTag["SuspenseComponent"] = 13] = "SuspenseComponent";
  WorkTag[WorkTag["MemoComponent"] = 14] = "MemoComponent";
  WorkTag[WorkTag["SimpleMemoComponent"] = 15] = "SimpleMemoComponent";
  WorkTag[WorkTag["LazyComponent"] = 16] = "LazyComponent";
  WorkTag[WorkTag["IncompleteClassComponent"] = 17] = "IncompleteClassComponent";
  WorkTag[WorkTag["DehydratedFragment"] = 18] = "DehydratedFragment";
  WorkTag[WorkTag["SuspenseListComponent"] = 19] = "SuspenseListComponent";
  WorkTag[WorkTag["FundamentalComponent"] = 20] = "FundamentalComponent";
  WorkTag[WorkTag["ScopeComponent"] = 21] = "ScopeComponent";
})(WorkTag || (WorkTag = {}));

var isDOMNode = function isDOMNode(e) {
  return e && typeof e.tagName === 'string' && e.nodeType === Node.ELEMENT_NODE;
};

var FiberNavigator =
/*#__PURE__*/
function () {
  function FiberNavigator() {
    (0, _classCallCheck2.default)(this, FiberNavigator);
    (0, _defineProperty2.default)(this, "__fiber", void 0);
  }

  (0, _createClass2.default)(FiberNavigator, [{
    key: "isEqual",
    //
    // Methods
    //
    value: function isEqual(fiberNav) {
      // TODO: do equality check on __fiber instead, however, see fromFiber TODO :/
      return !!fiberNav && fiberNav.instance === this.instance;
    }
  }, {
    key: "usesHook",
    value: function usesHook(name) {
      return this.__fiber._debugHookTypes.some(function (hook) {
        return hook === name;
      });
    }
  }, {
    key: "find",
    value: function find(condition, move) {
      var fiber = FiberNavigator.fromFiber(this.__fiber);

      while (fiber) {
        if (condition(fiber)) {
          return fiber;
        }

        fiber = move(fiber);
      }

      return null;
    }
  }, {
    key: "findOwner",
    value: function findOwner(condition) {
      return this.find(condition, function (fiber) {
        return fiber.owner;
      });
    }
  }, {
    key: "findParent",
    value: function findParent(condition) {
      return this.find(condition, function (fiber) {
        return fiber.parent;
      });
    } //
    // Component Types
    //

  }, {
    key: "name",
    get: function get() {
      return this.isClassComponent || this.isFunctionComponent ? this.__fiber.type.displayName || this.__fiber.type.name : this.isHostComponent ? this.__fiber.stateNode.constructor.name : null;
    }
  }, {
    key: "parent",
    get: function get() {
      return FiberNavigator.fromFiber(this.__fiber.return);
    }
  }, {
    key: "owner",
    get: function get() {
      return FiberNavigator.fromFiber(this.__fiber._debugOwner);
    }
  }, {
    key: "domNode",
    get: function get() {
      var fiber = this.__fiber;

      do {
        if (isDOMNode(fiber.stateNode)) {
          return fiber.stateNode;
        }

        fiber = fiber.child;
      } while (fiber);

      return null;
    }
  }, {
    key: "instance",
    get: function get() {
      return this.isClassComponent ? this.__fiber.stateNode : this.isFunctionComponent // TODO: assumes functional component w/useRef
      ? this.__fiber.memoizedState && this.__fiber.memoizedState.memoizedState && this.__fiber.memoizedState.memoizedState.current : null;
    }
  }, {
    key: "reactComponent",
    get: function get() {
      return this.isHostComponent ? this.owner.elementType : this.elementType;
    }
  }, {
    key: "elementType",
    get: function get() {
      return this.__fiber.elementType;
    }
  }, {
    key: "stardustDebug",
    get: function get() {
      return this.instance && this.instance.stardustDebug ? this.instance.stardustDebug : null;
    }
  }, {
    key: "jsxString",
    get: function get() {
      return "<".concat(this.name, " />");
    }
  }, {
    key: "isClassComponent",
    get: function get() {
      // React.Component subclasses have this flag
      // https://reactjs.org/docs/implementation-notes.html
      return typeof this.__fiber.type === 'function' && !!this.__fiber.type.prototype.isReactComponent;
    }
  }, {
    key: "isFunctionComponent",
    get: function get() {
      // React.Component subclasses have this flag
      // https://reactjs.org/docs/implementation-notes.html
      return typeof this.__fiber.type === 'function' && !this.__fiber.type.prototype.isReactComponent;
    }
  }, {
    key: "isHostComponent",
    get: function get() {
      // Host components are platform components (i.e. 'div' on web)
      // https://github.com/acdlite/react-fiber-architecture#type-and-key
      return typeof this.__fiber.type === 'string';
    } //
    // What this fiber component renders
    //

  }, {
    key: "isDOMComponent",
    get: function get() {
      return !!this.__fiber.child && FiberNavigator.fromFiber(this.__fiber.child).isHostComponent;
    } // https://github.com/facebook/react/blob/16.8.6/packages/react-dom/src/test-utils/ReactTestUtils.js#L193

  }, {
    key: "isCompositeComponent",
    get: function get() {
      return this.isDOMComponent ? false : !!this.instance && !!this.instance.render && !!this.instance.setState;
    }
  }]);
  return FiberNavigator;
}();

(0, _defineProperty2.default)(FiberNavigator, "domNodeToReactFiber", function (elm) {
  if (!elm) return null;

  for (var k in elm) {
    if (k.startsWith('__reactInternalInstance$')) {
      return elm[k];
    }
  }

  return null;
});
(0, _defineProperty2.default)(FiberNavigator, "fromFiber", function (fiber) {
  if (!fiber) return null;
  var fiberNavigator = new FiberNavigator();
  Object.defineProperty(fiberNavigator, '__fiber', {
    value: fiber,
    enumerable: false,
    writable: false,
    configurable: false
  });
  return fiberNavigator;
});
(0, _defineProperty2.default)(FiberNavigator, "fromDOMNode", function (domNode) {
  var fiber = FiberNavigator.domNodeToReactFiber(domNode);
  if (!fiber) return null;
  var fiberNavigator = new FiberNavigator();
  Object.defineProperty(fiberNavigator, '__fiber', {
    value: fiber,
    enumerable: false,
    writable: false,
    configurable: false
  });
  return fiberNavigator;
});
var _default = FiberNavigator;
exports.default = _default;
