// @stardust-ui/accessibility/dist/commonjs/behaviors/Accordion/accordionBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

/**
 * @specification
 * Adds attribute 'role=presentation' to 'root' slot.
 * Triggers 'moveNext' action with 'ArrowDown' on 'root'.
 * Triggers 'movePrevious' action with 'ArrowUp' on 'root'.
 * Triggers 'moveFirst' action with 'Home' on 'root'.
 * Triggers 'moveLast' action with 'End' on 'root'.
 */
var accordionBehavior = function accordionBehavior() {
  return {
    attributes: {
      root: {
        role: 'presentation'
      }
    },
    keyActions: {
      root: {
        moveNext: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowDown
          }]
        },
        movePrevious: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowUp
          }]
        },
        moveFirst: {
          keyCombinations: [{
            keyCode: keyboardKey.Home
          }]
        },
        moveLast: {
          keyCombinations: [{
            keyCode: keyboardKey.End
          }]
        }
      }
    }
  };
};

var _default = accordionBehavior;
exports.default = _default;
