// @stardust-ui/react/dist/commonjs/components/Grid/Grid.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var Grid =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Grid, _UIComponent);

  function Grid() {
    (0, _classCallCheck2.default)(this, Grid);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Grid).apply(this, arguments));
  }

  (0, _createClass2.default)(Grid, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, _lib.rtlTextContainer.getAttributes({
        forElements: [children, content]
      }), accessibility.attributes.root, unhandledProps), (0, _lib.childrenExist)(children) ? children : content);
    }
  }]);
  return Grid;
}(_lib.UIComponent);
/**
 * A Grid is a layout component that harmonizes negative space, by controlling both the row and column alignment.
 */


(0, _defineProperty2.default)(Grid, "displayName", 'Grid');
(0, _defineProperty2.default)(Grid, "className", 'ui-grid');
(0, _defineProperty2.default)(Grid, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  columns: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  content: customPropTypes.every([customPropTypes.disallow(['children']), PropTypes.oneOfType([PropTypes.arrayOf(customPropTypes.nodeContent), customPropTypes.nodeContent])]),
  rows: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}));
(0, _defineProperty2.default)(Grid, "defaultProps", {
  as: 'div'
});

var _default = (0, _types.withSafeTypeForAs)(Grid);

exports.default = _default;
