// @stardust-ui/react/dist/commonjs/components/Animation/Animation.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _lib = require("../../lib");

var _types = require("../../types");

var Animation =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Animation, _UIComponent);

  function Animation() {
    (0, _classCallCheck2.default)(this, Animation);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Animation).apply(this, arguments));
  }

  (0, _createClass2.default)(Animation, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var children = this.props.children;
      var child = (0, _lib.childrenExist)(children) && React.Children.only(children);
      var result = child ? React.cloneElement(child, {
        className: (0, _classnames.default)(child.props.className, classes.children)
      }) : '';
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, unhandledProps), result);
    }
  }]);
  return Animation;
}(_lib.UIComponent);
/**
 * An Animation provides animation effects to rendered elements.
 */


(0, _defineProperty2.default)(Animation, "create", void 0);
(0, _defineProperty2.default)(Animation, "className", 'ui-animation');
(0, _defineProperty2.default)(Animation, "displayName", 'Animation');
(0, _defineProperty2.default)(Animation, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  accessibility: false,
  animated: false,
  content: false,
  children: 'element'
}), {
  name: PropTypes.string,
  delay: PropTypes.string,
  direction: PropTypes.string,
  duration: PropTypes.string,
  fillMode: PropTypes.string,
  iterationCount: PropTypes.string,
  keyframeParams: PropTypes.object,
  playState: PropTypes.string,
  timingFunction: PropTypes.string
}));

var _default = (0, _types.withSafeTypeForAs)(Animation);

exports.default = _default;
