// @stardust-ui/react/dist/commonjs/themes/teams/components/Attachment/attachmentStyles.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _lib = require("../../../../lib");

var _Icon = _interopRequireDefault(require("../../../../components/Icon/Icon"));

var _getBorderFocusStyles = _interopRequireDefault(require("../../getBorderFocusStyles"));

var _getIconFillOrOutlineStyles = _interopRequireDefault(require("../../getIconFillOrOutlineStyles"));

var attachmentStyles = {
  root: function root(_ref) {
    var p = _ref.props,
        v = _ref.variables,
        siteVariables = _ref.theme.siteVariables;
    return Object.assign({
      position: 'relative',
      display: 'inline-flex',
      alignItems: 'center',
      width: '100%',
      maxWidth: (0, _lib.pxToRem)(440),
      minHeight: (0, _lib.pxToRem)(48),
      padding: v.padding,
      marginBottom: (0, _lib.pxToRem)(2),
      marginRight: (0, _lib.pxToRem)(2),
      background: v.backgroundColor,
      color: v.textColor,
      boxShadow: v.boxShadow,
      border: "".concat(siteVariables.borderWidth, " solid ").concat(v.borderColor),
      borderRadius: v.borderRadius
    }, (0, _getBorderFocusStyles.default)({
      siteVariables: siteVariables,
      borderRadius: v.borderRadius
    }), (p.actionable || p.onClick) && {
      cursor: 'pointer',
      ':hover': {
        background: v.backgroundColorHover,
        color: v.textColorHover
      }
    });
  },
  content: function content() {
    return {
      flex: 1
    };
  },
  header: function header(_ref2) {
    var v = _ref2.variables;
    return {
      fontSize: v.headerFontSize,
      fontWeight: v.headerFontWeight,
      lineHeight: v.headerLineHeight
    };
  },
  description: function description(_ref3) {
    var v = _ref3.variables;
    return {
      display: 'block',
      opacity: 0.5,
      fontSize: v.descriptionFontSize,
      fontWeight: v.descriptionFontWeight,
      lineHeight: v.descriptionLineHeight
    };
  },
  icon: function icon(_ref4) {
    var v = _ref4.variables;
    return {
      flex: '0 0 auto',
      marginRight: v.iconSpace
    };
  },
  action: function action(_ref5) {
    var p = _ref5.props,
        v = _ref5.variables,
        siteVariables = _ref5.theme.siteVariables;
    var iconFilledStyles = (0, _getIconFillOrOutlineStyles.default)({
      outline: false
    });
    var borderFocusStyles = (0, _getBorderFocusStyles.default)({
      siteVariables: siteVariables,
      borderRadius: v.borderRadius
    });
    return Object.assign((0, _defineProperty2.default)({}, "& .".concat(_Icon.default.className), {
      color: v.textColor // this breaks the color change on hover

    }), (0, _getIconFillOrOutlineStyles.default)({
      outline: true
    }), {
      ':hover': iconFilledStyles,
      ':focus': borderFocusStyles[':focus'],
      ':focus-visible': Object.assign({}, iconFilledStyles, borderFocusStyles[':focus-visible'])
    });
  },
  progress: function progress(_ref6) {
    var p = _ref6.props,
        v = _ref6.variables;
    return {
      transition: 'width 0.2s',
      position: 'absolute',
      display: 'block',
      bottom: 0,
      left: 0,
      width: "".concat(p.progress, "%"),
      maxWidth: '100%',
      height: (0, _lib.pxToRem)(v.progressHeight),
      background: v.progressColor
    };
  }
};
var _default = attachmentStyles;
exports.default = _default;
