// @stardust-ui/react/dist/commonjs/themes/teams/components/Attachment/attachmentVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lib = require("../../../../lib");

var _default = function _default(siteVariables) {
  return {
    padding: "".concat((0, _lib.pxToRem)(7), " ").concat((0, _lib.pxToRem)(3), " ").concat((0, _lib.pxToRem)(7), " ").concat((0, _lib.pxToRem)(11)),
    // padding set to 1px less to account for 1px border
    iconSpace: (0, _lib.pxToRem)(12),
    borderColor: siteVariables.colors.grey[200],
    borderRadius: (0, _lib.pxToRem)(3),
    backgroundColor: siteVariables.colors.grey[100],
    backgroundColorHover: siteVariables.colors.grey[200],
    textColor: siteVariables.colors.grey[750],
    textColorHover: siteVariables.colors.grey[750],
    boxShadow: siteVariables.shadowLevel1,
    progressColor: siteVariables.colors.green[200],
    progressHeight: 4,
    headerFontSize: siteVariables.fontSizes.medium,
    headerFontWeight: siteVariables.fontWeightSemibold,
    headerLineHeight: siteVariables.lineHeightMedium,
    descriptionFontSize: siteVariables.fontSizes.small,
    descriptionFontWeight: siteVariables.fontWeightRegular,
    descriptionLineHeight: siteVariables.lineHeightSmall
  };
};

exports.default = _default;
