// @stardust-ui/react/dist/commonjs/themes/teams/animations/timingFunctions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.bounce2 = exports.bounce1 = exports.expoInOut5 = exports.expoOut5 = exports.expoIn5 = exports.easeOutSmooth = exports.easeInSmooth = exports.easeInAlt = exports.easeOut60 = exports.easeIn60 = exports.easeOut70 = exports.easeIn70 = exports.easeOut80 = exports.easeIn80 = exports.easeOut90 = exports.easeIn90 = exports.shotgun = exports.easeEasy = exports.easeOut = exports.easeIn = void 0;
var easeIn = 'cubic-bezier(1.00,0.00,0.78,1.00)';
exports.easeIn = easeIn;
var easeOut = 'cubic-bezier(0.33,0.00,0.00,1.00)';
exports.easeOut = easeOut;
var easeEasy = 'cubic-bezier(0.33,0.00,0.67,1.00)';
exports.easeEasy = easeEasy;
var shotgun = 'cubic-bezier(0.00,0.00,0.00,1.00)';
exports.shotgun = shotgun;
var easeIn90 = 'cubic-bezier(0.90,0.00,0.78,1.00)';
exports.easeIn90 = easeIn90;
var easeOut90 = 'cubic-bezier(0.33,0.00,0.10,1.00)';
exports.easeOut90 = easeOut90;
var easeIn80 = 'cubic-bezier(0.80,0.00,0.78,1.0)';
exports.easeIn80 = easeIn80;
var easeOut80 = 'cubic-bezier(0.33,0.00,0.20,1.00)';
exports.easeOut80 = easeOut80;
var easeIn70 = 'cubic-bezier(0.70,0.00,0.78,1.00)';
exports.easeIn70 = easeIn70;
var easeOut70 = 'cubic-bezier(0.33,0.00,0.30,1.00)';
exports.easeOut70 = easeOut70;
var easeIn60 = 'cubic-bezier(0.60,0.00,0.78,1.00)';
exports.easeIn60 = easeIn60;
var easeOut60 = 'cubic-bezier(0.33,0.00,0.40,1.00)';
exports.easeOut60 = easeOut60;
var easeInAlt = 'cubic-bezier(1.00,0.00,0.67,1.00)';
exports.easeInAlt = easeInAlt;
var easeInSmooth = 'cubic-bezier(0.60,0.00,0.67,1.00)';
exports.easeInSmooth = easeInSmooth;
var easeOutSmooth = 'cubic-bezier(0.60,0.00,0.00,1.00)';
exports.easeOutSmooth = easeOutSmooth;
var expoIn5 = 'cubic-bezier(0.67,0.01,0.83,0.21)';
exports.expoIn5 = expoIn5;
var expoOut5 = 'cubic-bezier(0.17,0.8,0.33,0.99)';
exports.expoOut5 = expoOut5;
var expoInOut5 = 'cubic-bezier(0.78,0,0.22,1)';
exports.expoInOut5 = expoInOut5;
var bounce1 = 'cubic-bezier(0.65, 0.00, 0.26, 1.31)';
exports.bounce1 = bounce1;
var bounce2 = 'cubic-bezier(1.00, 0.00, 0.35, 1.36)';
exports.bounce2 = bounce2;
