// @stardust-ui/react/dist/es/components/Header/HeaderDescription.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as React from 'react';
import { childrenExist, createShorthandFactory, UIComponent, commonPropTypes, rtlTextContainer } from '../../lib';
import { withSafeTypeForAs } from '../../types';

var HeaderDescription =
/*#__PURE__*/
function (_UIComponent) {
  _inherits(HeaderDescription, _UIComponent);

  function HeaderDescription() {
    _classCallCheck(this, HeaderDescription);

    return _possibleConstructorReturn(this, _getPrototypeOf(HeaderDescription).apply(this, arguments));
  }

  _createClass(HeaderDescription, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content;
      return React.createElement(ElementType, _extends({}, rtlTextContainer.getAttributes({
        forElements: [children, content]
      }), accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), childrenExist(children) ? children : content);
    }
  }]);

  return HeaderDescription;
}(UIComponent);

_defineProperty(HeaderDescription, "create", void 0);

_defineProperty(HeaderDescription, "className", 'ui-header__description');

_defineProperty(HeaderDescription, "displayName", 'HeaderDescription');

_defineProperty(HeaderDescription, "propTypes", Object.assign({}, commonPropTypes.createCommon({
  color: true
})));

_defineProperty(HeaderDescription, "defaultProps", {
  as: 'p'
});

HeaderDescription.create = createShorthandFactory({
  Component: HeaderDescription,
  mappedProp: 'content'
});
/**
 * A HeaderDescription provides more detailed information about the Header.
 */

export default withSafeTypeForAs(HeaderDescription);
