// @stardust-ui/react/dist/es/components/Header/Header.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _classCallCheck from "@babel/runtime/helpers/esm/classCallCheck";
import _createClass from "@babel/runtime/helpers/esm/createClass";
import _possibleConstructorReturn from "@babel/runtime/helpers/esm/possibleConstructorReturn";
import _getPrototypeOf from "@babel/runtime/helpers/esm/getPrototypeOf";
import _inherits from "@babel/runtime/helpers/esm/inherits";
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as customPropTypes from '@stardust-ui/react-proptypes';
import * as PropTypes from 'prop-types';
import * as React from 'react';
import { childrenExist, createShorthandFactory, UIComponent, commonPropTypes, rtlTextContainer } from '../../lib';
import HeaderDescription from './HeaderDescription';
import { withSafeTypeForAs } from '../../types';

var Header =
/*#__PURE__*/
function (_UIComponent) {
  _inherits(Header, _UIComponent);

  function Header() {
    _classCallCheck(this, Header);

    return _possibleConstructorReturn(this, _getPrototypeOf(Header).apply(this, arguments));
  }

  _createClass(Header, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          v = _ref.variables,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          description = _this$props.description,
          content = _this$props.content;
      var hasChildren = childrenExist(children);
      var contentElement = childrenExist(children) ? children : content;
      return React.createElement(ElementType, _extends({}, rtlTextContainer.getAttributes({
        forElements: [children, content],
        condition: !description
      }), accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), rtlTextContainer.createFor({
        element: contentElement,
        condition: !!description
      }), !hasChildren && HeaderDescription.create(description, {
        defaultProps: function defaultProps() {
          return {
            className: Header.slotClassNames.description,
            variables: Object.assign({}, v.descriptionColor && {
              color: v.descriptionColor
            })
          };
        }
      }));
    }
  }]);

  return Header;
}(UIComponent);

_defineProperty(Header, "displayName", 'Header');

_defineProperty(Header, "className", 'ui-header');

_defineProperty(Header, "slotClassNames", {
  description: "".concat(Header.className, "__description")
});

_defineProperty(Header, "create", void 0);

_defineProperty(Header, "propTypes", Object.assign({}, commonPropTypes.createCommon({
  color: true
}), {
  description: customPropTypes.itemShorthand,
  align: customPropTypes.align,
  rtlAttributes: PropTypes.func
}));

_defineProperty(Header, "defaultProps", {
  as: 'h1'
});

_defineProperty(Header, "Description", HeaderDescription);

Header.create = createShorthandFactory({
  Component: Header,
  mappedProp: 'content'
});
/**
 * A Header organises the content by declaring a content's topic.
 *
 * @accessibility
 * Headings communicate the organization of the content on the page. Web browsers, plug-ins, and assistive technologies can use them to provide in-page navigation.
 * Nest headings by their rank (or level). The most important heading has the rank 1 (<h1>), the least important heading rank 6 (<h6>). Headings with an equal or higher rank start a new section, headings with a lower rank start new subsections that are part of the higher ranked section.
 *
 * Other considerations:
 *  - when the description property is used in header, readers will narrate both header content and description within the element. In addition to that, both will be displayed in the list of headings.
 */

export default withSafeTypeForAs(Header);
