// @stardust-ui/react/dist/commonjs/components/Dropdown/DropdownSelectedItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _isNil2 = _interopRequireDefault(require("lodash/isNil"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _reactComponentRef = require("@stardust-ui/react-component-ref");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _keyboardKey = _interopRequireDefault(require("keyboard-key"));

var _types = require("../../types");

var _lib = require("../../lib");

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Image = _interopRequireDefault(require("../Image/Image"));

var _Label = _interopRequireDefault(require("../Label/Label"));

var _Box = _interopRequireDefault(require("../Box/Box"));

var DropdownSelectedItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(DropdownSelectedItem, _UIComponent);

  function DropdownSelectedItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, DropdownSelectedItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(DropdownSelectedItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "itemRef", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleKeyDown", function (e) {
      (0, _invoke2.default)(_this.props, 'onKeyDown', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleIconOverrides", function (props) {
      return function (predefinedProps) {
        return Object.assign({}, props, {
          onClick: function onClick(e, iconProps) {
            e.stopPropagation();
            (0, _invoke2.default)(_this.props, 'onRemove', e, _this.props);
            (0, _invoke2.default)(predefinedProps, 'onClick', e, iconProps);
          },
          onKeyDown: function onKeyDown(e, iconProps) {
            e.stopPropagation();

            if (_keyboardKey.default.getCode(e) === _keyboardKey.default.Enter) {
              (0, _invoke2.default)(_this.props, 'onRemove', e, _this.props);
            }

            (0, _invoke2.default)(predefinedProps, 'onKeyDown', e, iconProps);
          }
        });
      };
    });
    return _this;
  }

  (0, _createClass2.default)(DropdownSelectedItem, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      if (!prevProps.active && this.props.active) {
        this.itemRef.current.focus();
      }
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var _this2 = this;

      var unhandledProps = _ref.unhandledProps,
          classes = _ref.classes,
          styles = _ref.styles;
      var _this$props = this.props,
          active = _this$props.active,
          header = _this$props.header,
          icon = _this$props.icon,
          image = _this$props.image;

      var contentElement = _Box.default.create(header, {
        defaultProps: function defaultProps() {
          return {
            as: 'span',
            className: DropdownSelectedItem.slotClassNames.header,
            styles: styles.header
          };
        }
      });

      var renderIcon = (0, _isNil2.default)(icon) ? icon : function (render) {
        return render(icon, function (ComponentType, props) {
          return _Icon.default.create(icon, {
            defaultProps: function defaultProps() {
              return {
                'aria-label': "Remove ".concat(header, " from selection."),
                // TODO: Extract this in a behaviour.
                className: DropdownSelectedItem.slotClassNames.icon,
                styles: styles.icon
              };
            },
            overrideProps: _this2.handleIconOverrides(props)
          });
        });
      };
      var renderImage = (0, _isNil2.default)(image) ? image : function (render) {
        return render(image, function (ComponentType, props) {
          return _Image.default.create(image, {
            defaultProps: function defaultProps() {
              return {
                avatar: true,
                className: DropdownSelectedItem.slotClassNames.image,
                styles: styles.image
              };
            },
            overrideProps: props
          });
        });
      };
      return React.createElement(_reactComponentRef.Ref, {
        innerRef: this.itemRef
      }, React.createElement(_Label.default, (0, _extends2.default)({
        className: classes.root,
        tabIndex: active ? 0 : -1,
        styles: styles.root,
        circular: true,
        onClick: this.handleClick,
        onKeyDown: this.handleKeyDown,
        content: contentElement,
        icon: renderIcon,
        image: renderImage
      }, unhandledProps)));
    }
  }]);
  return DropdownSelectedItem;
}(_lib.UIComponent);

(0, _defineProperty2.default)(DropdownSelectedItem, "displayName", 'DropdownSelectedItem');
(0, _defineProperty2.default)(DropdownSelectedItem, "create", void 0);
(0, _defineProperty2.default)(DropdownSelectedItem, "slotClassNames", void 0);
(0, _defineProperty2.default)(DropdownSelectedItem, "className", 'ui-dropdown__selecteditem');
(0, _defineProperty2.default)(DropdownSelectedItem, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  accessibility: false,
  children: false
}), {
  active: PropTypes.bool,
  header: customPropTypes.itemShorthand,
  icon: customPropTypes.itemShorthandWithoutJSX,
  image: customPropTypes.itemShorthandWithoutJSX,
  onClick: PropTypes.func,
  onKeyDown: PropTypes.func,
  onRemove: PropTypes.func
}));
(0, _defineProperty2.default)(DropdownSelectedItem, "defaultProps", {
  icon: 'close'
});
DropdownSelectedItem.slotClassNames = {
  header: "".concat(DropdownSelectedItem.className, "__header"),
  icon: "".concat(DropdownSelectedItem.className, "__icon"),
  image: "".concat(DropdownSelectedItem.className, "__image")
};
DropdownSelectedItem.create = (0, _lib.createShorthandFactory)({
  Component: DropdownSelectedItem,
  mappedProp: 'header'
});
/**
 * A DropdownSelectedItem represents a selected item of 'multiple-selection' Dropdown.
 */

var _default = (0, _types.withSafeTypeForAs)(DropdownSelectedItem);

exports.default = _default;
