// @stardust-ui/react/dist/commonjs/components/Dropdown/DropdownItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _lib = require("../../lib");

var _types = require("../../types");

var _ListItem = _interopRequireDefault(require("../List/ListItem"));

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Image = _interopRequireDefault(require("../Image/Image"));

var _Box = _interopRequireDefault(require("../Box/Box"));

var DropdownItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(DropdownItem, _UIComponent);

  function DropdownItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, DropdownItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(DropdownItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    return _this;
  }

  (0, _createClass2.default)(DropdownItem, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var classes = _ref.classes,
          styles = _ref.styles,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          content = _this$props.content,
          header = _this$props.header,
          image = _this$props.image,
          accessibilityItemProps = _this$props.accessibilityItemProps,
          selected = _this$props.selected,
          checkable = _this$props.checkable,
          checkableIndicator = _this$props.checkableIndicator;
      return React.createElement(_ListItem.default, (0, _extends2.default)({
        className: DropdownItem.className,
        styles: styles.root,
        onClick: this.handleClick,
        header: _Box.default.create(header, {
          defaultProps: function defaultProps() {
            return {
              className: DropdownItem.slotClassNames.header,
              styles: styles.header
            };
          }
        }),
        media: _Image.default.create(image, {
          defaultProps: function defaultProps() {
            return {
              avatar: true,
              className: DropdownItem.slotClassNames.image,
              styles: styles.image
            };
          }
        }),
        content: _Box.default.create(content, {
          defaultProps: function defaultProps() {
            return {
              className: DropdownItem.slotClassNames.content,
              styles: styles.content
            };
          }
        }),
        endMedia: selected && checkable && {
          content: _Icon.default.create(checkableIndicator, {
            defaultProps: function defaultProps() {
              return {
                className: DropdownItem.slotClassNames.checkableIndicator,
                styles: styles.checkableIndicator
              };
            }
          }),
          styles: styles.endMedia
        },
        truncateContent: true,
        truncateHeader: true
      }, accessibilityItemProps, unhandledProps));
    }
  }]);
  return DropdownItem;
}(_lib.UIComponent);

(0, _defineProperty2.default)(DropdownItem, "displayName", 'DropdownItem');
(0, _defineProperty2.default)(DropdownItem, "create", void 0);
(0, _defineProperty2.default)(DropdownItem, "className", 'ui-dropdown__item');
(0, _defineProperty2.default)(DropdownItem, "slotClassNames", void 0);
(0, _defineProperty2.default)(DropdownItem, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  accessibility: false,
  children: false,
  content: false
}), {
  accessibilityItemProps: PropTypes.object,
  active: PropTypes.bool,
  content: customPropTypes.itemShorthand,
  checkable: PropTypes.bool,
  checkableIndicator: customPropTypes.itemShorthandWithoutJSX,
  header: customPropTypes.itemShorthand,
  image: customPropTypes.itemShorthandWithoutJSX,
  onClick: PropTypes.func,
  isFromKeyboard: PropTypes.bool,
  selected: PropTypes.bool
}));
DropdownItem.slotClassNames = {
  content: "".concat(DropdownItem.className, "__content"),
  header: "".concat(DropdownItem.className, "__header"),
  image: "".concat(DropdownItem.className, "__image"),
  checkableIndicator: "".concat(DropdownItem.className, "__checkable-indicator")
};
DropdownItem.create = (0, _lib.createShorthandFactory)({
  Component: DropdownItem,
  mappedProp: 'header'
});
/**
 * A DropdownItem represents an option of Dropdown list.
 * Displays an item with optional rich media metadata.
 */

var _default = (0, _types.withSafeTypeForAs)(DropdownItem);

exports.default = _default;
