// @stardust-ui/react/dist/commonjs/components/RadioGroup/RadioGroupItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var _reactComponentRef = require("@stardust-ui/react-component-ref");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _lib = require("../../lib");

var _Box = _interopRequireDefault(require("../Box/Box"));

var _types = require("../../types");

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var RadioGroupItem =
/*#__PURE__*/
function (_AutoControlledCompon) {
  (0, _inherits2.default)(RadioGroupItem, _AutoControlledCompon);

  function RadioGroupItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, RadioGroupItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(RadioGroupItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "elementRef", React.createRef());
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      performClick: function performClick(e) {
        e.preventDefault();

        _this.handleClick(e);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    return _this;
  }

  (0, _createClass2.default)(RadioGroupItem, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps, prevState) {
      var checked = this.state.checked;

      if (checked !== prevState.checked) {
        checked && this.props.shouldFocus && this.elementRef.current.focus();
        (0, _invoke2.default)(this.props, 'checkedChanged', undefined, Object.assign({}, this.props, {
          checked: checked
        }));
      }
    }
  }, {
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles,
          accessibility = _ref.accessibility;
      var _this$props = this.props,
          label = _this$props.label,
          icon = _this$props.icon;
      return React.createElement(_reactComponentRef.Ref, {
        innerRef: this.elementRef
      }, React.createElement(ElementType, (0, _extends2.default)({
        onClick: this.handleClick,
        className: classes.root
      }, accessibility.attributes.root, unhandledProps, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), _Icon.default.create(icon || 'stardust-circle', {
        defaultProps: function defaultProps() {
          return {
            size: 'small',
            styles: styles.icon
          };
        }
      }), _Box.default.create(label, {
        defaultProps: function defaultProps() {
          return {
            as: 'span'
          };
        }
      })));
    }
  }]);
  return RadioGroupItem;
}(_lib.AutoControlledComponent);

(0, _defineProperty2.default)(RadioGroupItem, "create", void 0);
(0, _defineProperty2.default)(RadioGroupItem, "displayName", 'RadioGroupItem');
(0, _defineProperty2.default)(RadioGroupItem, "className", 'ui-radiogroup__item');
(0, _defineProperty2.default)(RadioGroupItem, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: false
}), {
  checked: PropTypes.bool,
  defaultChecked: PropTypes.bool,
  disabled: PropTypes.bool,
  icon: customPropTypes.itemShorthandWithoutJSX,
  label: customPropTypes.itemShorthand,
  name: PropTypes.string,
  onClick: PropTypes.func,
  checkedChanged: PropTypes.func,
  shouldFocus: PropTypes.bool,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  vertical: PropTypes.bool
}));
(0, _defineProperty2.default)(RadioGroupItem, "defaultProps", {
  as: 'div',
  accessibility: _accessibility.radioGroupItemBehavior
});
(0, _defineProperty2.default)(RadioGroupItem, "autoControlledProps", ['checked']);
RadioGroupItem.create = (0, _lib.createShorthandFactory)({
  Component: RadioGroupItem,
  mappedProp: 'label'
});
/**
 * A RadioGroupItem represents single input element within a RadioGroup.
 *
 * @accessibility
 * Radio items need to be grouped to correctly handle accessibility.
 */

var _default = (0, _types.withSafeTypeForAs)(RadioGroupItem);

exports.default = _default;
