// @stardust-ui/accessibility/dist/commonjs/behaviors/Menu/menuBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _types = require("../../focusZone/types");

var _menuItemBehavior = _interopRequireDefault(require("./menuItemBehavior"));

var _menuDividerBehavior = _interopRequireDefault(require("./menuDividerBehavior"));

/**
 * @description
 * Implements ARIA Menu design pattern.
 * The 'menu' role is used to identify an element that creates a list of common actions or functions that a user can invoke.
 *
 * @specification
 * Adds role='menu'.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in horizontal direction.
 * When 'vertical' prop is used, provides keyboard navigation in vertical direction.
 * Keyboard navigation is circular.
 */
var menuBehavior = function menuBehavior(props) {
  return {
    attributes: {
      root: {
        role: 'menu'
      }
    },
    focusZone: {
      mode: _types.FocusZoneMode.Embed,
      props: {
        isCircularNavigation: true,
        shouldFocusInnerElementWhenReceivedFocus: true,
        direction: props.vertical ? _types.FocusZoneDirection.vertical : _types.FocusZoneDirection.horizontal
      }
    },
    childBehaviors: {
      item: _menuItemBehavior.default,
      divider: _menuDividerBehavior.default
    }
  };
};

var _default = menuBehavior;
exports.default = _default;
