// @stardust-ui/accessibility/dist/commonjs/behaviors/Menu/submenuBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _types = require("../../focusZone/types");

var _menuItemBehavior = _interopRequireDefault(require("./menuItemBehavior"));

/**
 * @description
 * The 'menu' role is used to identify an element that creates a list of common actions or functions that a user can invoke.
 *
 * @specification
 * Adds role='menu'.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in vertical direction.
 * Keyboard navigation is circular.
 * Component will get focus when mounted.
 */
var submenuBehavior = function submenuBehavior() {
  return {
    attributes: {
      root: {
        role: 'menu'
      }
    },
    focusZone: {
      mode: _types.FocusZoneMode.Embed,
      props: {
        isCircularNavigation: true,
        shouldFocusOnMount: true,
        direction: _types.FocusZoneDirection.vertical
      }
    },
    childBehaviors: {
      item: _menuItemBehavior.default
    }
  };
};

var _default = submenuBehavior;
exports.default = _default;
