// @stardust-ui/react/dist/commonjs/lib/createStardustComponent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _pick2 = _interopRequireDefault(require("lodash/pick"));

var _createComponent = _interopRequireDefault(require("./createComponent"));

var createComponent = function createComponent(_ref) {
  var displayName = _ref.displayName,
      className = _ref.className,
      _render = _ref.render,
      defaultProps = _ref.defaultProps,
      actionHandlers = _ref.actionHandlers;
  return (0, _createComponent.default)({
    displayName: displayName,
    className: className,
    render: function render(config, props) {
      var filteredConfig = (0, _pick2.default)(config, ['accessibility', 'classes', 'rtl', 'styles']);
      return _render(Object.assign({
        stardust: filteredConfig
      }, props));
    },
    defaultProps: defaultProps,
    actionHandlers: actionHandlers
  });
};

var _default = createComponent;
exports.default = _default;
