// @stardust-ui/react/dist/commonjs/lib/createComponent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _keys2 = _interopRequireDefault(require("lodash/keys"));

var React = _interopRequireWildcard(require("react"));

var _reactFela = require("react-fela");

var _renderComponent = _interopRequireDefault(require("./renderComponent"));

var _factories = require("./factories");

// @ts-ignore
var createComponent = function createComponent(_ref) {
  var _ref$displayName = _ref.displayName,
      displayName = _ref$displayName === void 0 ? 'StardustComponent' : _ref$displayName,
      _ref$className = _ref.className,
      className = _ref$className === void 0 ? 'ui-stardust-component' : _ref$className,
      _ref$shorthandPropNam = _ref.shorthandPropName,
      shorthandPropName = _ref$shorthandPropNam === void 0 ? 'children' : _ref$shorthandPropNam,
      _ref$defaultProps = _ref.defaultProps,
      defaultProps = _ref$defaultProps === void 0 ? {} : _ref$defaultProps,
      _ref$handledProps = _ref.handledProps,
      handledProps = _ref$handledProps === void 0 ? [] : _ref$handledProps,
      propTypes = _ref.propTypes,
      actionHandlers = _ref.actionHandlers,
      _render = _ref.render;
  var mergedDefaultProps = Object.assign({
    as: 'div'
  }, defaultProps);

  var StardustComponent = function StardustComponent(props) {
    // Stores debug information for component.
    // Note that this ref should go as the first one, to be discoverable by debug utils.
    var ref = React.useRef(null);
    var context = React.useContext(_reactFela.ThemeContext);
    return (0, _renderComponent.default)({
      className: className,
      displayName: displayName,
      handledProps: (0, _keys2.default)(propTypes).concat(handledProps),
      props: props,
      state: {},
      actionHandlers: actionHandlers,
      render: function render(config) {
        return _render(config, props);
      },
      saveDebug: function saveDebug(stardustDebug) {
        return ref.current = {
          stardustDebug: stardustDebug
        };
      }
    }, context);
  };

  StardustComponent.className = className;
  StardustComponent.create = (0, _factories.createShorthandFactory)({
    Component: mergedDefaultProps.as,
    mappedProp: shorthandPropName
  });
  StardustComponent.displayName = displayName;
  StardustComponent.propTypes = propTypes; // TODO: generate prop types

  StardustComponent.defaultProps = mergedDefaultProps;
  return StardustComponent;
};

var _default = createComponent;
exports.default = _default;
