// @stardust-ui/react/dist/commonjs/lib/UIComponent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _keys2 = _interopRequireDefault(require("lodash/keys"));

var _difference2 = _interopRequireDefault(require("lodash/difference"));

var React = _interopRequireWildcard(require("react"));

var _reactFela = require("react-fela");

var _renderComponent = _interopRequireDefault(require("./renderComponent"));

// @ts-ignore We have this export in package, but it is not present in typings
// TODO @Bugaa92: deprecated by createComponent.tsx
var UIComponent =
/*#__PURE__*/
function (_React$Component) {
  (0, _inherits2.default)(UIComponent, _React$Component);
  (0, _createClass2.default)(UIComponent, null, [{
    key: "handledProps",

    /** Array of props to exclude from list of handled ones. */
    get: function get() {
      if (!this._handledPropsCache) {
        this._handledPropsCache = (0, _difference2.default)((0, _keys2.default)(this.propTypes), this.unhandledProps).sort();
      }

      return this._handledPropsCache;
    }
  }]);

  function UIComponent(props, context) {
    var _this;

    (0, _classCallCheck2.default)(this, UIComponent);
    _this = (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(UIComponent).call(this, props, context));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "childClass", _this.constructor);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "stardustDebug", null);

    if (process.env.NODE_ENV !== 'production') {
      var child = _this.constructor;
      var childName = child.name;

      if (typeof _this.renderComponent !== 'function') {
        throw new Error("".concat(childName, " extending UIComponent is missing a renderComponent() method."));
      }
    }

    _this.renderComponent = _this.renderComponent.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(UIComponent, [{
    key: "renderComponent",
    value: function renderComponent(config) {
      throw new Error('renderComponent is not implemented.');
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return (0, _renderComponent.default)({
        className: this.childClass.className,
        displayName: this.childClass.displayName,
        handledProps: this.childClass.handledProps,
        props: this.props,
        state: this.state,
        actionHandlers: this.actionHandlers,
        render: this.renderComponent,
        saveDebug: function saveDebug(updatedDebug) {
          return _this2.stardustDebug = updatedDebug;
        }
      }, this.context);
    }
  }]);
  return UIComponent;
}(React.Component);

(0, _defineProperty2.default)(UIComponent, "defaultProps", void 0);
(0, _defineProperty2.default)(UIComponent, "displayName", void 0);
(0, _defineProperty2.default)(UIComponent, "className", void 0);
(0, _defineProperty2.default)(UIComponent, "contextType", _reactFela.ThemeContext);
(0, _defineProperty2.default)(UIComponent, "propTypes", void 0);
(0, _defineProperty2.default)(UIComponent, "unhandledProps", []);
(0, _defineProperty2.default)(UIComponent, "_handledPropsCache", undefined);
var _default = UIComponent;
exports.default = _default;
