// @stardust-ui/accessibility/dist/es/behaviors/Toolbar/toolbarMenuItemRadioBehavior.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import * as keyboardKey from 'keyboard-key';
import { IS_FOCUSABLE_ATTRIBUTE } from '../../attributes';

/**
 * @specification
 *  Adds attribute 'aria-checked=true' based on the property 'active'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 *  Adds role='menuitemradio'.
 */
var toolbarMenuItemRadioBehavior = function toolbarMenuItemRadioBehavior(props) {
  var _root;

  return {
    attributes: {
      root: (_root = {}, _defineProperty(_root, IS_FOCUSABLE_ATTRIBUTE, true), _defineProperty(_root, 'aria-checked', props.active), _defineProperty(_root, 'aria-disabled', props.disabled), _defineProperty(_root, "role", 'menuitemradio'), _root)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Enter
          }, {
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

export default toolbarMenuItemRadioBehavior;
