// @stardust-ui/react-component-nesting-registry/dist/commonjs/hooks/useNestingRoot.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var React = _interopRequireWildcard(require("react"));

var _NestingContext = _interopRequireDefault(require("../NestingContext"));

var _RefStack = _interopRequireDefault(require("../lib/RefStack"));

var registrySet = new _RefStack.default();

var useNestingRoot = function useNestingRoot() {
  var _React$useState = React.useState(registrySet),
      _React$useState2 = (0, _slicedToArray2.default)(_React$useState, 1),
      registry = _React$useState2[0];

  var parentRef = React.useRef(null);
  var nestedProps = React.useMemo(function () {
    return {
      value: {
        getContextRefs: registry.getContextRefs,
        register: registry.register,
        unregister: registry.unregister
      }
    };
  }, []);
  var getRefs = React.useCallback(function () {
    return registry.getContextRefs(parentRef);
  }, []);
  React.useEffect(function () {
    registry.register(parentRef);
    return function () {
      return registry.unregister(parentRef);
    };
  }, []);
  return {
    NestedComponent: _NestingContext.default.Provider,
    nestedProps: nestedProps,
    getRefs: getRefs,
    isRoot: true,
    ref: parentRef
  };
};

var _default = useNestingRoot;
exports.default = _default;
