// @stardust-ui/react-component-nesting-registry/dist/commonjs/hooks/useNestingChild.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var React = _interopRequireWildcard(require("react"));

var _NestingContext = _interopRequireDefault(require("../NestingContext"));

var useNestingChild = function useNestingChild() {
  var nestingContext = React.useContext(_NestingContext.default);
  var childRef = React.useRef(null);
  var getRefs = React.useCallback(function () {
    return nestingContext.getContextRefs(childRef);
  }, []);
  React.useEffect(function () {
    nestingContext.register(childRef);
    return function () {
      return nestingContext.unregister(childRef);
    };
  }, []);
  return {
    NestedComponent: React.Fragment,
    nestedProps: null,
    getRefs: getRefs,
    isRoot: false,
    ref: childRef
  };
};

var _default = useNestingChild;
exports.default = _default;
