// @stardust-ui/react/dist/commonjs/components/Tooltip/TooltipContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var _lib = require("../../lib");

var _types = require("../../types");

var TooltipContent =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(TooltipContent, _UIComponent);

  function TooltipContent() {
    (0, _classCallCheck2.default)(this, TooltipContent);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(TooltipContent).apply(this, arguments));
  }

  (0, _createClass2.default)(TooltipContent, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content,
          open = _this$props.open,
          pointing = _this$props.pointing,
          pointerRef = _this$props.pointerRef;
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, _lib.rtlTextContainer.getAttributes({
        forElements: [children, content]
      }), accessibility.attributes.root, unhandledProps), open && pointing && React.createElement("div", {
        className: classes.pointer,
        ref: pointerRef
      }), React.createElement("div", {
        className: classes.content
      }, (0, _lib.childrenExist)(children) ? children : content));
    }
  }]);
  return TooltipContent;
}(_lib.UIComponent);

(0, _defineProperty2.default)(TooltipContent, "create", void 0);
(0, _defineProperty2.default)(TooltipContent, "displayName", 'TooltipContent');
(0, _defineProperty2.default)(TooltipContent, "className", 'ui-tooltip__content');
(0, _defineProperty2.default)(TooltipContent, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon(), {
  placement: PropTypes.string,
  pointing: PropTypes.bool,
  pointerRef: customPropTypes.ref
}));
TooltipContent.create = (0, _lib.createShorthandFactory)({
  Component: TooltipContent,
  mappedProp: 'content'
});
/**
 * A TooltipContent contains the content of a Tooltip component.
 */

var _default = (0, _types.withSafeTypeForAs)(TooltipContent);

exports.default = _default;
