// @stardust-ui/accessibility/dist/commonjs/behaviors/Checkbox/checkboxBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

/**
 * @specification
 * Adds role='checkbox'. This allows screen readers to handle the component as a checkbox button.
 * Adds attribute 'aria-checked=true' based on the property 'checked'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 * Adds attribute 'tabIndex=0' to 'root' slot.
 */
var checkboxBehavior = function checkboxBehavior(props) {
  return {
    attributes: {
      root: {
        'aria-checked': !!props.checked,
        'aria-disabled': props.disabled,
        role: 'checkbox',
        tabIndex: 0
      }
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

var _default = checkboxBehavior;
exports.default = _default;
