// @stardust-ui/react/dist/commonjs/components/Accordion/AccordionTitle.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _noop2 = _interopRequireDefault(require("lodash/noop"));

var _accessibility = require("@stardust-ui/accessibility");

var _reactComponentRef = require("@stardust-ui/react-component-ref");

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Layout = _interopRequireDefault(require("../Layout/Layout"));

var AccordionTitle =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(AccordionTitle, _UIComponent);

  function AccordionTitle() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, AccordionTitle);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(AccordionTitle)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionHandlers", {
      performClick: function performClick(e) {
        e.preventDefault();

        _this.handleClick(e);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleFocus", function (e) {
      e.stopPropagation();
      (0, _invoke2.default)(_this.props, 'onFocus', e, _this.props);
    });
    return _this;
  }

  (0, _createClass2.default)(AccordionTitle, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          styles = _ref.styles,
          accessibility = _ref.accessibility;
      var _this$props = this.props,
          contentRef = _this$props.contentRef,
          children = _this$props.children,
          content = _this$props.content,
          indicator = _this$props.indicator,
          active = _this$props.active;
      var defaultIndicator = {
        name: active ? 'stardust-arrow-down' : 'stardust-arrow-end'
      };
      var indicatorWithDefaults = indicator === undefined ? defaultIndicator : indicator;
      var contentElement = React.createElement(_reactComponentRef.Ref, {
        innerRef: contentRef
      }, React.createElement(_Layout.default, (0, _extends2.default)({
        onFocus: this.handleFocus,
        onClick: this.handleClick,
        className: AccordionTitle.slotClassNames.content
      }, accessibility.attributes.content, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.content, unhandledProps), {
        start: _Icon.default.create(indicatorWithDefaults, {
          defaultProps: function defaultProps() {
            return {
              styles: styles.indicator
            };
          }
        }),
        main: _lib.rtlTextContainer.createFor({
          element: content
        })
      })));
      return React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, _lib.rtlTextContainer.getAttributes({
        forElements: [children]
      }), accessibility.attributes.root, unhandledProps, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), (0, _lib.childrenExist)(children) ? children : contentElement);
    }
  }]);
  return AccordionTitle;
}(_lib.UIComponent);

(0, _defineProperty2.default)(AccordionTitle, "displayName", 'AccordionTitle');
(0, _defineProperty2.default)(AccordionTitle, "create", void 0);
(0, _defineProperty2.default)(AccordionTitle, "className", 'ui-accordion__title');
(0, _defineProperty2.default)(AccordionTitle, "slotClassNames", void 0);
(0, _defineProperty2.default)(AccordionTitle, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon(), {
  accordionContentId: PropTypes.string,
  active: PropTypes.bool,
  contentRef: customPropTypes.ref,
  canBeCollapsed: PropTypes.bool,
  index: PropTypes.number,
  onClick: PropTypes.func,
  indicator: customPropTypes.itemShorthand
}));
(0, _defineProperty2.default)(AccordionTitle, "defaultProps", {
  accessibility: _accessibility.accordionTitleBehavior,
  as: 'dt',
  contentRef: _noop2.default
});
AccordionTitle.create = (0, _lib.createShorthandFactory)({
  Component: AccordionTitle,
  mappedProp: 'content'
});
AccordionTitle.slotClassNames = {
  content: "".concat(AccordionTitle.className, "__content")
  /**
   * An AccordionTitle represents the title of Accordion's item that can be interacted with to expand or collapse the item's content.
   */

};

var _default = (0, _types.withSafeTypeForAs)(AccordionTitle);

exports.default = _default;
