// @stardust-ui/react/dist/commonjs/components/Accordion/AccordionContent.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var AccordionContent =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(AccordionContent, _UIComponent);

  function AccordionContent() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, AccordionContent);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(AccordionContent)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleClick", function (e) {
      (0, _invoke2.default)(_this.props, 'onClick', e, _this.props);
    });
    return _this;
  }

  (0, _createClass2.default)(AccordionContent, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          accessibility = _ref.accessibility;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content;
      return React.createElement(ElementType, (0, _extends2.default)({
        onClick: this.handleClick
      }, _lib.rtlTextContainer.getAttributes({
        forElements: [children, content]
      }), accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), (0, _lib.childrenExist)(children) ? children : content);
    }
  }]);
  return AccordionContent;
}(_lib.UIComponent);

(0, _defineProperty2.default)(AccordionContent, "displayName", 'AccordionContent');
(0, _defineProperty2.default)(AccordionContent, "create", void 0);
(0, _defineProperty2.default)(AccordionContent, "className", 'ui-accordion__content');
(0, _defineProperty2.default)(AccordionContent, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon(), {
  accordionTitleId: PropTypes.string,
  active: PropTypes.bool,
  onClick: PropTypes.func
}));
(0, _defineProperty2.default)(AccordionContent, "defaultProps", {
  accessibility: _accessibility.accordionContentBehavior,
  as: 'dd'
});
AccordionContent.create = (0, _lib.createShorthandFactory)({
  Component: AccordionContent,
  mappedProp: 'content'
});
/**
 * An AccordionContent displays content hosted in an Accordion.
 */

var _default = (0, _types.withSafeTypeForAs)(AccordionContent);

exports.default = _default;
