// @stardust-ui/react/dist/commonjs/themes/teams/components/Provider/providerBoxStyles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _default = {
  root: function root(_ref) {
    var v = _ref.variables;
    return {
      background: v.background,
      color: v.color,
      textAlign: 'left',
      '& ::-webkit-scrollbar': {
        height: v.scrollbarHeight,
        width: v.scrollbarWidth,
        ':disabled': {
          display: 'none'
        }
      },
      '& ::-webkit-scrollbar-thumb': {
        borderRadius: v.scrollbarThumbBorderRadius,
        border: "solid ".concat(v.scrollbarThumbBorderSize, " transparent"),
        backgroundClip: 'content-box',
        backgroundColor: v.scrollbarThumbBackgroundColor,
        ':hover': {
          backgroundColor: v.scrollbarThumbHoverBackgroundColor,
          border: "solid ".concat(v.scrollbarThumbHoverBorderSize, " transparent")
        }
      },
      '& ::-webkit-scrollbar-track': {
        background: 'transparent'
      }
    };
  }
};
exports.default = _default;
