// @stardust-ui/react/dist/commonjs/themes/teams-dark/components/Menu/menuVariables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _menuVariables = require("../../../teams/components/Menu/menuVariables");

var _colorUtils = require("../../../colorUtils");

var _default = function _default(siteVars) {
  return {
    colorScheme: (0, _colorUtils.pickValuesFromColorScheme)((0, _colorUtils.extendColorScheme)(siteVars.colorScheme, {
      default: {
        borderActive: siteVars.colors.grey[600],
        backgroundFocus: siteVars.colors.grey[500],
        backgroundActive: siteVars.colorScheme.default.backgroundActive1,
        foregroundDisabled: siteVars.colorScheme.default.foregroundDisabled1
      },
      brand: {
        foregroundHover: siteVars.colors.white,
        backgroundHover: siteVars.colors.brand[300],
        borderActive: siteVars.colors.brand[400],
        foregroundActive: siteVars.colors.white,
        foregroundFocus: siteVars.colors.white,
        backgroundFocus: siteVars.colors.brand[300],
        foregroundDisabled: siteVars.colorScheme.brand.foregroundDisabled1
      }
    }), _menuVariables.menuColorAreas),
    color: siteVars.colors.grey[250],
    colorActive: siteVars.colors.white,
    primaryBorderColor: siteVars.colors.grey[600],
    pointingIndicatorBackgroundColor: siteVars.colors.brand[400],
    verticalBackgroundColor: siteVars.colors.grey[600],
    verticalBackgroundColorFocus: siteVars.colors.grey[550],
    iconOnlyColorActive: siteVars.colors.brand[400]
  };
};

exports.default = _default;
