// @stardust-ui/accessibility/dist/commonjs/behaviors/Radio/radioGroupBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

/**
 * @description
 * Implements ARIA Radio Group design pattern.
 * @specification
 *  Adds role='radiogroup'. This allows screen readers to handle the component as a radio group.
 *  Triggers 'nextItem' action with 'ArrowDown' or 'ArrowRight' on 'root'.
 *  Triggers 'prevItem' action with 'ArrowUp' or 'ArrowLeft' on 'root'.
 */
var radioGroupBehavior = function radioGroupBehavior() {
  return {
    attributes: {
      root: {
        role: 'radiogroup'
      }
    },
    keyActions: {
      root: {
        nextItem: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowDown
          }, {
            keyCode: keyboardKey.ArrowRight
          }]
        },
        prevItem: {
          keyCombinations: [{
            keyCode: keyboardKey.ArrowUp
          }, {
            keyCode: keyboardKey.ArrowLeft
          }]
        }
      }
    }
  };
};

var _default = radioGroupBehavior;
exports.default = _default;
