// @stardust-ui/accessibility/dist/commonjs/behaviors/Radio/radioGroupItemBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

/**
 * @specification
 *  Adds role='radio'. This allows screen readers to handle the component as a radio button.
 *  Adds attribute 'aria-checked=true' based on the property 'checked'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabled'. This can be overriden by providing 'aria-disabled' property directly to the component.
 *  Triggers 'performClick' action with 'Spacebar' on 'root'.
 *  Implements roving tabIndex.
 */
var radioGroupItemBehavior = function radioGroupItemBehavior(props) {
  return {
    attributes: {
      root: {
        role: 'radio',
        tabIndex: props.checked ? 0 : -1,
        'aria-checked': props.checked,
        'aria-disabled': props.disabled
      }
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

var _default = radioGroupItemBehavior;
exports.default = _default;
