// @stardust-ui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarItemBehavior.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

var _buttonBehavior = _interopRequireDefault(require("../Button/buttonBehavior"));

/**
 * @specification
 * Adds role='button' if element type is other than 'button'. This allows screen readers to handle the component as a button.
 * Adds attribute 'tabIndex=0' if element type is other than 'button'.
 * Adds attribute 'aria-disabled=true' based on the property 'disabled'. This can be overriden by providing 'aria-disabled' property directly to the component.
 * Adds attribute 'aria-haspopup=menu' to 'root' slot if 'menu' property is set.
 * Adds attribute 'aria-haspopup=dialog' to 'root' slot if 'popup' property is set.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 * Triggers 'closeMenuAndFocusTrigger' action with 'Escape' on 'wrapper', when toolbar button has submenu and it is opened.
 * Triggers 'doNotNavigateNextToolbarItem' action with 'ArrowLeft' or 'ArrowRight' on 'wrapper', when toolbar button has submenu and it is opened.
 */
var toolbarItemBehavior = function toolbarItemBehavior(props) {
  var behaviorData = (0, _buttonBehavior.default)(props);
  behaviorData.attributes.root = Object.assign({}, behaviorData.attributes.root, {
    'aria-haspopup': props.popup ? 'dialog' : props.menu ? 'menu' : undefined
  });
  behaviorData.keyActions.wrapper = Object.assign({}, behaviorData.keyActions.wrapper, {
    performWrapperClick: {
      keyCombinations: [{
        keyCode: keyboardKey.Enter
      }, {
        keyCode: keyboardKey.Spacebar
      }]
    },
    closeMenuAndFocusTrigger: {
      keyCombinations: props.menu && props.menuOpen ? [{
        keyCode: keyboardKey.Escape
      }, {
        keyCode: keyboardKey.Tab,
        shiftKey: true
      }] : null
    },
    doNotNavigateNextToolbarItem: {
      keyCombinations: props.menu && props.menuOpen ? [{
        keyCode: keyboardKey.ArrowLeft
      }, {
        keyCode: keyboardKey.ArrowRight
      }] : null
    }
  });
  return behaviorData;
};

var _default = toolbarItemBehavior;
exports.default = _default;
