// @stardust-ui/accessibility/dist/commonjs/behaviors/Toolbar/toolbarMenuItemRadioBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

var _attributes = require("../../attributes");

/**
 * @specification
 *  Adds attribute 'aria-checked=true' based on the property 'active'.
 *  Adds attribute 'aria-disabled=true' based on the property 'disabled'.
 *  Adds role='menuitemradio'.
 */
var toolbarMenuItemRadioBehavior = function toolbarMenuItemRadioBehavior(props) {
  var _root;

  return {
    attributes: {
      root: (_root = {}, (0, _defineProperty2.default)(_root, _attributes.IS_FOCUSABLE_ATTRIBUTE, true), (0, _defineProperty2.default)(_root, 'aria-checked', props.active), (0, _defineProperty2.default)(_root, 'aria-disabled', props.disabled), (0, _defineProperty2.default)(_root, "role", 'menuitemradio'), _root)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Enter
          }, {
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

var _default = toolbarMenuItemRadioBehavior;
exports.default = _default;
