// @stardust-ui/react/dist/es/themes/teams/components/Dropdown/dropdownVariables.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import { pxToRem } from '../../../../lib';

var _map = [3, 12].map(function (v) {
  return pxToRem(v);
}),
    _map2 = _slicedToArray(_map, 2),
    cornerRadius = _map2[0],
    _12px_asRem = _map2[1];

export default (function (siteVars) {
  return {
    backgroundColor: siteVars.colors.grey[100],
    backgroundColorHover: siteVars.colors.grey[150],
    borderColor: 'transparent',
    borderColorFocus: siteVars.colors.brand[600],
    borderWidth: '0px',
    containerBorderRadius: "".concat(cornerRadius),
    openAboveContainerBorderRadius: "0 0 ".concat(cornerRadius, " ").concat(cornerRadius),
    openBelowContainerBorderRadius: "".concat(cornerRadius, " ").concat(cornerRadius, " 0 0"),
    searchBorderBottomWidth: pxToRem(2),
    color: siteVars.bodyColor,
    selectedItemColor: siteVars.bodyColor,
    comboboxPaddingButton: "0 ".concat(_12px_asRem),
    comboboxFlexBasis: pxToRem(50),
    aboveListBorderRadius: "".concat(cornerRadius, " ").concat(cornerRadius, " 0 0"),
    belowListBorderRadius: "0 0 ".concat(cornerRadius, " ").concat(cornerRadius),
    listBackgroundColor: siteVars.colors.white,
    listBorderColor: 'transparent',
    listBorderWidth: '0px',
    listPadding: "".concat(pxToRem(8), " 0 ").concat(pxToRem(6)),
    listBoxShadow: siteVars.shadowLevel3,
    listMaxHeight: pxToRem(296),
    listItemFocusBorderWidth: pxToRem(1),
    listItemBackgroundColor: siteVars.colors.white,
    listItemColorHover: siteVars.colors.grey[750],
    listItemBackgroundColorHover: siteVars.colors.grey[100],
    listItemBackgroundColorActive: siteVars.colors.grey[100],
    listItemColorActive: siteVars.colors.grey[750],
    listItemSelectedFontWeight: siteVars.fontWeightSemibold,
    listItemSelectedColor: siteVars.colors.grey[750],
    selectedItemBackgroundColor: 'undefined',
    selectedItemColorFocus: siteVars.bodyColor,
    selectedItemBackgroundColorFocus: siteVars.colors.brand[200],
    selectedItemsMaxHeight: pxToRem(82),
    toggleIndicatorSize: pxToRem(32),
    triggerButtonColorHover: siteVars.bodyColor,
    width: pxToRem(356),
    // these should only apply when there is content in the image/media slot:
    listItemHeaderFontSize: siteVars.fontSizes.medium,
    listItemHeaderColor: siteVars.colors.grey[1000],
    listItemContentFontSize: siteVars.fontSizes.small,
    listItemContentColor: siteVars.colors.grey[450]
  };
});
