// @stardust-ui/react/dist/commonjs/components/Toolbar/ToolbarCustomItem.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var _accessibility = require("@stardust-ui/accessibility");

var React = _interopRequireWildcard(require("react"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var _lib = require("../../lib");

var _types = require("../../types");

var ToolbarCustomItem =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(ToolbarCustomItem, _UIComponent);

  function ToolbarCustomItem() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ToolbarCustomItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ToolbarCustomItem)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleBlur", function (e) {
      if (_this.props.focusable) {
        (0, _invoke2.default)(_this.props, 'onBlur', e, _this.props);
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleFocus", function (e) {
      if (_this.props.focusable) {
        (0, _invoke2.default)(_this.props, 'onFocus', e, _this.props);
      }
    });
    return _this;
  }

  (0, _createClass2.default)(ToolbarCustomItem, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          variables = _ref.variables,
          accessibility = _ref.accessibility,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          content = _this$props.content,
          focusable = _this$props.focusable;
      return React.createElement(ElementType, (0, _extends2.default)({}, accessibility.attributes.root, (0, _defineProperty2.default)({}, _accessibility.IS_FOCUSABLE_ATTRIBUTE, focusable), unhandledProps, {
        className: classes.root,
        onBlur: this.handleBlur,
        onFocus: this.handleFocus
      }), (0, _lib.childrenExist)(children) ? children : content);
    }
  }]);
  return ToolbarCustomItem;
}(_lib.UIComponent);

(0, _defineProperty2.default)(ToolbarCustomItem, "displayName", 'ToolbarCustomItem');
(0, _defineProperty2.default)(ToolbarCustomItem, "className", 'ui-toolbar__customitem');
(0, _defineProperty2.default)(ToolbarCustomItem, "create", void 0);
(0, _defineProperty2.default)(ToolbarCustomItem, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon(), {
  fitted: PropTypes.oneOfType([PropTypes.bool, PropTypes.oneOf(['horizontally', 'vertically'])]),
  focusable: PropTypes.bool,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func
}));
ToolbarCustomItem.create = (0, _lib.createShorthandFactory)({
  Component: ToolbarCustomItem,
  mappedProp: 'content'
});
/**
 * A ToolbarCustomItem renders Toolbar item as a non-actionable `div` with custom content inside.
 */

var _default = (0, _types.withSafeTypeForAs)(ToolbarCustomItem);

exports.default = _default;
