// @stardust-ui/react/dist/commonjs/components/Toolbar/ToolbarMenuDivider.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var ToolbarMenuDivider =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(ToolbarMenuDivider, _UIComponent);

  function ToolbarMenuDivider() {
    (0, _classCallCheck2.default)(this, ToolbarMenuDivider);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(ToolbarMenuDivider).apply(this, arguments));
  }

  (0, _createClass2.default)(ToolbarMenuDivider, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          accessibility = _ref.accessibility;
      return React.createElement(ElementType, (0, _extends2.default)({}, accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }));
    }
  }]);
  return ToolbarMenuDivider;
}(_lib.UIComponent);

(0, _defineProperty2.default)(ToolbarMenuDivider, "displayName", 'ToolbarMenuDivider');
(0, _defineProperty2.default)(ToolbarMenuDivider, "create", void 0);
(0, _defineProperty2.default)(ToolbarMenuDivider, "className", 'ui-toolbar__menudivider');
(0, _defineProperty2.default)(ToolbarMenuDivider, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon()));
(0, _defineProperty2.default)(ToolbarMenuDivider, "defaultProps", {
  as: 'li'
});
ToolbarMenuDivider.create = (0, _lib.createShorthandFactory)({
  Component: ToolbarMenuDivider,
  mappedProp: 'content'
});
/**
 * A ToolbarMenuDivider adds non-actionable separator between items of ToolbarMenu.
 */

var _default = (0, _types.withSafeTypeForAs)(ToolbarMenuDivider);

exports.default = _default;
