// @stardust-ui/react/dist/commonjs/components/Toolbar/ToolbarMenuRadioGroup.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _map2 = _interopRequireDefault(require("lodash/map"));

var _invoke2 = _interopRequireDefault(require("lodash/invoke"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var PropTypes = _interopRequireWildcard(require("prop-types"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var _accessibility = require("@stardust-ui/accessibility");

var _ToolbarMenuItem = _interopRequireDefault(require("./ToolbarMenuItem"));

var _mergeThemes = require("../../lib/mergeThemes");

var _Box = _interopRequireDefault(require("../Box/Box"));

var ToolbarMenuRadioGroup =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(ToolbarMenuRadioGroup, _UIComponent);

  function ToolbarMenuRadioGroup() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2.default)(this, ToolbarMenuRadioGroup);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2.default)(this, (_getPrototypeOf2 = (0, _getPrototypeOf3.default)(ToolbarMenuRadioGroup)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "handleItemOverrides", function (variables) {
      return function (predefinedProps) {
        return {
          onClick: function onClick(e, itemProps) {
            (0, _invoke2.default)(predefinedProps, 'onClick', e, itemProps);
            (0, _invoke2.default)(_this.props, 'onItemClick', e, itemProps);
          },
          variables: (0, _mergeThemes.mergeComponentVariables)(variables, predefinedProps.variables),
          wrapper: null
        };
      };
    });
    return _this;
  }

  (0, _createClass2.default)(ToolbarMenuRadioGroup, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var _this2 = this;

      var ElementType = _ref.ElementType,
          classes = _ref.classes,
          unhandledProps = _ref.unhandledProps,
          accessibility = _ref.accessibility,
          styles = _ref.styles;
      var _this$props = this.props,
          activeIndex = _this$props.activeIndex,
          items = _this$props.items,
          variables = _this$props.variables,
          wrapper = _this$props.wrapper;
      var content = React.createElement(ElementType, (0, _extends2.default)({
        className: classes.root
      }, accessibility.attributes.root, unhandledProps, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.root, unhandledProps)), (0, _map2.default)(items, function (item, index) {
        return _ToolbarMenuItem.default.create(item, {
          defaultProps: function defaultProps() {
            return {
              accessibility: _accessibility.toolbarMenuItemRadioBehavior,
              as: 'li',
              active: activeIndex === index,
              index: index
            };
          },
          overrideProps: _this2.handleItemOverrides(variables)
        });
      }));
      return _Box.default.create(wrapper, {
        defaultProps: function defaultProps() {
          return Object.assign({
            as: 'li',
            className: ToolbarMenuRadioGroup.slotClassNames.wrapper,
            styles: styles.wrapper
          }, accessibility.attributes.wrapper, (0, _lib.applyAccessibilityKeyHandlers)(accessibility.keyHandlers.wrapper, wrapper));
        },
        overrideProps: {
          children: content
        }
      });
    }
  }]);
  return ToolbarMenuRadioGroup;
}(_lib.UIComponent);

(0, _defineProperty2.default)(ToolbarMenuRadioGroup, "displayName", 'ToolbarMenuRadioGroup');
(0, _defineProperty2.default)(ToolbarMenuRadioGroup, "create", void 0);
(0, _defineProperty2.default)(ToolbarMenuRadioGroup, "className", 'ui-toolbars');
(0, _defineProperty2.default)(ToolbarMenuRadioGroup, "slotClassNames", {
  wrapper: "".concat(ToolbarMenuRadioGroup.className, "__wrapper")
});
(0, _defineProperty2.default)(ToolbarMenuRadioGroup, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon(), {
  activeIndex: PropTypes.number,
  items: customPropTypes.collectionShorthand,
  onItemClick: PropTypes.func,
  wrapper: customPropTypes.itemShorthand
}));
(0, _defineProperty2.default)(ToolbarMenuRadioGroup, "defaultProps", {
  as: 'ul',
  accessibility: _accessibility.toolbarMenuRadioGroupBehavior,
  wrapper: {}
});
ToolbarMenuRadioGroup.create = (0, _lib.createShorthandFactory)({
  Component: ToolbarMenuRadioGroup
});
/**
 * A ToolbarMenuRadioGroup renders ToolbarMenuItem as a group of mutually exclusive options.
 */

var _default = (0, _types.withSafeTypeForAs)(ToolbarMenuRadioGroup);

exports.default = _default;
