// @stardust-ui/docs-components/dist/commonjs/knobs/useRangeKnob.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _parseRangeValue = _interopRequireDefault(require("./lib/parseRangeValue"));

var _useKnob = _interopRequireDefault(require("./useKnob"));

var useRangeKnob = function useRangeKnob(options) {
  var _options$initialValue = options.initialValue,
      initialValue = _options$initialValue === void 0 ? 3 : _options$initialValue,
      _options$min = options.min,
      min = _options$min === void 0 ? 0 : _options$min,
      _options$max = options.max,
      max = _options$max === void 0 ? (0, _parseRangeValue.default)(initialValue) : _options$max,
      _options$step = options.step,
      step = _options$step === void 0 ? 1 : _options$step,
      rest = (0, _objectWithoutProperties2.default)(options, ["initialValue", "min", "max", "step"]);
  var unit = "".concat(initialValue).replace("".concat((0, _parseRangeValue.default)(initialValue)), '');
  return (0, _useKnob.default)(Object.assign({
    initialValue: initialValue,
    unit: unit,
    min: (0, _parseRangeValue.default)(min),
    max: (0, _parseRangeValue.default)(max),
    step: (0, _parseRangeValue.default)(step),
    type: 'range'
  }, rest));
};

var _default = useRangeKnob;
exports.default = _default;
