// @stardust-ui/accessibility/dist/commonjs/behaviors/List/selectableListBehavior.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _types = require("../../focusZone/types");

/**
 * @description
 * The listbox role is used to identify an element that creates a list from which a user may select one or more items.
 *
 * @specification
 * Adds role='listbox'.
 * Adds attribute 'tabIndex=-1' to 'root' slot.
 * Adds attribute 'aria-orientation=horizontal' to 'root' slot if 'horizontal' property is true. Does not set the attribute otherwise.
 * Embeds component into FocusZone.
 * Provides arrow key navigation in bidirectionalDomOrder direction.
 */
var selectableListBehavior = function selectableListBehavior(props) {
  return {
    attributes: {
      root: Object.assign({
        role: 'listbox',
        tabIndex: -1
      }, props.horizontal && {
        'aria-orientation': 'horizontal'
      })
    },
    focusZone: {
      mode: _types.FocusZoneMode.Embed,
      props: {
        shouldFocusInnerElementWhenReceivedFocus: true,
        direction: _types.FocusZoneDirection.bidirectionalDomOrder
      }
    }
  };
};

var _default = selectableListBehavior;
exports.default = _default;
