// @stardust-ui/accessibility/dist/commonjs/behaviors/List/selectableListItemBehavior.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var keyboardKey = _interopRequireWildcard(require("keyboard-key"));

var _attributes = require("../../attributes");

/**
 * @specification
 * Adds role='option'. This role is used for a selectable item in a list.
 * Adds attribute 'data-is-focusable=true' to 'root' slot.
 * Adds attribute 'aria-selected=true' based on the property 'selected'. Based on this screen readers will recognize the selected state of the item.
 * Triggers 'performClick' action with 'Enter' or 'Spacebar' on 'root'.
 */
var selectableListItemBehavior = function selectableListItemBehavior(props) {
  return {
    attributes: {
      root: (0, _defineProperty2.default)({
        role: 'option',
        'aria-selected': !!props.selected
      }, _attributes.IS_FOCUSABLE_ATTRIBUTE, true)
    },
    keyActions: {
      root: {
        performClick: {
          keyCombinations: [{
            keyCode: keyboardKey.Enter
          }, {
            keyCode: keyboardKey.Spacebar
          }]
        }
      }
    }
  };
};

var _default = selectableListItemBehavior;
exports.default = _default;
