// @stardust-ui/react/dist/commonjs/components/Reaction/Reaction.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var customPropTypes = _interopRequireWildcard(require("@stardust-ui/react-proptypes"));

var React = _interopRequireWildcard(require("react"));

var _lib = require("../../lib");

var _types = require("../../types");

var _Icon = _interopRequireDefault(require("../Icon/Icon"));

var _Box = _interopRequireDefault(require("../Box/Box"));

var _ReactionGroup = _interopRequireDefault(require("./ReactionGroup"));

var Reaction =
/*#__PURE__*/
function (_UIComponent) {
  (0, _inherits2.default)(Reaction, _UIComponent);

  function Reaction() {
    (0, _classCallCheck2.default)(this, Reaction);
    return (0, _possibleConstructorReturn2.default)(this, (0, _getPrototypeOf2.default)(Reaction).apply(this, arguments));
  }

  (0, _createClass2.default)(Reaction, [{
    key: "renderComponent",
    value: function renderComponent(_ref) {
      var accessibility = _ref.accessibility,
          ElementType = _ref.ElementType,
          classes = _ref.classes,
          styles = _ref.styles,
          unhandledProps = _ref.unhandledProps;
      var _this$props = this.props,
          children = _this$props.children,
          icon = _this$props.icon,
          content = _this$props.content;
      return React.createElement(ElementType, (0, _extends2.default)({}, _lib.rtlTextContainer.getAttributes({
        forElements: [children]
      }), accessibility.attributes.root, unhandledProps, {
        className: classes.root
      }), (0, _lib.childrenExist)(children) ? children : React.createElement(React.Fragment, null, _Icon.default.create(icon, {
        defaultProps: function defaultProps() {
          return {
            className: Reaction.slotClassNames.icon,
            styles: styles.icon
          };
        }
      }), _Box.default.create(content, {
        defaultProps: function defaultProps() {
          return {
            className: Reaction.slotClassNames.content,
            styles: styles.content
          };
        }
      })));
    }
  }]);
  return Reaction;
}(_lib.UIComponent);

(0, _defineProperty2.default)(Reaction, "create", void 0);
(0, _defineProperty2.default)(Reaction, "className", 'ui-reaction');
(0, _defineProperty2.default)(Reaction, "slotClassNames", void 0);
(0, _defineProperty2.default)(Reaction, "displayName", 'Reaction');
(0, _defineProperty2.default)(Reaction, "propTypes", Object.assign({}, _lib.commonPropTypes.createCommon({
  content: 'shorthand'
}), {
  icon: customPropTypes.itemShorthandWithoutJSX
}));
(0, _defineProperty2.default)(Reaction, "defaultProps", {
  as: 'span'
});
(0, _defineProperty2.default)(Reaction, "Group", _ReactionGroup.default);
Reaction.create = (0, _lib.createShorthandFactory)({
  Component: Reaction,
  mappedProp: 'content'
});
Reaction.slotClassNames = {
  icon: "".concat(Reaction.className, "__icon"),
  content: "".concat(Reaction.className, "__content")
  /**
   * A Reaction indicates user's emotion or perception.
   * Used to display user's reaction for entity in Chat (e.g. message).
   */

};

var _default = (0, _types.withSafeTypeForAs)(Reaction);

exports.default = _default;
